// @stratio/egeo/esm5/lib/st-dynamic-table/st-dynamic-table.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/st-dynamic-table/st-dynamic-table.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
/*
 * © 2017 Stratio Big Data Inc., Sucursal en España.
 *
 * This software is licensed under the Apache License, Version 2.0.
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the terms of the License for more details.
 *
 * SPDX-License-Identifier: Apache-2.0.
 */
import { ChangeDetectionStrategy, ChangeDetectorRef, Component, EventEmitter, Input, Output, TemplateRef } from '@angular/core';
import { StEgeo, StRequired } from '../decorators/require-decorators';
import { Order } from '../st-table/shared/order';
import { StDynamicTableUtils } from './utils/st-dynamic-table.utils';
import { StTableIconClasses } from '../st-table/st-table.interface';
/**
 * \@description {Component} [Dynamic Table]
 *
 * The table component has been designed to be able to create a table deducing its columns using a json schema
 *
 * * \@model
 *
 *   [StDynamicTableHeader] {./st-dynamic-table.model.ts#StDynamicTableHeader}
 *   [StDynamicTableUISpecification] {./st-dynamic-table.model.ts#StDynamicTableUISpecification}
 *   [StDynamicTableUserInterface] {./st-dynamic-table.model.ts#StDynamicTableUserInterface}
 *
 * \@example
 *
 * {html}
 *
 * ```
 * <st-dynamic-table [jsonSchema]="jsonSchema"
 *    [items]="sortedUsers"
 *    [sortable]="true"
 *    [filterable]="true"
 *    [hasHoverMenu]="true"
 *    [currentOrder]="currentOrder"
 *    [activeHoverMenu]="activeHoverMenuPosition"
 *    [templateContentFilter]="filterContent"
 *    (changeOrder)="onSortTable($event)"
 *    (showHoverMenu)="onShowHoverMenu($event)">
 *        <st-dropdown-menu st-dynamic-table-hover-menu
 *          class="hover-menu"
 *          [items]="rowActions"
 *          [active]="activeHoverMenuPosition[3] !== undefined"
 *          [visualMode]="1">
 *       </st-dropdown-menu>
 * </st-dynamic-table>
 * ```
 *
 */
var StDynamicTableComponent = /** @class */ (function () {
    function StDynamicTableComponent(_cd) {
        this._cd = _cd;
        /**
         * \@Input {boolean} [header=true] Boolean to show or hide the header
         */
        this.header = true;
        /**
         * \@Input {boolean} [sortable=true] Boolean to make sortable the table, To enable sorting of columns use
         * the new "sortable" field inside stTableHeader model
         */
        this.sortable = true;
        /**
         * \@Input {boolean} [filterable=false] Boolean to make filterable the table, To enable filtering of columns use
         * the new "filterable" field inside stTableHeader model (necesary define filterConfig).
         */
        this.filterable = false;
        /**
         * \@Input {boolean} [selectable=false] Boolean to show or hide a checkboxes in the first cell of rows
         */
        this.selectable = false;
        /**
         * \@Input {boolean} [selectableAll=false] Boolean to show or hide a checkbox in the header to select or
         *  deselect all rows
         */
        this.selectableAll = false;
        /**
         * \@Input {boolean[]} [selected=''] Boolean list to indicate if a row is selected
         */
        this.selected = [];
        /**
         * \@Input {boolean} [fixedHeader=false] Boolean to fix the table header
         */
        this.fixedHeader = false;
        /**
         * \@Input {StTableIconClasses} [iconClasses=''] List of icon classes
         */
        this.iconClasses = new StTableIconClasses();
        /**
         * \@Input {boolean[]} [statusFilter=] List of status filter by column, needed with templateContentFilter
         */
        this.statusFilter = [];
        /**
         * \@Input {string} [hoverButton='icon-ellipsis'] It specifies the icon class of the hover button displayed when user puts mouse over a row
         */
        this.hoverButton = 'icon-ellipsis';
        /**
         * \@Output {Order} [changeOrder=''] Event emitted with the new order which has to be applied to the table rows
         */
        this.changeOrder = new EventEmitter();
        /**
         * \@Output {boolean} [selectAll=''] Event emitted  when user interacts with the checkbox to select or deselect
         * all rows
         */
        this.selectAll = new EventEmitter();
        /**
         * \@Output {EventEmitter<StTableHeader[]>} [fields=] Event emitted when header fields are being loaded
         */
        this.updateFields = new EventEmitter();
        /**
         * \@Output {string} [clickFilter=] Event emitted when using filters custom template
         */
        this.clickFilter = new EventEmitter();
        /**
         * \@Output {StTableHeader[]} [selectFilters=] Event emitted  when user interacts with filter button without a custom template
         */
        this.selectFilters = new EventEmitter();
        /**
         * \@Output {EventEmitter<number} [showHoverMenu=] Event emitted when user clicks on hover button of a row
         */
        this.showHoverMenu = new EventEmitter();
        /**
         * \@Output {Object(checked: boolean, row: number)} [selectRow=] Event emitted when user clicks on checkbox of a row
         */
        this.selectRow = new EventEmitter();
        /**
         * \@Output {Object(checked: boolean, row: number)} [clickCell=] Event emitted when user clicks on checkbox of a row
         */
        this.clickCell = new EventEmitter();
        /**
         * \@Output {StDynamicTableFkEvent} [clickFk=] Event emitted when user clicks on Fk cell
         */
        this.clickFk = new EventEmitter();
        this.fields = [];
    }
    Object.defineProperty(StDynamicTableComponent.prototype, "templateContentFilter", {
        /** @Input {TemplateRef} [templateContentFilter=undefined] Reference to paint a custom template inside popover content */
        get: /**
         * \@Input {TemplateRef} [templateContentFilter=undefined] Reference to paint a custom template inside popover content
         * @return {?}
         */
        function () {
            return this._templateContentFilter;
        },
        set: /**
         * @param {?} _templateRef
         * @return {?}
         */
        function (_templateRef) {
            this._templateContentFilter = _templateRef;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(StDynamicTableComponent.prototype, "jsonSchema", {
        /** @Input {JSONSchema4} [jsonSchema=] Json schema to define its structure */
        get: /**
         * \@Input {JSONSchema4} [jsonSchema=] Json schema to define its structure
         * @return {?}
         */
        function () {
            return this._jsonSchema;
        },
        set: /**
         * @param {?} _jsonSchema
         * @return {?}
         */
        function (_jsonSchema) {
            this._jsonSchema = _jsonSchema;
            this.fields = StDynamicTableUtils.getHeaderFieldsFromJsonSchema(this._jsonSchema, this.uiDefinitions);
            this.updateFields.emit(this.fields);
            this._cd.markForCheck();
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} selectedFilter
     * @return {?}
     */
    StDynamicTableComponent.prototype.onFilterClick = /**
     * @param {?} selectedFilter
     * @return {?}
     */
    function (selectedFilter) {
        this.clickFilter.emit(selectedFilter);
    };
    /**
     * @param {?} order
     * @return {?}
     */
    StDynamicTableComponent.prototype.onChangeOrder = /**
     * @param {?} order
     * @return {?}
     */
    function (order) {
        if (order) {
            this.changeOrder.emit(order);
        }
    };
    /**
     * @param {?} checked
     * @return {?}
     */
    StDynamicTableComponent.prototype.onSelectAll = /**
     * @param {?} checked
     * @return {?}
     */
    function (checked) {
        this.selectAll.emit(checked);
    };
    /**
     * @param {?} selectedFilters
     * @return {?}
     */
    StDynamicTableComponent.prototype.onSelectedFilters = /**
     * @param {?} selectedFilters
     * @return {?}
     */
    function (selectedFilters) {
        this.selectFilters.emit(selectedFilters);
    };
    /**
     * @param {?} row
     * @return {?}
     */
    StDynamicTableComponent.prototype.onShowHoverMenu = /**
     * @param {?} row
     * @return {?}
     */
    function (row) {
        this.showHoverMenu.emit(row);
    };
    /**
     * @return {?}
     */
    StDynamicTableComponent.prototype.onLeaveRow = /**
     * @return {?}
     */
    function () {
        this.showHoverMenu.emit(undefined);
    };
    /**
     * @param {?} item
     * @param {?} field
     * @return {?}
     */
    StDynamicTableComponent.prototype.onClickFk = /**
     * @param {?} item
     * @param {?} field
     * @return {?}
     */
    function (item, field) {
        this.clickFk.emit({
            fk: field.fk,
            value: item[field.id]
        });
    };
    /**
     * @param {?} item
     * @param {?} field
     * @return {?}
     */
    StDynamicTableComponent.prototype.getCellContent = /**
     * @param {?} item
     * @param {?} field
     * @return {?}
     */
    function (item, field) {
        if (field.group) {
            return field.group.split(' - ')
                .map((/**
             * @param {?} _groupKey
             * @return {?}
             */
            function (_groupKey) { return item[_groupKey]; }))
                .join(' - ');
        }
        return item[field.id];
    };
    /**
     * @param {?} field
     * @return {?}
     */
    StDynamicTableComponent.prototype.getCellClasses = /**
     * @param {?} field
     * @return {?}
     */
    function (field) {
        /** @type {?} */
        var classes = {};
        if (field.reference) {
            classes.clickable = true;
        }
        return classes;
    };
    /**
     * @param {?} field
     * @return {?}
     */
    StDynamicTableComponent.prototype.getCellStyles = /**
     * @param {?} field
     * @return {?}
     */
    function (field) {
        /** @type {?} */
        var uiDefinition = this.uiDefinitions && this.uiDefinitions[field.id];
        return uiDefinition && uiDefinition.styles;
    };
    /**
     * @param {?} checkboxEvent
     * @param {?} row
     * @return {?}
     */
    StDynamicTableComponent.prototype.onSelectRow = /**
     * @param {?} checkboxEvent
     * @param {?} row
     * @return {?}
     */
    function (checkboxEvent, row) {
        this.selectRow.emit({ checked: checkboxEvent.checked, row: row });
    };
    /**
     * @param {?} row
     * @param {?} fieldId
     * @param {?} label
     * @return {?}
     */
    StDynamicTableComponent.prototype.onClickCellLabel = /**
     * @param {?} row
     * @param {?} fieldId
     * @param {?} label
     * @return {?}
     */
    function (row, fieldId, label) {
        this.clickCell.emit({ row: row, fieldId: fieldId, label: label });
    };
    StDynamicTableComponent.ctorParameters = function () { return [
        { type: ChangeDetectorRef }
    ]; };
    StDynamicTableComponent.decorators = [
        { type: Component, args: [{
                    selector: 'st-dynamic-table',
                    template: "<!--\n\n    \u00A9 2017 Stratio Big Data Inc., Sucursal en Espa\u00F1a.\n\n    This software is licensed under the Apache License, Version 2.0.\n    This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;\n    without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.\n    See the terms of the License for more details.\n\n    SPDX-License-Identifier: Apache-2.0.\n\n-->\n\n<st-table *ngIf=\"fields && fields.length && items\"\n          class=\"st-dynamic-table\"\n          [fields]=\"fields\"\n          [header]=\"header\"\n          [sortable]=\"sortable\"\n          [filterable]=\"filterable\"\n          [selectableAll]=\"selectableAll\"\n          [hasHoverMenu]=\"hasHoverMenu\"\n          [fixedHeader]=\"fixedHeader\"\n          [currentOrder]=\"currentOrder\"\n          [customClasses]=\"customClasses\"\n          [selectedAll]=\"selectedAll\"\n          [templateContentFilter]=\"templateContentFilter\"\n          [statusFilter]=\"statusFilter\"\n          [iconClasses]=\"iconClasses\"\n          (selectAll)=\"onSelectAll($event)\"\n          (changeOrder)=\"onChangeOrder($event)\"\n          (selectFilters)=\"onSelectedFilters($event)\"\n          (clickFilter)=\"onFilterClick($event)\">\n   <tbody>\n   <tr st-table-row *ngFor=\"let item of items; let row = index\" [selected]=\"selected[row]\" (mouseleave)=\"onLeaveRow()\">\n      <td st-table-cell st-table-row-content\n          *ngFor=\"let field of fields; let cell = index\"\n          [ngClass]=\"getCellClasses(fields[cell])\"\n          [ngStyle]=\"getCellStyles(fields[cell])\">\n\n         <st-checkbox *ngIf=\"selectable && !cell\"\n                      class=\"st-table__checkbox\"\n                      [checked]=\"selected[row]\"\n                      (change)=\"onSelectRow($event, row)\">\n         </st-checkbox>\n         <span (click)=\"onClickCellLabel(row, field.id, item[field.id])\" *ngIf=\"!field.fk; else hasFk\">\n            {{item[field.id]}}\n         </span>\n         <ng-template #hasFk>\n            <a (click)=\"onClickFk(item, field)\" class=\"hasFk\">{{getCellContent(item, field)}}</a>\n         </ng-template>\n      </td>\n      <td st-table-row-hover class=\"st-table-hover\" *ngIf=\"hasHoverMenu\">\n         <i class=\"icon {{hoverButton}}\" (click)=\"onShowHoverMenu(row)\"></i>\n         <ng-content select='[st-dynamic-table-hover-menu]' *ngIf=\"activeHoverMenu === row\"></ng-content>\n      </td>\n   </tr>\n   </tbody>\n</st-table>\n",
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    styles: ["@charset \"UTF-8\";:host{width:100%}:host .st-dynamic-table{max-height:100%;min-width:100%}:host .st-table-hover>i{cursor:pointer}:host .hasFk{color:var(--egeo-theme-action,#0776df);cursor:pointer}"]
                }] }
    ];
    /** @nocollapse */
    StDynamicTableComponent.ctorParameters = function () { return [
        { type: ChangeDetectorRef }
    ]; };
    StDynamicTableComponent.propDecorators = {
        items: [{ type: Input }],
        uiDefinitions: [{ type: Input }],
        qaTag: [{ type: Input }],
        header: [{ type: Input }],
        sortable: [{ type: Input }],
        filterable: [{ type: Input }],
        selectable: [{ type: Input }],
        selectableAll: [{ type: Input }],
        selected: [{ type: Input }],
        currentOrder: [{ type: Input }],
        customClasses: [{ type: Input }],
        fixedHeader: [{ type: Input }],
        iconClasses: [{ type: Input }],
        templateContentFilter: [{ type: Input }],
        statusFilter: [{ type: Input }],
        activeHoverMenu: [{ type: Input }],
        hasHoverMenu: [{ type: Input }],
        hoverButton: [{ type: Input }],
        selectedAll: [{ type: Input }],
        changeOrder: [{ type: Output }],
        selectAll: [{ type: Output }],
        updateFields: [{ type: Output }],
        clickFilter: [{ type: Output }],
        selectFilters: [{ type: Output }],
        showHoverMenu: [{ type: Output }],
        selectRow: [{ type: Output }],
        clickCell: [{ type: Output }],
        clickFk: [{ type: Output }],
        jsonSchema: [{ type: Input }]
    };
    tslib_1.__decorate([
        StRequired(),
        tslib_1.__metadata("design:type", Array)
    ], StDynamicTableComponent.prototype, "items", void 0);
    tslib_1.__decorate([
        StRequired(),
        tslib_1.__metadata("design:type", Object),
        tslib_1.__metadata("design:paramtypes", [Object])
    ], StDynamicTableComponent.prototype, "jsonSchema", null);
    /**
     * \@description {Component} [Dynamic Table]
     *
     * The table component has been designed to be able to create a table deducing its columns using a json schema
     *
     * * \@model
     *
     *   [StDynamicTableHeader] {./st-dynamic-table.model.ts#StDynamicTableHeader}
     *   [StDynamicTableUISpecification] {./st-dynamic-table.model.ts#StDynamicTableUISpecification}
     *   [StDynamicTableUserInterface] {./st-dynamic-table.model.ts#StDynamicTableUserInterface}
     *
     * \@example
     *
     * {html}
     *
     * ```
     * <st-dynamic-table [jsonSchema]="jsonSchema"
     *    [items]="sortedUsers"
     *    [sortable]="true"
     *    [filterable]="true"
     *    [hasHoverMenu]="true"
     *    [currentOrder]="currentOrder"
     *    [activeHoverMenu]="activeHoverMenuPosition"
     *    [templateContentFilter]="filterContent"
     *    (changeOrder)="onSortTable($event)"
     *    (showHoverMenu)="onShowHoverMenu($event)">
     *        <st-dropdown-menu st-dynamic-table-hover-menu
     *          class="hover-menu"
     *          [items]="rowActions"
     *          [active]="activeHoverMenuPosition[3] !== undefined"
     *          [visualMode]="1">
     *       </st-dropdown-menu>
     * </st-dynamic-table>
     * ```
     *
     */
    StDynamicTableComponent = tslib_1.__decorate([
        StEgeo(),
        tslib_1.__metadata("design:paramtypes", [ChangeDetectorRef])
    ], StDynamicTableComponent);
    return StDynamicTableComponent;
}());
export { StDynamicTableComponent };
if (false) {
    /**
     * \@Input {Object(key: string, value: any)[]} [items=''] Item list displayed as table rows
     * @type {?}
     */
    StDynamicTableComponent.prototype.items;
    /**
     * \@Input {StDynamicTableUserInterface} [uiDefinitions=''] UI definition for each field
     * @type {?}
     */
    StDynamicTableComponent.prototype.uiDefinitions;
    /**
     * \@Input {string} [qaTag=''] Prefix used to generate the id values for qa tests
     * @type {?}
     */
    StDynamicTableComponent.prototype.qaTag;
    /**
     * \@Input {boolean} [header=true] Boolean to show or hide the header
     * @type {?}
     */
    StDynamicTableComponent.prototype.header;
    /**
     * \@Input {boolean} [sortable=true] Boolean to make sortable the table, To enable sorting of columns use
     * the new "sortable" field inside stTableHeader model
     * @type {?}
     */
    StDynamicTableComponent.prototype.sortable;
    /**
     * \@Input {boolean} [filterable=false] Boolean to make filterable the table, To enable filtering of columns use
     * the new "filterable" field inside stTableHeader model (necesary define filterConfig).
     * @type {?}
     */
    StDynamicTableComponent.prototype.filterable;
    /**
     * \@Input {boolean} [selectable=false] Boolean to show or hide a checkboxes in the first cell of rows
     * @type {?}
     */
    StDynamicTableComponent.prototype.selectable;
    /**
     * \@Input {boolean} [selectableAll=false] Boolean to show or hide a checkbox in the header to select or
     *  deselect all rows
     * @type {?}
     */
    StDynamicTableComponent.prototype.selectableAll;
    /**
     * \@Input {boolean[]} [selected=''] Boolean list to indicate if a row is selected
     * @type {?}
     */
    StDynamicTableComponent.prototype.selected;
    /**
     * \@Input {Order} [currentOrder=''] It specifies what is the current order applied to the table
     * @type {?}
     */
    StDynamicTableComponent.prototype.currentOrder;
    /**
     * \@Input {string} [customClasses=] Classes for adding styles to table tag from outside. These can be: separated-rows
     * @type {?}
     */
    StDynamicTableComponent.prototype.customClasses;
    /**
     * \@Input {boolean} [fixedHeader=false] Boolean to fix the table header
     * @type {?}
     */
    StDynamicTableComponent.prototype.fixedHeader;
    /**
     * \@Input {StTableIconClasses} [iconClasses=''] List of icon classes
     * @type {?}
     */
    StDynamicTableComponent.prototype.iconClasses;
    /**
     * \@Input {boolean[]} [statusFilter=] List of status filter by column, needed with templateContentFilter
     * @type {?}
     */
    StDynamicTableComponent.prototype.statusFilter;
    /**
     * \@Input {number} [activeHoverMenu=] Position of the current active hover menu
     * @type {?}
     */
    StDynamicTableComponent.prototype.activeHoverMenu;
    /**
     * \@Input {boolean} [hasHoverMenu=] It specifies if a menu has to be displayed when user puts the mouse over
     * the rows. Remember to add a cell with the selector st-table-row-hover for adding content to the menu
     * @type {?}
     */
    StDynamicTableComponent.prototype.hasHoverMenu;
    /**
     * \@Input {string} [hoverButton='icon-ellipsis'] It specifies the icon class of the hover button displayed when user puts mouse over a row
     * @type {?}
     */
    StDynamicTableComponent.prototype.hoverButton;
    /**
     * \@Input {boolean} [selectedAll=] It specifies if all rows are selected
     * @type {?}
     */
    StDynamicTableComponent.prototype.selectedAll;
    /**
     * \@Output {Order} [changeOrder=''] Event emitted with the new order which has to be applied to the table rows
     * @type {?}
     */
    StDynamicTableComponent.prototype.changeOrder;
    /**
     * \@Output {boolean} [selectAll=''] Event emitted  when user interacts with the checkbox to select or deselect
     * all rows
     * @type {?}
     */
    StDynamicTableComponent.prototype.selectAll;
    /**
     * \@Output {EventEmitter<StTableHeader[]>} [fields=] Event emitted when header fields are being loaded
     * @type {?}
     */
    StDynamicTableComponent.prototype.updateFields;
    /**
     * \@Output {string} [clickFilter=] Event emitted when using filters custom template
     * @type {?}
     */
    StDynamicTableComponent.prototype.clickFilter;
    /**
     * \@Output {StTableHeader[]} [selectFilters=] Event emitted  when user interacts with filter button without a custom template
     * @type {?}
     */
    StDynamicTableComponent.prototype.selectFilters;
    /**
     * \@Output {EventEmitter<number} [showHoverMenu=] Event emitted when user clicks on hover button of a row
     * @type {?}
     */
    StDynamicTableComponent.prototype.showHoverMenu;
    /**
     * \@Output {Object(checked: boolean, row: number)} [selectRow=] Event emitted when user clicks on checkbox of a row
     * @type {?}
     */
    StDynamicTableComponent.prototype.selectRow;
    /**
     * \@Output {Object(checked: boolean, row: number)} [clickCell=] Event emitted when user clicks on checkbox of a row
     * @type {?}
     */
    StDynamicTableComponent.prototype.clickCell;
    /**
     * \@Output {StDynamicTableFkEvent} [clickFk=] Event emitted when user clicks on Fk cell
     * @type {?}
     */
    StDynamicTableComponent.prototype.clickFk;
    /** @type {?} */
    StDynamicTableComponent.prototype.fields;
    /**
     * @type {?}
     * @private
     */
    StDynamicTableComponent.prototype._jsonSchema;
    /**
     * @type {?}
     * @private
     */
    StDynamicTableComponent.prototype._templateContentFilter;
    /**
     * @type {?}
     * @private
     */
    StDynamicTableComponent.prototype._cd;
}
