// @stratio/egeo/esm2015/lib/st-draggable-card/st-draggable-card.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/st-draggable-card/st-draggable-card.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/*
 * © 2017 Stratio Big Data Inc., Sucursal en España.
 *
 * This software is licensed under the Apache License, Version 2.0.
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the terms of the License for more details.
 *
 * SPDX-License-Identifier: Apache-2.0.
 */
import { Component, EventEmitter, Input, Output, HostListener, HostBinding, ChangeDetectionStrategy, ViewChild } from '@angular/core';
/**
 * \@description {Component} [Draggable card]
 *
 * The draggable card component is a custom draggable item.
 *
 * \@example
 *
 * {html}
 *
 * ```
 *
 *  <st-draggable-card class='card' id='1' (onDragStart)='drag($event)'>
 *      <p>Draggable card 1</p>
 *  </st-draggable-card>
 *
 *
 * ```
 *
 */
export class StDraggableCardComponent {
    constructor() {
        /**
         * \@Output {any} [dragStart] Event emitted when user start dragging card
         */
        this.onDragStart = new EventEmitter();
        /**
         * \@Output {any} [dragEnd] Event emitted when user drops dragging card
         */
        this.onDragEnd = new EventEmitter();
        /**
         * \@Input {boolean} [draggable=true] Boolean to enable or disable dragging
         */
        this.draggable = true;
        this._dragging = false;
    }
    /**
     * @return {?}
     */
    get dragging() {
        return this._dragging;
    }
    /**
     * @param {?} event
     * @return {?}
     */
    dragStartHandle(event) {
        this._dragging = true;
        setTimeout((/**
         * @return {?}
         */
        () => this.card.nativeElement.classList.add('shadow')));
        event.dataTransfer.effectAllowed = 'all';
        event.dataTransfer.dropEffect = 'move';
        this.onDragStart.emit(event);
        event.stopPropagation();
    }
    /**
     * @param {?} event
     * @return {?}
     */
    dragEndHandle(event) {
        this.card.nativeElement.classList.remove('shadow');
        this._dragging = false;
        event.stopPropagation();
        this.onDragEnd.emit(event);
    }
}
StDraggableCardComponent.decorators = [
    { type: Component, args: [{
                selector: 'st-draggable-card',
                template: "<!--\n\n    \u00A9 2017 Stratio Big Data Inc., Sucursal en Espa\u00F1a.\n\n    This software is licensed under the Apache License, Version 2.0.\n    This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;\n    without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.\n    See the terms of the License for more details.\n\n    SPDX-License-Identifier: Apache-2.0.\n\n-->\n\n<div #card class=\"st-draggable-card\" [id]=\"id\" [ngClass]=\"{dragging: dragging}\">\n   <span class=\"drag-button\">\n        <i class=\"icon-drag\"></i>\n   </span>\n   <div class=\"content\">\n      <ng-content></ng-content>\n   </div>\n</div>\n",
                changeDetection: ChangeDetectionStrategy.OnPush,
                styles: ["@charset \"UTF-8\";.st-draggable-card{width:100%;height:100%;padding:30px 30px 30px 15px;display:flex}.st-draggable-card .drag-button{display:block;vertical-align:middle;margin-right:15px;-ms-grid-row-align:center;align-self:center}.st-draggable-card .content{display:block}"]
            }] }
];
StDraggableCardComponent.propDecorators = {
    id: [{ type: Input, args: ['id',] }],
    onDragStart: [{ type: Output }],
    onDragEnd: [{ type: Output }],
    draggable: [{ type: HostBinding, args: ['draggable',] }, { type: Input }],
    card: [{ type: ViewChild, args: ['card', { static: false },] }],
    dragStartHandle: [{ type: HostListener, args: ['dragstart', ['$event'],] }],
    dragEndHandle: [{ type: HostListener, args: ['dragend', ['$event'],] }]
};
if (false) {
    /**
     * \@Input {string} [id] Card id
     * @type {?}
     */
    StDraggableCardComponent.prototype.id;
    /**
     * \@Output {any} [dragStart] Event emitted when user start dragging card
     * @type {?}
     */
    StDraggableCardComponent.prototype.onDragStart;
    /**
     * \@Output {any} [dragEnd] Event emitted when user drops dragging card
     * @type {?}
     */
    StDraggableCardComponent.prototype.onDragEnd;
    /**
     * \@Input {boolean} [draggable=true] Boolean to enable or disable dragging
     * @type {?}
     */
    StDraggableCardComponent.prototype.draggable;
    /** @type {?} */
    StDraggableCardComponent.prototype.card;
    /**
     * @type {?}
     * @private
     */
    StDraggableCardComponent.prototype._dragging;
}
