// @stratio/egeo/esm5/lib/st-modal/st-modal-buttons/st-modal-buttons.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/st-modal/st-modal-buttons/st-modal-buttons.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/*
 * © 2017 Stratio Big Data Inc., Sucursal en España.
 *
 * This software is licensed under the Apache License, Version 2.0.
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the terms of the License for more details.
 *
 * SPDX-License-Identifier: Apache-2.0.
 */
import { Component, ChangeDetectionStrategy, Input, Output, EventEmitter } from '@angular/core';
var StModalButtonsComponent = /** @class */ (function () {
    function StModalButtonsComponent() {
        this.buttonConfig = [];
        this.fullscreen = false;
        this.click = new EventEmitter();
        this.hasButtonsValue = false;
    }
    /**
     * @return {?}
     */
    StModalButtonsComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.hasButtonsValue = this.hasButtons();
    };
    /**
     * @return {?}
     */
    StModalButtonsComponent.prototype.hasButtons = /**
     * @return {?}
     */
    function () {
        return this.buttonConfig && this.buttonConfig.length > 0;
    };
    /**
     * @param {?} button
     * @return {?}
     */
    StModalButtonsComponent.prototype.getButtonClass = /**
     * @param {?} button
     * @return {?}
     */
    function (button) {
        /** @type {?} */
        var classes = button && button.classes ? button.classes : '';
        return classes.trim();
    };
    /**
     * @param {?} button
     * @return {?}
     */
    StModalButtonsComponent.prototype.getButtonId = /**
     * @param {?} button
     * @return {?}
     */
    function (button) {
        return button && button.label ? "st-modal-button-" + this.getButtonLabelWithoutSpaces(button.label) : null;
    };
    /**
     * @param {?} label
     * @return {?}
     */
    StModalButtonsComponent.prototype.getButtonLabelWithoutSpaces = /**
     * @param {?} label
     * @return {?}
     */
    function (label) {
        return label.replace(/\s+/g, '_');
    };
    /** INTERACTION WITH MODAL */
    /**
     * INTERACTION WITH MODAL
     * @param {?} button
     * @param {?} event
     * @return {?}
     */
    StModalButtonsComponent.prototype.clickButton = /**
     * INTERACTION WITH MODAL
     * @param {?} button
     * @param {?} event
     * @return {?}
     */
    function (button, event) {
        if (button.response) {
            button.response(event);
        }
        else if (button.responseValue !== undefined) {
            this.click.emit({ response: button.responseValue, close: button.closeOnClick });
        }
    };
    StModalButtonsComponent.decorators = [
        { type: Component, args: [{
                    selector: 'st-modal-buttons',
                    template: "<!--\n\n    \u00A9 2017 Stratio Big Data Inc., Sucursal en Espa\u00F1a.\n\n    This software is licensed under the Apache License, Version 2.0.\n    This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;\n    without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.\n    See the terms of the License for more details.\n\n    SPDX-License-Identifier: Apache-2.0.\n\n-->\n<div *ngIf=\"hasButtonsValue\" class=\"st-modal-buttons\">\n   <button *ngFor=\"let button of buttonConfig\" [attr.id]=\"getButtonId(button)\" class=\"button\" [ngClass]=\"getButtonClass(button)\" (click)=\"clickButton(button, $event)\"\n       type=\"button\">\n      <i *ngIf=\"button.leftIcon\" [ngClass]=\"button.leftIcon\"></i>\n      <span>{{button.label}}</span>\n      <i *ngIf=\"button.rightIcon\" [ngClass]=\"button.rightIcon\"></i>\n   </button>\n</div>\n",
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    styles: ["@charset \"UTF-8\";:host{display:flex;width:100%;flex-grow:1}.st-modal-buttons{display:flex;flex-direction:row;justify-content:flex-end;width:100%}"]
                }] }
    ];
    StModalButtonsComponent.propDecorators = {
        buttonConfig: [{ type: Input }],
        fullscreen: [{ type: Input }],
        click: [{ type: Output }]
    };
    return StModalButtonsComponent;
}());
export { StModalButtonsComponent };
if (false) {
    /** @type {?} */
    StModalButtonsComponent.prototype.buttonConfig;
    /** @type {?} */
    StModalButtonsComponent.prototype.fullscreen;
    /** @type {?} */
    StModalButtonsComponent.prototype.click;
    /** @type {?} */
    StModalButtonsComponent.prototype.hasButtonsValue;
}
