// @stratio/egeo/esm5/lib/st-switch/st-switch.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/st-switch/st-switch.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
/*
 * © 2017 Stratio Big Data Inc., Sucursal en España.
 *
 * This software is licensed under the Apache License, Version 2.0.
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the terms of the License for more details.
 *
 * SPDX-License-Identifier: Apache-2.0.
 */
import { Component, Input, ChangeDetectionStrategy, forwardRef, ChangeDetectorRef, Output, EventEmitter } from '@angular/core';
import { NG_VALUE_ACCESSOR, NG_VALIDATORS } from '@angular/forms';
import { StEgeo } from '../decorators/require-decorators';
var StSwitchComponent = /** @class */ (function () {
    function StSwitchComponent(_cd) {
        this._cd = _cd;
        this.change = new EventEmitter();
    }
    StSwitchComponent_1 = StSwitchComponent;
    Object.defineProperty(StSwitchComponent.prototype, "value", {
        get: /**
         * @return {?}
         */
        function () {
            return this._value;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._value = value;
            this._cd.markForCheck();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(StSwitchComponent.prototype, "disabled", {
        get: /**
         * @return {?}
         */
        function () {
            return this._disabled;
        },
        set: /**
         * @param {?} disabled
         * @return {?}
         */
        function (disabled) {
            this._disabled = disabled;
            this._cd.markForCheck();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(StSwitchComponent.prototype, "labelQaTag", {
        get: /**
         * @return {?}
         */
        function () {
            return (this.qaTag || this.name) + '-label';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(StSwitchComponent.prototype, "relatedInput", {
        get: /**
         * @return {?}
         */
        function () {
            return this.name + "-input";
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} control
     * @return {?}
     */
    StSwitchComponent.prototype.validate = /**
     * @param {?} control
     * @return {?}
     */
    function (control) { };
    // load external change
    // load external change
    /**
     * @param {?} value
     * @return {?}
     */
    StSwitchComponent.prototype.writeValue = 
    // load external change
    /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        if (!this._disabled) {
            this._value = value;
            this.change.emit(this._value);
            if (this.registeredOnChange) {
                this.registeredOnChange(value);
            }
        }
    };
    // internal change callback
    // internal change callback
    /**
     * @param {?} fn
     * @return {?}
     */
    StSwitchComponent.prototype.registerOnChange = 
    // internal change callback
    /**
     * @param {?} fn
     * @return {?}
     */
    function (fn) {
        this.registeredOnChange = fn;
    };
    /**
     * @param {?} fn
     * @return {?}
     */
    StSwitchComponent.prototype.registerOnTouched = /**
     * @param {?} fn
     * @return {?}
     */
    function (fn) {
    };
    /**
     * @param {?} disable
     * @return {?}
     */
    StSwitchComponent.prototype.setDisabledState = /**
     * @param {?} disable
     * @return {?}
     */
    function (disable) {
        this.disabled = disable;
    };
    /**
     * @param {?} event
     * @return {?}
     */
    StSwitchComponent.prototype.onChange = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        event.stopPropagation();
        /** @type {?} */
        var value = ((/** @type {?} */ (event.target))).checked;
        this._value = value;
        this.writeValue(value);
    };
    var StSwitchComponent_1;
    StSwitchComponent.ctorParameters = function () { return [
        { type: ChangeDetectorRef }
    ]; };
    StSwitchComponent.decorators = [
        { type: Component, args: [{
                    selector: 'st-switch',
                    host: { class: 'st-switch' },
                    template: "<!--\n\n    \u00A9 2017 Stratio Big Data Inc., Sucursal en Espa\u00F1a.\n\n    This software is licensed under the Apache License, Version 2.0.\n    This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;\n    without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.\n    See the terms of the License for more details.\n\n    SPDX-License-Identifier: Apache-2.0.\n\n-->\n<label st-label [attr.for]=\"relatedInput\" [attr.title]=\"contextualHelp\" [attr.id]=\"labelQaTag\" >\n   {{label}}\n</label>\n<div class=\"st-switch__toggle sth-switch__toggle\" (click)=\"writeValue(!value)\" [attr.id]=\"labelQaTag\" [ngClass]=\"{ 'st-switch--disabled': disabled, 'st-switch--on': value, 'st-switch--off': !value}\">\n   <div class=\"st-switch__box sth-switch__box\">\n      <span class=\"st-switch__circle sth-switch__circle\"></span>\n   </div>\n   <input class=\"st-switch__input\" [name]=\"name\" [attr.type]=\"'checkbox'\" [checked]=\"value\" (click)=\"onChange($event)\" [attr.id]=\"relatedInput\"\n   [disabled]=\"disabled\" />\n</div>\n",
                    providers: [
                        { provide: NG_VALUE_ACCESSOR, useExisting: forwardRef((/**
                             * @return {?}
                             */
                            function () { return StSwitchComponent_1; })), multi: true },
                        { provide: NG_VALIDATORS, useExisting: forwardRef((/**
                             * @return {?}
                             */
                            function () { return StSwitchComponent_1; })), multi: true }
                    ],
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    styles: ["@charset \"UTF-8\";.st-switch__input{visibility:hidden;position:absolute}.st-switch__toggle{margin-left:15px;vertical-align:middle;position:relative;height:25px;width:45px;display:inline-block;cursor:pointer}.st-switch__box{border-radius:42px;height:100%;width:100%}.st-switch__circle{width:19px;height:19px;top:3px;left:3px;border-radius:100%;position:absolute;transition:.3s}.st-switch--disabled{opacity:.5}.st-switch--on .st-switch__circle{margin-left:21px}"]
                }] }
    ];
    /** @nocollapse */
    StSwitchComponent.ctorParameters = function () { return [
        { type: ChangeDetectorRef }
    ]; };
    StSwitchComponent.propDecorators = {
        qaTag: [{ type: Input }],
        label: [{ type: Input }],
        name: [{ type: Input }],
        contextualHelp: [{ type: Input }],
        change: [{ type: Output }],
        value: [{ type: Input }],
        disabled: [{ type: Input }]
    };
    StSwitchComponent = StSwitchComponent_1 = tslib_1.__decorate([
        StEgeo(),
        tslib_1.__metadata("design:paramtypes", [ChangeDetectorRef])
    ], StSwitchComponent);
    return StSwitchComponent;
}());
export { StSwitchComponent };
if (false) {
    /** @type {?} */
    StSwitchComponent.prototype.qaTag;
    /** @type {?} */
    StSwitchComponent.prototype.label;
    /** @type {?} */
    StSwitchComponent.prototype.name;
    /** @type {?} */
    StSwitchComponent.prototype.contextualHelp;
    /** @type {?} */
    StSwitchComponent.prototype.change;
    /**
     * @type {?}
     * @private
     */
    StSwitchComponent.prototype._value;
    /**
     * @type {?}
     * @private
     */
    StSwitchComponent.prototype._disabled;
    /**
     * @type {?}
     * @private
     */
    StSwitchComponent.prototype.registeredOnChange;
    /**
     * @type {?}
     * @private
     */
    StSwitchComponent.prototype._cd;
}
