// @stratio/egeo/esm5/lib/st-modal/st-modal.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/st-modal/st-modal.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/*
 * © 2017 Stratio Big Data Inc., Sucursal en España.
 *
 * This software is licensed under the Apache License, Version 2.0.
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the terms of the License for more details.
 *
 * SPDX-License-Identifier: Apache-2.0.
 */
import { ComponentFactoryResolver, Injectable } from '@angular/core';
import { Subject } from 'rxjs';
/* local dependencies */
import { StModalComponent } from './st-modal.component';
import { StModalResponse, StModalBasicType } from './st-modal.model';
var StModalService = /** @class */ (function () {
    function StModalService(_cfr) {
        this._cfr = _cfr;
        this._containerRef = undefined;
        this.dynamicModal = undefined;
    }
    Object.defineProperty(StModalService.prototype, "container", {
        /* External API */
        set: /* External API */
        /**
         * @param {?} container
         * @return {?}
         */
        function (container) {
            this._containerRef = container;
        },
        enumerable: true,
        configurable: true
    });
    // - Public methods
    // - Public methods
    /**
     * @param {?} config
     * @param {?=} component
     * @return {?}
     */
    StModalService.prototype.show = 
    // - Public methods
    /**
     * @param {?} config
     * @param {?=} component
     * @return {?}
     */
    function (config, component) {
        /** @type {?} */
        var errors = this.canCreateModal(config, component);
        if (errors && errors.length > 0) {
            throw new Error(errors.join(' '));
        }
        this.notifyButtonInteraction = new Subject();
        this.createModal(this.createConfig(config), component);
        return this.notifyButtonInteraction.asObservable();
    };
    /**
     * @param {?} type
     * @param {?} modalTitle
     * @param {?} messageTitle
     * @param {?} message
     * @param {?} okButton
     * @param {?=} cancelButton
     * @param {?=} maxWidth
     * @param {?=} minWidth
     * @param {?=} icon
     * @return {?}
     */
    StModalService.prototype.showBasicModal = /**
     * @param {?} type
     * @param {?} modalTitle
     * @param {?} messageTitle
     * @param {?} message
     * @param {?} okButton
     * @param {?=} cancelButton
     * @param {?=} maxWidth
     * @param {?=} minWidth
     * @param {?=} icon
     * @return {?}
     */
    function (type, modalTitle, messageTitle, message, okButton, cancelButton, maxWidth, minWidth, icon) {
        if (cancelButton === void 0) { cancelButton = ''; }
        if (maxWidth === void 0) { maxWidth = 600; }
        if (minWidth === void 0) { minWidth = 400; }
        if (icon === void 0) { icon = ''; }
        /** @type {?} */
        var iconStatus;
        /** @type {?} */
        var buttons = [{
                label: okButton,
                classes: (type === StModalBasicType.DELETE) ? 'button-critical' : 'button-primary',
                responseValue: StModalResponse.YES,
                closeOnClick: true
            }];
        if (type !== StModalBasicType.INFO) {
            buttons.unshift({
                label: cancelButton,
                classes: (type === StModalBasicType.DELETE) ? 'button-borderless' : 'button-secondary',
                responseValue: StModalResponse.NO,
                closeOnClick: true
            });
        }
        switch (type) {
            case StModalBasicType.DELETE:
                iconStatus = 'icon-circle-cross';
                break;
            case StModalBasicType.WARNING:
                iconStatus = 'icon-alert';
                break;
            default:
                iconStatus = icon;
                break;
        }
        return this.show({
            fullscreen: false,
            message: message,
            messageTitle: messageTitle,
            modalTitle: modalTitle,
            buttons: buttons,
            maxWidth: maxWidth,
            minWidth: minWidth,
            iconStatus: iconStatus
        });
    };
    /**
     * @return {?}
     */
    StModalService.prototype.close = /**
     * @return {?}
     */
    function () {
        this.destroy();
    };
    /* INTERNAL METHODS FOR WORK WITH MODALS */
    /* INTERNAL METHODS FOR WORK WITH MODALS */
    /**
     * @private
     * @param {?} modalConfig
     * @param {?=} component
     * @return {?}
     */
    StModalService.prototype.createModal = /* INTERNAL METHODS FOR WORK WITH MODALS */
    /**
     * @private
     * @param {?} modalConfig
     * @param {?=} component
     * @return {?}
     */
    function (modalConfig, component) {
        /** @type {?} */
        var stModalFactory = this._cfr.resolveComponentFactory(StModalComponent);
        if (stModalFactory) {
            this._containerRef.clear();
            this.dynamicModal = this._containerRef.createComponent(stModalFactory);
            this.bindVars(modalConfig, component);
        }
    };
    /**
     * @private
     * @return {?}
     */
    StModalService.prototype.destroy = /**
     * @private
     * @return {?}
     */
    function () {
        if (this.dynamicModal) {
            this.dynamicModal.destroy();
            this.dynamicModal = undefined;
            this.notifyButtonInteraction.next(StModalResponse.CLOSE);
            this.notifyButtonInteraction.complete();
        }
    };
    /**
     * @private
     * @param {?} modalConfig
     * @param {?} component
     * @return {?}
     */
    StModalService.prototype.bindVars = /**
     * @private
     * @param {?} modalConfig
     * @param {?} component
     * @return {?}
     */
    function (modalConfig, component) {
        this.dynamicModal.instance.component = component;
        this.dynamicModal.instance.click.subscribe(this.notify.bind(this));
        this.dynamicModal.instance.modalConfig = modalConfig;
        this.dynamicModal.changeDetectorRef.detectChanges();
    };
    /**
     * @private
     * @param {?} buttonResponse
     * @return {?}
     */
    StModalService.prototype.notify = /**
     * @private
     * @param {?} buttonResponse
     * @return {?}
     */
    function (buttonResponse) {
        this.notifyButtonInteraction.next(buttonResponse.response);
        if (buttonResponse.close) {
            this.close();
        }
    };
    /**
     * @private
     * @param {?} config
     * @param {?=} component
     * @return {?}
     */
    StModalService.prototype.canCreateModal = /**
     * @private
     * @param {?} config
     * @param {?=} component
     * @return {?}
     */
    function (config, component) {
        /** @type {?} */
        var errors = [];
        if (!this._containerRef) {
            errors.push("[ERROR]: StModalService => Cant find container, are you sure you declarate in MAIN APP component in html and typescript?");
        }
        if (this.dynamicModal !== undefined) {
            errors.push("[ERROR]: StModalService => Can't create modal beacause already exists one. Are you sure that you call close method?)");
        }
        if (!component && !config.message && !config.html) {
            errors.push("[ERROR]: StModalService => Can't find message, html or component to show in modal");
        }
        return errors;
    };
    /**
     * @private
     * @param {?} config
     * @return {?}
     */
    StModalService.prototype.createConfig = /**
     * @private
     * @param {?} config
     * @return {?}
     */
    function (config) {
        /** @type {?} */
        var defaultConfig = {
            fullscreen: false,
            inputs: {},
            outputs: {},
            modalTitle: 'Default title',
            messageTitle: 'Default subtitle',
            buttons: [],
            message: undefined,
            html: undefined,
            maxWidth: undefined,
            minWidth: undefined,
            empty: false
        };
        /** @type {?} */
        var checkedConfig = Object.assign({}, defaultConfig, config);
        checkedConfig.buttons = this.checkButtons(checkedConfig.buttons);
        return checkedConfig;
    };
    /**
     * @private
     * @param {?} buttons
     * @return {?}
     */
    StModalService.prototype.checkButtons = /**
     * @private
     * @param {?} buttons
     * @return {?}
     */
    function (buttons) {
        return buttons.map((/**
         * @param {?} button
         * @return {?}
         */
        function (button) { return Object.assign({}, { label: 'Button', closeOnClick: false }, button); }));
    };
    StModalService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    StModalService.ctorParameters = function () { return [
        { type: ComponentFactoryResolver }
    ]; };
    return StModalService;
}());
export { StModalService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    StModalService.prototype._containerRef;
    /**
     * @type {?}
     * @private
     */
    StModalService.prototype.dynamicModal;
    /**
     * @type {?}
     * @private
     */
    StModalService.prototype.notifyButtonInteraction;
    /**
     * @type {?}
     * @private
     */
    StModalService.prototype._cfr;
}
