// @stratio/egeo/esm5/lib/st-modal/st-modal.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/st-modal/st-modal.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/*
 * © 2017 Stratio Big Data Inc., Sucursal en España.
 *
 * This software is licensed under the Apache License, Version 2.0.
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the terms of the License for more details.
 *
 * SPDX-License-Identifier: Apache-2.0.
 */
import { ChangeDetectionStrategy, Component, ComponentFactoryResolver, EventEmitter, Input, Output, ViewChild, ViewContainerRef } from '@angular/core';
import { StModalConfig, StModalResponse } from './st-modal.model';
import { StWindowRefService } from '../utils/window-service';
var StModalComponent = /** @class */ (function () {
    function StModalComponent(cfr, windowRef) {
        this.cfr = cfr;
        this.windowRef = windowRef;
        this.click = new EventEmitter();
        this.defaultMaxWidth = 600;
        this.defaultMinWidth = 400;
    }
    Object.defineProperty(StModalComponent.prototype, "hasIcon", {
        get: /**
         * @return {?}
         */
        function () {
            return this.modalConfig.iconStatus;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(StModalComponent.prototype, "isFullscreen", {
        get: /**
         * @return {?}
         */
        function () {
            return this.modalConfig.fullscreen;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(StModalComponent.prototype, "title", {
        get: /**
         * @return {?}
         */
        function () {
            return this.modalConfig.modalTitle;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(StModalComponent.prototype, "buttons", {
        get: /**
         * @return {?}
         */
        function () {
            return this.modalConfig.buttons || [];
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(StModalComponent.prototype, "isMessageModal", {
        get: /**
         * @return {?}
         */
        function () {
            return this.modalConfig.message && this.modalConfig.message.length > 0;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(StModalComponent.prototype, "isComplexMessageModal", {
        get: /**
         * @return {?}
         */
        function () {
            return this.modalConfig.html && this.modalConfig.html.length > 0;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(StModalComponent.prototype, "html", {
        get: /**
         * @return {?}
         */
        function () {
            return this.modalConfig.html;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(StModalComponent.prototype, "message", {
        get: /**
         * @return {?}
         */
        function () {
            return this.modalConfig.message;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(StModalComponent.prototype, "messageTitle", {
        get: /**
         * @return {?}
         */
        function () {
            return this.modalConfig.messageTitle;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(StModalComponent.prototype, "modalStyles", {
        get: /**
         * @return {?}
         */
        function () {
            /** @type {?} */
            var maxWidth = this.modalConfig.maxWidth || this.defaultMaxWidth;
            /** @type {?} */
            var minWidth = this.modalConfig.minWidth || this.defaultMinWidth;
            /** @type {?} */
            var width = this.getModalActualWidth(maxWidth, minWidth);
            return { 'max-width': maxWidth + "px", 'min-width': minWidth + "px", 'width': width + "px" };
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(StModalComponent.prototype, "emptyModal", {
        get: /**
         * @return {?}
         */
        function () {
            return this.modalConfig && this.modalConfig.empty;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(StModalComponent.prototype, "showCloseBtn", {
        get: /**
         * @return {?}
         */
        function () {
            return this.modalConfig.showCloseBtn;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    StModalComponent.prototype.onClose = /**
     * @return {?}
     */
    function () {
        this.click.emit({
            response: StModalResponse.CLOSE,
            close: true
        });
    };
    /** DYNAMIC MODAL BODY COMPONENT LOAD */
    /**
     * DYNAMIC MODAL BODY COMPONENT LOAD
     * @return {?}
     */
    StModalComponent.prototype.ngAfterViewInit = /**
     * DYNAMIC MODAL BODY COMPONENT LOAD
     * @return {?}
     */
    function () {
        this.target = this.emptyModal ? this.targetEmpty : this.targetContent;
        if (this.component && !(this.modalConfig.html || this.modalConfig.message)) {
            this.loadBody();
        }
        this.windowRef.nativeWindow.document.body.classList.add('st-modal-overlay');
    };
    /**
     * @return {?}
     */
    StModalComponent.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        if (this.componentRef) {
            this.componentRef.destroy();
        }
        this.windowRef.nativeWindow.document.body.classList.remove('st-modal-overlay');
    };
    /**
     * @private
     * @param {?} maxWidth
     * @param {?=} minWidth
     * @return {?}
     */
    StModalComponent.prototype.getModalActualWidth = /**
     * @private
     * @param {?} maxWidth
     * @param {?=} minWidth
     * @return {?}
     */
    function (maxWidth, minWidth) {
        /** @type {?} */
        var screenWidth = this.windowRef.nativeWindow.screen.width;
        return screenWidth > maxWidth ? maxWidth : (screenWidth < minWidth ? minWidth : screenWidth);
    };
    /**
     * @private
     * @return {?}
     */
    StModalComponent.prototype.loadBody = /**
     * @private
     * @return {?}
     */
    function () {
        if (!this.componentRef) {
            this.target.clear();
            /** @type {?} */
            var compFactory = this.cfr.resolveComponentFactory(this.component);
            this.componentRef = this.target.createComponent(compFactory);
            this.bindModalInputs();
        }
    };
    /**
     * @private
     * @return {?}
     */
    StModalComponent.prototype.bindModalInputs = /**
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        Object.keys(this.modalConfig.inputs).forEach((/**
         * @param {?} key
         * @return {?}
         */
        function (key) {
            _this.componentRef.instance[key] = ((/** @type {?} */ (_this.modalConfig.inputs)))[key];
        }));
        Object.keys(this.modalConfig.outputs).forEach((/**
         * @param {?} key
         * @return {?}
         */
        function (key) {
            _this.componentRef.instance[key].subscribe(((/** @type {?} */ (_this.modalConfig.outputs)))[key]);
        }));
        this.componentRef.changeDetectorRef.detectChanges();
    };
    StModalComponent.decorators = [
        { type: Component, args: [{
                    selector: 'st-modal',
                    template: "<!--\n\n    \u00A9 2017 Stratio Big Data Inc., Sucursal en Espa\u00F1a.\n\n    This software is licensed under the Apache License, Version 2.0.\n    This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;\n    without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.\n    See the terms of the License for more details.\n\n    SPDX-License-Identifier: Apache-2.0.\n\n-->\n<div class=\"st-modal-container\">\n   <section *ngIf=\"emptyModal\" class=\"st-modal\" [ngClass]=\"{'st-modal-fullscreen': isFullscreen}\" [ngStyle]=\"modalStyles\" [hidden]=\"emptyModal\">\n      <div #stModalBodyEmpty></div>\n   </section>\n   <section *ngIf=\"!emptyModal\" class=\"st-modal\" [ngClass]=\"{'st-modal-fullscreen': isFullscreen}\" [ngStyle]=\"modalStyles\" [hidden]=\"!emptyModal\">\n      <div class=\"st-modal-header\">\n         <div class=\"container\">\n            <span class=\"status-icon\" [ngClass]=\"hasIcon\" *ngIf=\"hasIcon\"></span>\n            <p class=\"title\">{{title}}</p> <span class=\"icon-cross close-button\" *ngIf=\"showCloseBtn\" (click)=\"onClose()\"></span>\n\n            <st-modal-buttons *ngIf=\"isFullscreen\" [buttonConfig]=\"buttons\" [fullscreen]=\"isFullscreen\" (click)=\"click.emit($event)\"></st-modal-buttons>\n         </div>\n      </div>\n      <div class=\"st-modal-body\">\n         <div class=\"container\">\n\n            <div *ngIf=\"isMessageModal\" class=\"message\">\n               <h1 id=\"st-modal-message-plain-title\" class=\"st-modal-message-plain-title\">{{messageTitle}}</h1>\n               <p id=\"st-modal-message-plain-message\" class=\"st-modal-message-plain-message\">{{message}}</p>\n            </div>\n            <div *ngIf=\"isComplexMessageModal\" [innerHTML]=\"html\" id=\"st-modal-message-html\"></div>\n            <div #stModalBody id=\"st-modal-message-component\"></div>\n\n         </div>\n      </div>\n      <footer class=\"st-modal-footer\" *ngIf=\"buttons && buttons.length > 0\">\n         <div class=\"container\">\n            <st-modal-buttons *ngIf=\"!isFullscreen\" [buttonConfig]=\"buttons\" (click)=\"click.emit($event)\"></st-modal-buttons>\n         </div>\n      </footer>\n   </section>\n</div>\n\n",
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    styles: ["@charset \"UTF-8\";.st-modal-container{display:flex;height:100vh;left:0;position:fixed;top:0;width:100%}.st-modal-container .st-modal{align-self:center;display:flex;flex-direction:column;margin:auto}.st-modal-container .st-modal.st-modal-fullscreen{min-height:100vh;max-height:100vh;min-width:100%;max-width:100%}.st-modal-container .st-modal .st-modal-header{display:flex;flex-direction:row;height:70px;width:100%}.st-modal-container .st-modal .st-modal-header .container{align-items:center;flex-wrap:nowrap}.st-modal-container .st-modal .st-modal-header .container .status-icon{padding-right:1.05rem}.st-modal-container .st-modal st-modal-buttons{display:flex;justify-content:flex-end;margin-left:auto}.st-modal-container .st-modal .container{min-width:auto}"]
                }] }
    ];
    /** @nocollapse */
    StModalComponent.ctorParameters = function () { return [
        { type: ComponentFactoryResolver },
        { type: StWindowRefService }
    ]; };
    StModalComponent.propDecorators = {
        modalConfig: [{ type: Input }],
        component: [{ type: Input }],
        click: [{ type: Output }],
        targetContent: [{ type: ViewChild, args: ['stModalBody', { read: ViewContainerRef, static: false },] }],
        targetEmpty: [{ type: ViewChild, args: ['stModalBodyEmpty', { read: ViewContainerRef, static: false },] }]
    };
    return StModalComponent;
}());
export { StModalComponent };
if (false) {
    /** @type {?} */
    StModalComponent.prototype.modalConfig;
    /** @type {?} */
    StModalComponent.prototype.component;
    /** @type {?} */
    StModalComponent.prototype.click;
    /** @type {?} */
    StModalComponent.prototype.targetContent;
    /** @type {?} */
    StModalComponent.prototype.targetEmpty;
    /** @type {?} */
    StModalComponent.prototype.defaultMaxWidth;
    /** @type {?} */
    StModalComponent.prototype.defaultMinWidth;
    /** @type {?} */
    StModalComponent.prototype.target;
    /**
     * @type {?}
     * @private
     */
    StModalComponent.prototype.componentRef;
    /**
     * @type {?}
     * @private
     */
    StModalComponent.prototype.cfr;
    /**
     * @type {?}
     * @private
     */
    StModalComponent.prototype.windowRef;
}
