// @stratio/egeo/esm2015/lib/st-header/st-header.model.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/st-header/st-header.model.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/*
 * © 2017 Stratio Big Data Inc., Sucursal en España.
 *
 * This software is licensed under the Apache License, Version 2.0.
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the terms of the License for more details.
 *
 * SPDX-License-Identifier: Apache-2.0.
 */
import { StDropDownMenuItem } from '../st-dropdown-menu/st-dropdown-menu.interface';
export class StHeaderMenuOption {
}
if (false) {
    /** @type {?} */
    StHeaderMenuOption.prototype.icon;
    /** @type {?} */
    StHeaderMenuOption.prototype.label;
    /** @type {?} */
    StHeaderMenuOption.prototype.subMenus;
    /** @type {?} */
    StHeaderMenuOption.prototype.link;
    /** @type {?} */
    StHeaderMenuOption.prototype.external;
    /** @type {?} */
    StHeaderMenuOption.prototype.openInNewPage;
}
export class StHeaderSubMenuOption {
}
if (false) {
    /** @type {?} */
    StHeaderSubMenuOption.prototype.label;
    /** @type {?} */
    StHeaderSubMenuOption.prototype.link;
    /** @type {?} */
    StHeaderSubMenuOption.prototype.external;
    /** @type {?} */
    StHeaderSubMenuOption.prototype.openInNewPage;
}
export class StHeaderSelection {
}
if (false) {
    /** @type {?} */
    StHeaderSelection.prototype.link;
    /** @type {?} */
    StHeaderSelection.prototype.external;
    /** @type {?} */
    StHeaderSelection.prototype.openInNewPage;
}
export class StHeaderMenuItem extends StDropDownMenuItem {
}
if (false) {
    /** @type {?} */
    StHeaderMenuItem.prototype.selection;
}
export class StHeaderMenuOptionSchema {
}
if (false) {
    /** @type {?} */
    StHeaderMenuOptionSchema.prototype.icon;
    /** @type {?} */
    StHeaderMenuOptionSchema.prototype.label;
    /** @type {?} */
    StHeaderMenuOptionSchema.prototype.subMenus;
    /** @type {?} */
    StHeaderMenuOptionSchema.prototype.link;
    /** @type {?} */
    StHeaderMenuOptionSchema.prototype.external;
    /** @type {?} */
    StHeaderMenuOptionSchema.prototype.openInNewPage;
}
export class StHeaderSubMenuOptionSchema {
}
if (false) {
    /** @type {?} */
    StHeaderSubMenuOptionSchema.prototype.label;
    /** @type {?} */
    StHeaderSubMenuOptionSchema.prototype.link;
    /** @type {?} */
    StHeaderSubMenuOptionSchema.prototype.external;
    /** @type {?} */
    StHeaderSubMenuOptionSchema.prototype.openInNewPage;
}
