// @stratio/egeo/esm2015/lib/st-bubble-on-ellipsis/st-bubble-on-ellipsis.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/st-bubble-on-ellipsis/st-bubble-on-ellipsis.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/*
 * © 2017 Stratio Big Data Inc., Sucursal en España.
 *
 * This software is licensed under the Apache License, Version 2.0.
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the terms of the License for more details.
 *
 * SPDX-License-Identifier: Apache-2.0.
 */
import { ChangeDetectorRef, Component, ElementRef, Input, ViewChild } from '@angular/core';
import { get as _get } from 'lodash';
/**
 * \@description {Component} [StBubbleOnEllipsis]
 *
 * This component displays a bubble below a content if its width is longer than container
 *
 * \@example
 *
 * {html}
 *
 * ```
 *      <st-bubble-on-ellipsis
 *         [text]="'Text for bubble'"
 *         [openToLeft]="true"
 *         [maxWidth]="'40%'">
 *          Any text
 *      </st-bubble-on-ellipsis>
 * ```
 */
export class StBubbleOnEllipsisComponent {
    /**
     * @param {?} _cd
     */
    constructor(_cd) {
        this._cd = _cd;
        /**
         * \@Input {string} [lines=1] number of lines where ellipsis is placed
         */
        this.lines = 1;
    }
    /**
     * \@Input {boolean} [openToLeft=true] when true, bubble is displayed with the arrow to the right
     * @return {?}
     */
    get openToLeft() {
        return this._openToLeft;
    }
    /**
     * @param {?} _openToLeft
     * @return {?}
     */
    set openToLeft(_openToLeft) {
        this._openToLeft = _openToLeft;
        this.offset = this._openToLeft ? { x: 38, y: 7 } : { x: 0, y: 7 };
        this._cd.markForCheck();
    }
    /**
     * @return {?}
     */
    onShowBubble() {
        if (_get(this.bubbleTrigger, 'nativeElement.parentElement.offsetWidth') < _get(this.bubbleTrigger, 'nativeElement.parentElement.scrollWidth')
            || _get(this.bubbleTrigger, 'nativeElement.parentElement.offsetHeight') < _get(this.bubbleTrigger, 'nativeElement.parentElement.scrollHeight')) {
            this.visible = true;
        }
    }
    /**
     * @return {?}
     */
    onHideBubble() {
        this.visible = false;
    }
}
StBubbleOnEllipsisComponent.decorators = [
    { type: Component, args: [{
                selector: 'st-bubble-on-ellipsis',
                template: "<!--\n\n    \u00A9 2017 Stratio Big Data Inc., Sucursal en Espa\u00F1a.\n\n    This software is licensed under the Apache License, Version 2.0.\n    This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;\n    without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.\n    See the terms of the License for more details.\n\n    SPDX-License-Identifier: Apache-2.0.\n\n-->\n\n<span class=\"bubble-trigger\" #bubbleTrigger\n      (mouseenter)=\"onShowBubble()\"\n      (mouseleave)=\"onHideBubble()\">\n   <ng-content>\n   </ng-content>\n</span>\n\n<st-bubble class=\"bubble\" [ngClass]=\"{'bubble--open-to-left': openToLeft}\"\n           [hidden]=\"!visible\"\n           [small]=\"true\"\n           [text]=\"text\"\n           [animation]=\"false\"\n           [showArrow]=\"true\"\n           [offset]=\"offset\"\n           [openToLeft]=\"openToLeft\"\n           [minWidth]=\"minWidth\"\n           [maxWidth]=\"maxWidth\"\n>\n   <span bubble-button></span>\n</st-bubble>\n\n",
                host: {
                    '[class.multi-line]': 'lines > 1',
                    '[style.-webkit-line-clamp]': 'lines'
                },
                styles: ["@charset \"UTF-8\";:host{display:block;overflow:hidden;text-overflow:ellipsis;white-space:nowrap}:host.multi-line{display:block;display:-webkit-box;word-break:break-word;-webkit-box-orient:vertical;overflow:hidden;text-overflow:ellipsis;white-space:normal}.bubble-trigger{text-overflow:ellipsis}.bubble{display:block}.bubble--open-to-left{margin-left:50%}"]
            }] }
];
/** @nocollapse */
StBubbleOnEllipsisComponent.ctorParameters = () => [
    { type: ChangeDetectorRef }
];
StBubbleOnEllipsisComponent.propDecorators = {
    text: [{ type: Input }],
    minWidth: [{ type: Input }],
    maxWidth: [{ type: Input }],
    lines: [{ type: Input }],
    bubbleTrigger: [{ type: ViewChild, args: ['bubbleTrigger', { static: false },] }],
    openToLeft: [{ type: Input }]
};
if (false) {
    /**
     * \@Input {string} [text=] Text of the bubble
     * @type {?}
     */
    StBubbleOnEllipsisComponent.prototype.text;
    /**
     * \@Input {string} [minWidth=] min width for bubble
     * @type {?}
     */
    StBubbleOnEllipsisComponent.prototype.minWidth;
    /**
     * \@Input {string} [maxWidth=] max width for bubble
     * @type {?}
     */
    StBubbleOnEllipsisComponent.prototype.maxWidth;
    /**
     * \@Input {string} [lines=1] number of lines where ellipsis is placed
     * @type {?}
     */
    StBubbleOnEllipsisComponent.prototype.lines;
    /** @type {?} */
    StBubbleOnEllipsisComponent.prototype.bubbleTrigger;
    /** @type {?} */
    StBubbleOnEllipsisComponent.prototype.offset;
    /** @type {?} */
    StBubbleOnEllipsisComponent.prototype.visible;
    /**
     * @type {?}
     * @private
     */
    StBubbleOnEllipsisComponent.prototype._openToLeft;
    /**
     * @type {?}
     * @private
     */
    StBubbleOnEllipsisComponent.prototype._cd;
}
