// @stratio/egeo/esm5/lib/st-menu/st-menu.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/st-menu/st-menu.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/*
 * © 2017 Stratio Big Data Inc., Sucursal en España.
 *
 * This software is licensed under the Apache License, Version 2.0.
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the terms of the License for more details.
 *
 * SPDX-License-Identifier: Apache-2.0.
 */
import { Component, Output, EventEmitter, Input, ChangeDetectionStrategy, ChangeDetectorRef } from '@angular/core';
import { FormControl } from '@angular/forms';
import { StMenuPosition } from './st-menu.model';
/**
 * \@description {Component} Menu
 *
 * This components shows a dropdown menu with nested submenus
 *
 * \@model
 *
 *   [Menu model] {./st-menu.model.ts#StMenuModel}
 *   [Menu status] {./st-menu.model.ts#StMenuStatus}
 *   [Menu position] {./st-menu.model.ts#StMenuPosition}
 *
 * \@example
 *
 * {html}
 *
 * ```
 *    <st-menu
 *       qaTag="menu-example"
 *       [menuOptions]="options"
 *       [dynamicHeight]="true"
 *       (selectedOption)="onSelectMenuOption(event)">
 *    </st-menu>
 * ```
 * @template T
 */
var StMenuComponent = /** @class */ (function () {
    function StMenuComponent(_cd) {
        this._cd = _cd;
        /**
         * \@input {StMenuPosition} current menu position: left | right
         */
        this.position = StMenuPosition.right;
        /**
         * \@input {string} search input custom label
         */
        this.searchLabel = 'Search';
        /**
         * \@input {boolean} shows menu dropdown animation
         */
        this.animate = true;
        /**
         * \@input {boolean} disable menu toggle
         */
        this.disabled = false;
        /**
         * \@output {StMenuModel<T>} [selectedOption=''] Notify the selected menu option
         */
        this.selectedOption = new EventEmitter();
        /**
         * \@output {string} [searchChange=''] Notify when the search input value changes
         */
        this.searchChange = new EventEmitter();
        this.showMenu = false;
    }
    /**
     * @return {?}
     */
    StMenuComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        if (this.search) {
            this.searchFormControl = new FormControl('');
            this._formControlSubscription = this.searchFormControl.valueChanges.subscribe((/**
             * @param {?} value
             * @return {?}
             */
            function (value) {
                _this.searchChange.emit(value);
            }));
        }
    };
    /**
     * @return {?}
     */
    StMenuComponent.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        if (this._formControlSubscription) {
            this._formControlSubscription.unsubscribe();
        }
    };
    /**
     * Toggle menu activation
     * @param event
     */
    /**
     * Toggle menu activation
     * @param {?} event
     * @return {?}
     */
    StMenuComponent.prototype.activateMenu = /**
     * Toggle menu activation
     * @param {?} event
     * @return {?}
     */
    function (event) {
        var _this = this;
        if (this.showMenu) {
            this.hideMenu();
        }
        else {
            setTimeout((/**
             * @return {?}
             */
            function () {
                _this.showMenu = true;
                _this._cd.markForCheck();
            }));
        }
    };
    /**
     * Emits the selected menu option from childs
     * @param event: StMenuModel<T>
     */
    /**
     * Emits the selected menu option from childs
     * @param {?} event
     * @return {?}
     */
    StMenuComponent.prototype.selectedMenuOption = /**
     * Emits the selected menu option from childs
     * @param {?} event
     * @return {?}
     */
    function (event) {
        this.selectedOption.emit(event);
        this.hideMenu();
    };
    /**
     * Hides the menu and reset the search input value if it's enabled
     */
    /**
     * Hides the menu and reset the search input value if it's enabled
     * @return {?}
     */
    StMenuComponent.prototype.hideMenu = /**
     * Hides the menu and reset the search input value if it's enabled
     * @return {?}
     */
    function () {
        this.showMenu = false;
        if (this.searchFormControl) {
            this.searchFormControl.setValue('');
        }
        this._cd.markForCheck();
    };
    StMenuComponent.decorators = [
        { type: Component, args: [{
                    selector: 'st-menu',
                    template: "<div class=\"st-menu\" [ngClass]=\"{'st-menu--disabled': disabled}\">\n  <div class=\"st-menu__action\">\n    <div class=\"st-menu__link\"\n      [ngClass]=\"{'st-menu__button--disabled': disabled}\"\n      [id]=\"qaTag\"\n      (click)=\"!disabled && activateMenu($event)\" >\n      <ng-content></ng-content>\n    </div>\n    <div class=\"st-menu__options\"\n      *ngIf=\"showMenu\"\n      clickOutside\n      (clickOutside)=\"hideMenu()\"\n      [ngClass]=\"{'st-menu__options--left': position === 'left'}\">\n      <st-menu-options\n        [searchFormControl]=\"searchFormControl\"\n        [searchLabel]=\"searchLabel\"\n        [menuOptions]=\"menuOptions\"\n        [animate]=\"animate\"\n        [dynamicHeight]=\"dynamicHeight\"\n        [activeOption]=\"activeOption\"\n        [position]=\"position\"\n        [marginTop]=\"marginTop\"\n        [qaTag]=\"qaTag\"\n        (selectedOption)=\"selectedMenuOption($event)\"></st-menu-options>\n    </div>\n  </div>\n</div>\n",
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    styles: ["@charset \"UTF-8\";:host{display:inline-block}.st-menu__action{position:relative;display:inline-block}.st-menu__link{cursor:pointer;display:inline-block;position:relative}.st-menu__options{position:absolute;z-index:7500;-webkit-transform:translate3d(0,0,0);top:0;height:100%;margin-top:5px}.st-menu__options--left{right:0}.st-menu--disabled{cursor:not-allowed}.st-menu__button--disabled{pointer-events:none}"]
                }] }
    ];
    /** @nocollapse */
    StMenuComponent.ctorParameters = function () { return [
        { type: ChangeDetectorRef }
    ]; };
    StMenuComponent.propDecorators = {
        qaTag: [{ type: Input }],
        menuOptions: [{ type: Input }],
        activeOption: [{ type: Input }],
        position: [{ type: Input }],
        search: [{ type: Input }],
        searchLabel: [{ type: Input }],
        dynamicHeight: [{ type: Input }],
        marginTop: [{ type: Input }],
        animate: [{ type: Input }],
        disabled: [{ type: Input }],
        selectedOption: [{ type: Output }],
        searchChange: [{ type: Output }]
    };
    return StMenuComponent;
}());
export { StMenuComponent };
if (false) {
    /**
     * \@input {string} [qaTag=''] For set id for tests
     * @type {?}
     */
    StMenuComponent.prototype.qaTag;
    /**
     * \@input {StMenuModel<T>[]} list of menu items to show
     * @type {?}
     */
    StMenuComponent.prototype.menuOptions;
    /**
     * \@input {any} current selected option
     * @type {?}
     */
    StMenuComponent.prototype.activeOption;
    /**
     * \@input {StMenuPosition} current menu position: left | right
     * @type {?}
     */
    StMenuComponent.prototype.position;
    /**
     * \@input {boolean} show search input
     * @type {?}
     */
    StMenuComponent.prototype.search;
    /**
     * \@input {string} search input custom label
     * @type {?}
     */
    StMenuComponent.prototype.searchLabel;
    /**
     * \@input {boolean} Show menu in full height mode
     * @type {?}
     */
    StMenuComponent.prototype.dynamicHeight;
    /**
     * \@input {number} adjust menu options position
     * @type {?}
     */
    StMenuComponent.prototype.marginTop;
    /**
     * \@input {boolean} shows menu dropdown animation
     * @type {?}
     */
    StMenuComponent.prototype.animate;
    /**
     * \@input {boolean} disable menu toggle
     * @type {?}
     */
    StMenuComponent.prototype.disabled;
    /**
     * \@output {StMenuModel<T>} [selectedOption=''] Notify the selected menu option
     * @type {?}
     */
    StMenuComponent.prototype.selectedOption;
    /**
     * \@output {string} [searchChange=''] Notify when the search input value changes
     * @type {?}
     */
    StMenuComponent.prototype.searchChange;
    /** @type {?} */
    StMenuComponent.prototype.showMenu;
    /** @type {?} */
    StMenuComponent.prototype.searchFormControl;
    /**
     * @type {?}
     * @private
     */
    StMenuComponent.prototype._formControlSubscription;
    /**
     * @type {?}
     * @private
     */
    StMenuComponent.prototype._cd;
}
