// @stratio/egeo/esm2015/lib/utils/event-window-manager.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/utils/event-window-manager.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @abstract
 */
export class EventWindowManager {
    /**
     * @param {?} _renderer
     * @param {?} _cd
     */
    constructor(_renderer, _cd) {
        this._renderer = _renderer;
        this._cd = _cd;
    }
    /**
     * @protected
     * @return {?}
     */
    openElement() {
        if (this.isActive) {
            this.closeElement();
        }
        else {
            this.isActive = !this.isActive;
            setTimeout((/**
             * @return {?}
             */
            () => {
                this.globalListener = this._renderer.listen('document', 'click', this.onClickOutside.bind(this));
                this.forceClose = this._renderer.listen('document', 'stForceClose', this.onForceClose.bind(this));
            }));
        }
    }
    /**
     * @protected
     * @return {?}
     */
    closeElement() {
        if (this.isActive) {
            this.isActive = !this.isActive;
            this._cd.markForCheck();
            if (this.globalListener !== undefined && typeof (this.globalListener) === 'function') {
                this.globalListener();
                this.forceClose();
            }
        }
    }
    /**
     * @private
     * @param {?} event
     * @return {?}
     */
    onClickOutside(event) {
        if (!this._eventElementRef.nativeElement.contains(event.target)) {
            this.closeElement();
        }
    }
    /**
     * @private
     * @param {?} event
     * @return {?}
     */
    onForceClose(event) {
        this.closeElement();
    }
}
if (false) {
    /** @type {?} */
    EventWindowManager.prototype.isActive;
    /** @type {?} */
    EventWindowManager.prototype._eventElementRef;
    /**
     * @type {?}
     * @private
     */
    EventWindowManager.prototype.globalListener;
    /**
     * @type {?}
     * @private
     */
    EventWindowManager.prototype.forceClose;
    /**
     * @type {?}
     * @private
     */
    EventWindowManager.prototype._renderer;
    /**
     * @type {?}
     * @private
     */
    EventWindowManager.prototype._cd;
    /**
     * @abstract
     * @return {?}
     */
    EventWindowManager.prototype.ngOnDestroy = function () { };
}
