// @stratio/egeo/esm2015/lib/utils/demo-generator/demo-generator.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/utils/demo-generator/demo-generator.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/*
 * © 2017 Stratio Big Data Inc., Sucursal en España.
 *
 * This software is licensed under the Apache License, Version 2.0.
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the terms of the License for more details.
 *
 * SPDX-License-Identifier: Apache-2.0.
 */
import { ChangeDetectionStrategy, Component, ComponentFactoryResolver, ViewChild, ViewContainerRef } from '@angular/core';
import { DemoGeneratorProviders } from './demo-generator.interface';
export class StDemoGenerator {
    /**
     * @param {?} cfr
     * @param {?} config
     */
    constructor(cfr, config) {
        this.cfr = cfr;
        this.config = config;
        this.inputs = config.inputs;
        this.outputs = config.outputs;
        this.component = config.components[0];
    }
    /**
     * DYNAMIC BODY DEMO COMPONENT LOAD
     * @return {?}
     */
    ngOnDestroy() {
        if (this.componentRef) {
            this.componentRef.destroy();
        }
    }
    /**
     * @return {?}
     */
    ngAfterViewInit() {
        if (this.component) {
            this.loadDemo();
        }
    }
    /**
     * @private
     * @return {?}
     */
    loadDemo() {
        if (!this.componentRef) {
            this.target.clear();
            /** @type {?} */
            let compFactory = this.cfr.resolveComponentFactory(this.component);
            this.componentRef = this.target.createComponent(compFactory);
            this.bindDemoInputsAndOutputs();
        }
    }
    // tslint:disable:
    /**
     * @private
     * @return {?}
     */
    bindDemoInputsAndOutputs() {
        // Inputs its not necessary to be defined
        if (this.inputs) {
            Object.keys(this.inputs).map((/**
             * @param {?} key
             * @return {?}
             */
            key => this.componentRef.instance[key] = ((/** @type {?} */ (this.inputs)))[key]));
        }
        // Inputs must be defined for subscribe to them
        if (this.outputs) {
            Object.keys(this.outputs).map((/**
             * @param {?} key
             * @return {?}
             */
            key => {
                if (((/** @type {?} */ (this.componentRef.instance))).hasOwnProperty(key)) {
                    this.componentRef.instance[key].subscribe(((/** @type {?} */ (this.outputs)))[key]);
                }
            }));
        }
        this.componentRef.changeDetectorRef.detectChanges();
    }
}
StDemoGenerator.decorators = [
    { type: Component, args: [{
                selector: 'st-demo-body',
                template: '<div #stDemoBody></div>',
                changeDetection: ChangeDetectionStrategy.OnPush,
                styles: [`
      :host {
         width: 100%;
         display: block;
      }
   `]
            }] }
];
/** @nocollapse */
StDemoGenerator.ctorParameters = () => [
    { type: ComponentFactoryResolver },
    { type: DemoGeneratorProviders }
];
StDemoGenerator.propDecorators = {
    target: [{ type: ViewChild, args: ['stDemoBody', { read: ViewContainerRef, static: false },] }]
};
if (false) {
    /** @type {?} */
    StDemoGenerator.prototype.target;
    /** @type {?} */
    StDemoGenerator.prototype.inputs;
    /** @type {?} */
    StDemoGenerator.prototype.outputs;
    /** @type {?} */
    StDemoGenerator.prototype.component;
    /**
     * @type {?}
     * @private
     */
    StDemoGenerator.prototype.componentRef;
    /**
     * @type {?}
     * @private
     */
    StDemoGenerator.prototype.cfr;
    /**
     * @type {?}
     * @private
     */
    StDemoGenerator.prototype.config;
}
