// @stratio/egeo/esm5/lib/st-form/st-form.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/st-form/st-form.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/*
 * © 2017 Stratio Big Data Inc., Sucursal en España.
 *
 * This software is licensed under the Apache License, Version 2.0.
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the terms of the License for more details.
 *
 * SPDX-License-Identifier: Apache-2.0.
 */
import { ChangeDetectionStrategy, Component, EventEmitter, forwardRef, Input, Output, ViewChild } from '@angular/core';
import { NG_VALIDATORS, NG_VALUE_ACCESSOR, NgForm } from '@angular/forms';
import { FORM_UI_COMPONENT } from './shared/ui-component.interface';
/**
 * \@description {Component} [Dynamic form]
 *
 * The form component allows to generate forms dynamically using a JSON schema.
 *
 * \@example
 *
 * {html}
 *
 * ```
 * <st-form [schema]="jsonSchema" [(ngModel)]="model" #formModel="ngModel">
 * </st-form>
 *
 * ```
 *
 */
var StFormComponent = /** @class */ (function () {
    function StFormComponent() {
        /**
         * \@Input {string} [nestingLevel=0] This informs about the nesting level of the form. This input is only used for design purposes
         */
        this.nestingLevel = 0;
        /**
         * \@Input {string} [disabledSectionMessage='for this instance is disabled.']
         *  Message displayed when a section is disabled. This is always displayed after the section name
         */
        this.disabledSectionMessage = 'for this instance is disabled.';
        /**
         * \@Input {number} [sectionDescriptionLevel=-1] Level of the section to which description is displayed.
         * By default, section descriptions are not displayed.
         */
        this.sectionDescriptionLevel = -1;
        /**
         * \@Input {boolean} [showTooltips=-1] Enable or disable displaying of tooltips
         * By default, tooltips are displayed
         */
        this.showTooltips = true;
        /**
         * \@Output {any} [valueChange=] Event emitted when value is changed. This emits the current form value
         */
        this.valueChange = new EventEmitter();
        this.showCollapsedSectionFields = false;
        this.innerValue = {};
        this._value = {};
        this._parentFieldSubscription = [];
        this.onTouched = (/**
         * @return {?}
         */
        function () {
        });
    }
    Object.defineProperty(StFormComponent.prototype, "schema", {
        /** @Input {JSONSchema4 [schema=] JSON schema needed to generate the form */
        get: /**
         * \@Input {JSONSchema4 [schema=] JSON schema needed to generate the form
         * @return {?}
         */
        function () {
            return this._schema;
        },
        set: /**
         * @param {?} schema
         * @return {?}
         */
        function (schema) {
            this._schema = schema;
            this._value = {};
            if (this._schema.dependencies) {
                this._parentFields = Object.keys(this._schema.dependencies);
            }
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    StFormComponent.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        if (!this.forceValidations) {
            this.form.form.markAsPristine();
        }
    };
    /**
     * @return {?}
     */
    StFormComponent.prototype.ngAfterViewChecked = /**
     * @return {?}
     */
    function () {
        var _this = this;
        if (this._parentFields && this.form.control && this.form.control.controls) {
            var _loop_1 = function (i) {
                /** @type {?} */
                var parentField = this_1._parentFields[i];
                if (!this_1._parentFieldSubscription[i] && this_1.form.control.controls[parentField]) {
                    this_1._parentFieldSubscription[i] = this_1.form.control.controls[this_1._parentFields[i]].valueChanges.subscribe((/**
                     * @param {?} value
                     * @return {?}
                     */
                    function (value) {
                        if (!value) {
                            /** @type {?} */
                            var childrenFields = _this.schema.dependencies[parentField];
                            for (var j = 0; j < childrenFields.length; ++j) {
                                if (_this.form.controls[childrenFields[j]]) {
                                    _this._value[childrenFields[j]] = undefined;
                                }
                            }
                        }
                    }));
                }
            };
            var this_1 = this;
            for (var i = 0; i < this._parentFields.length; ++i) {
                _loop_1(i);
            }
        }
    };
    // Function to call when the value changes.
    // Function to call when the value changes.
    /**
     * @param {?} _
     * @return {?}
     */
    StFormComponent.prototype.onChange = 
    // Function to call when the value changes.
    /**
     * @param {?} _
     * @return {?}
     */
    function (_) {
    };
    /**
     * @param {?} control
     * @return {?}
     */
    StFormComponent.prototype.validate = /**
     * @param {?} control
     * @return {?}
     */
    function (control) {
        var _this = this;
        /** @type {?} */
        var errors = null;
        if (this.form) {
            Object.keys(this.form.controls).forEach((/**
             * @param {?} propertyName
             * @return {?}
             */
            function (propertyName) {
                if (_this.form.controls[propertyName] && _this.form.controls[propertyName].errors) {
                    if (!errors) {
                        errors = {};
                    }
                    errors[propertyName] = _this.form.controls[propertyName].errors;
                }
            }));
            this.form.control.setErrors(errors);
        }
        return errors;
    };
    /**
     * @param {?} propertyName
     * @return {?}
     */
    StFormComponent.prototype.isRequired = /**
     * @param {?} propertyName
     * @return {?}
     */
    function (propertyName) {
        return propertyName && this.schema.required && this.schema.required.indexOf(propertyName) !== -1;
    };
    /**
     * @return {?}
     */
    StFormComponent.prototype.isCollapsedSection = /**
     * @return {?}
     */
    function () {
        return this.schema.type === 'object' && this.schema.ui &&
            (this.schema.ui.component === FORM_UI_COMPONENT.SHOW_MORE || this.schema.ui.component === FORM_UI_COMPONENT.ACCORDION);
    };
    /**
     * @return {?}
     */
    StFormComponent.prototype.iShowMoreSection = /**
     * @return {?}
     */
    function () {
        return this.schema.type === 'object' && this.schema.ui && this.schema.ui.component === FORM_UI_COMPONENT.SHOW_MORE;
    };
    /**
     * @param {?} propertyName
     * @return {?}
     */
    StFormComponent.prototype.isAParentField = /**
     * @param {?} propertyName
     * @return {?}
     */
    function (propertyName) {
        return this.schema.dependencies && this.schema.dependencies[propertyName] && this.schema.dependencies[propertyName].length > 0;
    };
    /**
     * @return {?}
     */
    StFormComponent.prototype.getOptionalButtonLabel = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var label = 'Additional options';
        if (this.parentName || this.schema.title) {
            label += ' of ';
            label += this.parentName || this.schema.title;
        }
        return label;
    };
    /**
     * @return {?}
     */
    StFormComponent.prototype.onChangeOptionalFieldsVisibility = /**
     * @return {?}
     */
    function () {
        this.showCollapsedSectionFields = !this.showCollapsedSectionFields;
    };
    /**
     * @param {?} propertyName
     * @return {?}
     */
    StFormComponent.prototype.getFieldClasses = /**
     * @param {?} propertyName
     * @return {?}
     */
    function (propertyName) {
        return {
            'hidden': this.isCollapsedSection() && !this.showCollapsedSectionFields,
            'parent-field': this.isAParentField(propertyName) || (this.isASwitchSection() && this.isTheFirstField(propertyName))
        };
    };
    /**
     * @param {?} propertyName
     * @return {?}
     */
    StFormComponent.prototype.fieldHasToBeCreated = /**
     * @param {?} propertyName
     * @return {?}
     */
    function (propertyName) {
        /** @type {?} */
        var createField = true;
        /** @type {?} */
        var parentField = this.getParentField(propertyName);
        if (((parentField && !this._value[parentField]) || (this.isInADisabledSection() && !this.isTheFirstField(propertyName)))
            || !this.fulfillDependencyVisibility(propertyName)) {
            createField = false;
            this._value[propertyName] = undefined;
        }
        return createField;
    };
    /**
     * @param {?} propertyName
     * @return {?}
     */
    StFormComponent.prototype.isRelatedField = /**
     * @param {?} propertyName
     * @return {?}
     */
    function (propertyName) {
        return this.schema.properties[propertyName].ui && this.schema.properties[propertyName].ui.relatedTo;
    };
    // When value is received from outside
    // When value is received from outside
    /**
     * @param {?} value
     * @return {?}
     */
    StFormComponent.prototype.writeValue = 
    // When value is received from outside
    /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        if (value) {
            this.onChange(value);
            this.innerValue = value;
        }
    };
    /**
     * @param {?} value
     * @param {?} property
     * @return {?}
     */
    StFormComponent.prototype.onChangeProperty = /**
     * @param {?} value
     * @param {?} property
     * @return {?}
     */
    function (value, property) {
        var _this = this;
        this._value[property] = value;
        this.valueChange.emit(this._value);
        setTimeout((/**
         * @return {?}
         */
        function () {
            _this.onChange(_this._value);
        }));
    };
    // Registry the change function to propagate internal model changes
    // Registry the change function to propagate internal model changes
    /**
     * @param {?} fn
     * @return {?}
     */
    StFormComponent.prototype.registerOnChange = 
    // Registry the change function to propagate internal model changes
    /**
     * @param {?} fn
     * @return {?}
     */
    function (fn) {
        this.onChange = fn;
    };
    // Registry the touch function to propagate internal touch events TODO: make this function.
    // Registry the touch function to propagate internal touch events TODO: make this function.
    /**
     * @param {?} fn
     * @return {?}
     */
    StFormComponent.prototype.registerOnTouched = 
    // Registry the touch function to propagate internal touch events TODO: make this function.
    /**
     * @param {?} fn
     * @return {?}
     */
    function (fn) {
        this.onTouched = fn;
    };
    // Allows Angular to disable the form.
    // Allows Angular to disable the form.
    /**
     * @param {?} disable
     * @return {?}
     */
    StFormComponent.prototype.setDisabledState = 
    // Allows Angular to disable the form.
    /**
     * @param {?} disable
     * @return {?}
     */
    function (disable) {
        if (disable) {
            this.form.control.disable();
        }
        else {
            this.form.control.enable();
        }
    };
    /**
     * @return {?}
     */
    StFormComponent.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        if (this._parentFieldSubscription) {
            for (var i = 0; i < this._parentFieldSubscription.length; ++i) {
                this._parentFieldSubscription[i].unsubscribe();
            }
        }
    };
    /**
     * @return {?}
     */
    StFormComponent.prototype.isInADisabledSection = /**
     * @return {?}
     */
    function () {
        if (this.isASwitchSection()) {
            /** @type {?} */
            var sectionEnabler = Object.keys(this.schema.properties)[0];
            return !(this.form && this.form.controls[sectionEnabler] && this.form.controls[sectionEnabler].value);
        }
        return false;
    };
    /**
     * @return {?}
     */
    StFormComponent.prototype.isASwitchSection = /**
     * @return {?}
     */
    function () {
        return this.schema.ui && this.schema.ui.component === FORM_UI_COMPONENT.SWITCH;
    };
    /**
     * @return {?}
     */
    StFormComponent.prototype.isAnAccordion = /**
     * @return {?}
     */
    function () {
        return this.schema.ui && this.schema.ui.component === FORM_UI_COMPONENT.ACCORDION;
    };
    /**
     * @return {?}
     */
    StFormComponent.prototype.onClickTitle = /**
     * @return {?}
     */
    function () {
        if (this.isAnAccordion()) {
            this.onChangeOptionalFieldsVisibility();
        }
    };
    /**
     * @param {?} propertyName
     * @return {?}
     */
    StFormComponent.prototype.isTheFirstField = /**
     * @param {?} propertyName
     * @return {?}
     */
    function (propertyName) {
        return propertyName === Object.keys(this.schema.properties)[0];
    };
    /**
     * @private
     * @param {?} propertyName
     * @return {?}
     */
    StFormComponent.prototype.getParentField = /**
     * @private
     * @param {?} propertyName
     * @return {?}
     */
    function (propertyName) {
        var _this = this;
        /** @type {?} */
        var parentField = undefined;
        if (this.schema.dependencies) {
            Object.keys(this.schema.dependencies).forEach((/**
             * @param {?} key
             * @return {?}
             */
            function (key) {
                if (_this.schema.dependencies[key].indexOf(propertyName) !== -1) {
                    parentField = key;
                }
            }));
        }
        return parentField;
    };
    /**
     * @private
     * @param {?} propertyName
     * @return {?}
     */
    StFormComponent.prototype.fulfillDependencyVisibility = /**
     * @private
     * @param {?} propertyName
     * @return {?}
     */
    function (propertyName) {
        /** @type {?} */
        var fulfill = true;
        /** @type {?} */
        var propertySchema = this.schema.properties[propertyName];
        if (propertySchema.ui && propertySchema.ui.visible) {
            /** @type {?} */
            var keys = Object.keys(propertySchema.ui.visible);
            /** @type {?} */
            var i = 0;
            while (fulfill && i < keys.length && propertySchema.ui.visible.hasOwnProperty(keys[i])) {
                if (this._value[keys[i]] !== propertySchema.ui.visible[keys[i]]) {
                    fulfill = false;
                }
                ++i;
            }
        }
        return fulfill;
    };
    StFormComponent.decorators = [
        { type: Component, args: [{
                    selector: 'st-form',
                    template: "<!--\n\n    \u00A9 2017 Stratio Big Data Inc., Sucursal en Espa\u00F1a.\n\n    This software is licensed under the Apache License, Version 2.0.\n    This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;\n    without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.\n    See the terms of the License for more details.\n\n    SPDX-License-Identifier: Apache-2.0.\n\n-->\n\n<form #form=\"ngForm\" *ngIf=\"innerValue\"\n      [ngClass]=\"{'category': nestingLevel === 0,\n                  'section': nestingLevel === 1,\n                  'subsection': nestingLevel > 1,\n                  'accordion': isAnAccordion(),\n                   'accordion--open': isAnAccordion() && showCollapsedSectionFields}\"\n      novalidate>\n   <h1 *ngIf=\"schema.title && !isASwitchSection()\" class=\"title\" (click)=\"onClickTitle()\">{{schema.title}}</h1>\n   <p *ngIf=\"schema.description && nestingLevel <= sectionDescriptionLevel\" class=\"description\">{{schema.description}}</p>\n   <span class=\"optional-fields-button\" *ngIf=\"iShowMoreSection()\">\n   <button class=\"button button-link\"\n           (click)=\"onChangeOptionalFieldsVisibility()\">\n      <span> <i\n         [ngClass]=\"{'icon-arrow2_down': showCollapsedSectionFields, 'icon-arrow2_right': !showCollapsedSectionFields}\"></i> {{getOptionalButtonLabel()}}</span>\n   </button>\n   </span>\n   <ng-container *ngFor=\"let property of schema.properties | stObjectToArray; let index = index;\">\n      <div class=\"form-field-container\">\n         <st-form-field *ngIf=\"property.value.type !== 'object' && fieldHasToBeCreated(property.key)\" class=\"form-field\"\n                        [schema]=\"property\"\n                        [name]=\"property.key\"\n                        [maxWidth]=\"textFieldMaxWidth\"\n                        [ngModel]=\"innerValue[property.key]\"\n                        [forceValidations]=\"forceValidations\"\n                        (ngModelChange)=\"onChangeProperty($event, property.key)\"\n                        [required]=\"isRequired(property.key)\"\n                        [hasDependencies]=\"isAParentField(property.key) || (isASwitchSection() && isTheFirstField(property.key))\"\n                        [ngClass]=\"getFieldClasses(property.key)\"\n                        [showTooltip]=\"showTooltips\"\n                        [qaTag]=\"property.key\"\n                        [errorMessages]=\"errorMessages\"\n                        [attr.id]=\"property.key + '-form-field'\">\n         </st-form-field>\n\n         <div *ngIf=\"property.value.type === 'object' && fieldHasToBeCreated(property.key)\">\n            <st-form\n               [hidden]=\"isCollapsedSection() && !showCollapsedSectionFields\"\n               [schema]=\"property.value\"\n               [parentName]=\"schema.title\"\n               [name]=\"property.key\"\n               [textFieldMaxWidth]=\"textFieldMaxWidth\"\n               [ngModel]=\"innerValue[property.key]\"\n               (ngModelChange)=\"onChangeProperty($event, property.key)\"\n               [attr.id]=\"property.key + '-section'\"\n               [nestingLevel]=\"nestingLevel + 1\"\n               [showTooltips]=\"showTooltips\"\n               [sectionDescriptionLevel]=\"sectionDescriptionLevel\"\n               [errorMessages]=\"errorMessages\">\n            </st-form>\n         </div>\n      </div>\n      <span class=\"line-break\" *ngIf=\"!isRelatedField(property.key)\"></span>\n   </ng-container>\n\n   <p *ngIf=\"isInADisabledSection()\" class=\"disabled-section-info\"> {{schema.title}} {{disabledSectionMessage}}</p>\n</form>\n",
                    host: { class: 'st-form' },
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    providers: [
                        { provide: NG_VALUE_ACCESSOR, useExisting: forwardRef((/**
                             * @return {?}
                             */
                            function () { return StFormComponent; })), multi: true },
                        { provide: NG_VALIDATORS, useExisting: forwardRef((/**
                             * @return {?}
                             */
                            function () { return StFormComponent; })), multi: true }
                    ],
                    styles: ["@charset \"UTF-8\";.form-field{display:block}.form-field.hidden{visibility:hidden;position:absolute}.section-title{padding:10px 0 20px}"]
                }] }
    ];
    StFormComponent.propDecorators = {
        parentName: [{ type: Input }],
        nestingLevel: [{ type: Input }],
        forceValidations: [{ type: Input }],
        disabledSectionMessage: [{ type: Input }],
        sectionDescriptionLevel: [{ type: Input }],
        textFieldMaxWidth: [{ type: Input }],
        errorMessages: [{ type: Input }],
        showTooltips: [{ type: Input }],
        valueChange: [{ type: Output }],
        form: [{ type: ViewChild, args: ['form', { static: false },] }],
        schema: [{ type: Input }]
    };
    return StFormComponent;
}());
export { StFormComponent };
if (false) {
    /**
     * \@Input {string} [parentName=] Name of the parent section. By default, it is undefined
     * @type {?}
     */
    StFormComponent.prototype.parentName;
    /**
     * \@Input {string} [nestingLevel=0] This informs about the nesting level of the form. This input is only used for design purposes
     * @type {?}
     */
    StFormComponent.prototype.nestingLevel;
    /**
     * \@Input {boolean} [forceValidations=] Boolean to force the field validations
     * @type {?}
     */
    StFormComponent.prototype.forceValidations;
    /**
     * \@Input {string} [disabledSectionMessage='for this instance is disabled.']
     *  Message displayed when a section is disabled. This is always displayed after the section name
     * @type {?}
     */
    StFormComponent.prototype.disabledSectionMessage;
    /**
     * \@Input {number} [sectionDescriptionLevel=-1] Level of the section to which description is displayed.
     * By default, section descriptions are not displayed.
     * @type {?}
     */
    StFormComponent.prototype.sectionDescriptionLevel;
    /**
     * \@Input {number} [textFieldMaxWidth=] Maximum width of a field needed to paint a input or textarea
     * @type {?}
     */
    StFormComponent.prototype.textFieldMaxWidth;
    /**
     * \@Input {StInputError} [errorMessages=] Field error translations
     * @type {?}
     */
    StFormComponent.prototype.errorMessages;
    /**
     * \@Input {boolean} [showTooltips=-1] Enable or disable displaying of tooltips
     * By default, tooltips are displayed
     * @type {?}
     */
    StFormComponent.prototype.showTooltips;
    /**
     * \@Output {any} [valueChange=] Event emitted when value is changed. This emits the current form value
     * @type {?}
     */
    StFormComponent.prototype.valueChange;
    /** @type {?} */
    StFormComponent.prototype.form;
    /** @type {?} */
    StFormComponent.prototype.showCollapsedSectionFields;
    /** @type {?} */
    StFormComponent.prototype.innerValue;
    /**
     * @type {?}
     * @private
     */
    StFormComponent.prototype._value;
    /**
     * @type {?}
     * @private
     */
    StFormComponent.prototype._parentFieldSubscription;
    /**
     * @type {?}
     * @private
     */
    StFormComponent.prototype._parentFields;
    /**
     * @type {?}
     * @private
     */
    StFormComponent.prototype._schema;
    /** @type {?} */
    StFormComponent.prototype.onTouched;
}
