// @stratio/egeo/esm5/lib/st-tag-input/st-tag-input.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/st-tag-input/st-tag-input.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
/*
 * © 2017 Stratio Big Data Inc., Sucursal en España.
 *
 * This software is licensed under the Apache License, Version 2.0.
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the terms of the License for more details.
 *
 * SPDX-License-Identifier: Apache-2.0.
 */
import { ChangeDetectionStrategy, ChangeDetectorRef, Component, ElementRef, forwardRef, HostBinding, Input, ViewChild } from '@angular/core';
import { NG_VALIDATORS, NG_VALUE_ACCESSOR } from '@angular/forms';
/**
 * \@description {Component} Tag Input
 *
 * This component is a text input box that automatically creates tags out of a typed text.
 *
 * \@example
 *
 * {html}
 *
 * ```
 * <st-tag-input
 *    class="st-form-field"
 *    name="tag-input-reactive"
 *    formControlName="tag-input-reactive"
 *    [autocompleteList]="filteredlist"
 *    [withAutocomplete]="true"
 *    [disabled]="disabled"
 *    [label]="'Tag Input with Reactive Form'"
 *    [id]="'tag-input-reactive'"
 *    [placeholder]="'Add tags separated by commas'"
 *    [tooltip]="'This is a Tag Input component tooltip'"
 *    [forbiddenValues]="['test']"
 *    (input)="onFilterList($event)">
 * </st-tag-input>
 * <st-tag-input
 *    class="st-form-field"
 *    name="tag-input-template-driven"
 *    [(ngModel)]="tags.templateDriven"
 *    [autocompleteList]="filteredlist"
 *    [withAutocomplete]="true"
 *    [disabled]="disabled"
 *    [label]="'Tag Input with Template Driven Form'"
 *    [id]="'tag-input-template-driven'"
 *    [placeholder]="'Add tags separated by commas'"
 *    [tooltip]="'This is a Tag Input component tooltip'"
 *    [regularExpression]="pattern"
 *    (input)="onFilterList($event)">
 * </st-tag-input>
 * ```
 */
var StTagInputComponent = /** @class */ (function () {
    function StTagInputComponent(_selectElement, _cd) {
        this._selectElement = _selectElement;
        this._cd = _cd;
        /**
         * \@input {string | null} [label=null] Label to show over the input. It is empty by default
         */
        this.label = null;
        /**
         * \@input {string | null} [tooltip=null] The tooltip to show  over the label. It is empty by default
         */
        this.tooltip = null;
        /**
         * \@input {string | null} [placeholder=null] The text that appears as placeholder of the input. It is empty by default
         */
        this.placeholder = null;
        /**
         * \@input {string | null} [errorMessage=null] Error message to show. It is empty by default
         */
        this.errorMessage = null;
        /**
         * \@input {string | null} [type=null] Type of the items
         */
        this.type = 'text';
        /**
         * \@input {boolean} [withAutocomplete=false] Enable autocomplete feature. It is false by default
         */
        this.withAutocomplete = false;
        /**
         * \@input {(StDropDownMenuItem | StDropDownMenuGroup)[]} [autocompleteList=Array()] List to be used for autocomplete feature. It is empty by default
         */
        this.autocompleteList = [];
        /**
         * \@input {boolean} [charsToShowAutocompleteList=Array()] List to be used for autocomplete feature. It is empty by default
         */
        this.charsToShowAutocompleteList = 1;
        /**
         * \@input {boolean} [allowFreeText=true] Boolean to allow user to type a free text or not
         */
        this.allowFreeText = true;
        /**
         * \@input {string[]} [forbiddenValues=Array()] A list of values that user can not type and if he types one of them,
         * tag input will be invalid. It is empty by default
         */
        this.forbiddenValues = [];
        /**
         * \@input {string} [regularExpression=] Regular expression to validate values. It is null by default
         */
        this.regularExpression = null;
        /**
         * \@Input {boolean} [forceValidations=false] If you specify it to 'true', the tag input checks the errors before being modified by user
         */
        this.forceValidations = false;
        this.expandedMenu = false;
        this.items = [];
        this.innerInputContent = '';
        this.isPristine = true;
        this.showErrorValue = false;
        this._focus = false;
        this._isDisabled = false;
        this._newElementInput = null;
        this._selected = null;
        this.onChange = (/**
         * @param {?} _
         * @return {?}
         */
        function (_) {
        });
        this.onTouched = (/**
         * @return {?}
         */
        function () {
        });
    }
    Object.defineProperty(StTagInputComponent.prototype, "disabled", {
        /** @input {boolean} [disabled=false] Disable the component. It is false by default */
        get: /**
         * \@input {boolean} [disabled=false] Disable the component. It is false by default
         * @return {?}
         */
        function () {
            return this._isDisabled;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._isDisabled = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(StTagInputComponent.prototype, "hasLabel", {
        get: /**
         * @return {?}
         */
        function () {
            return this.label !== null && this.label.length > 0;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(StTagInputComponent.prototype, "hasFocus", {
        get: /**
         * @return {?}
         */
        function () {
            return this._focus;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(StTagInputComponent.prototype, "hasPlaceholder", {
        get: /**
         * @return {?}
         */
        function () {
            return !this._focus && !this.items.length && this.placeholder && this.placeholder.length > 0;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(StTagInputComponent.prototype, "hasAutocomplete", {
        get: /**
         * @return {?}
         */
        function () {
            return this.expandedMenu && this.autocompleteList && this.autocompleteList.length > 0;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(StTagInputComponent.prototype, "disableValue", {
        get: /**
         * @return {?}
         */
        function () {
            return this._isDisabled === true ? '' : null;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(StTagInputComponent.prototype, "isValidInput", {
        get: /**
         * @return {?}
         */
        function () {
            /** @type {?} */
            var isForbidden = this.forbiddenValues.length && this.forbiddenValues.indexOf(this.innerInputContent) > -1;
            /** @type {?} */
            var isDuplicated = this.items.indexOf(this.innerInputContent) !== -1;
            /** @type {?} */
            var matchedPattern = this.regularExpression ? this._regularExp.test(this.innerInputContent) : true;
            return this.innerInputContent.length ? !isForbidden && !isDuplicated && matchedPattern : true;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(StTagInputComponent.prototype, "tagSelected", {
        get: /**
         * @return {?}
         */
        function () {
            return this._selected;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(StTagInputComponent.prototype, "selectId", {
        get: /**
         * @return {?}
         */
        function () {
            /** @type {?} */
            var select = this._selectElement.nativeElement;
            return select.getAttribute('id') !== null ? select.id : null;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(StTagInputComponent.prototype, "inputId", {
        get: /**
         * @return {?}
         */
        function () {
            return this.selectId !== null ? this.selectId + "-input" : null;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(StTagInputComponent.prototype, "labelId", {
        get: /**
         * @return {?}
         */
        function () {
            return this.selectId !== null ? this.selectId + "-label" : null;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(StTagInputComponent.prototype, "tagId", {
        get: /**
         * @return {?}
         */
        function () {
            return this.selectId !== null ? this.selectId + "-tag-" : null;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(StTagInputComponent.prototype, "listId", {
        get: /**
         * @return {?}
         */
        function () {
            return this.selectId !== null ? this.selectId + "-autocomplete" : null;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    StTagInputComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this._newElementInput = this.newElementInput.nativeElement;
        switch (this.type) {
            case 'number': {
                this.regularExpression = '^\\d+(\\.\\d+)?$';
                break;
            }
            case 'integer': {
                this.regularExpression = '^\\d+$';
                break;
            }
            default: {
                break;
            }
        }
        this._regularExp = new RegExp(this.regularExpression);
        this.showErrorValue = this.showError();
        this._cd.markForCheck();
    };
    /**
     * @param {?} changes
     * @return {?}
     */
    StTagInputComponent.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        this.checkAutocompleteMenuChange(changes);
        if (this.forceValidations) {
            this.writeValue(this.items);
        }
        this._cd.markForCheck();
    };
    /**
     * @param {?} data
     * @return {?}
     */
    StTagInputComponent.prototype.writeValue = /**
     * @param {?} data
     * @return {?}
     */
    function (data) {
        var e_1, _a;
        if (data && Array.isArray(data) && data.length) {
            this.items = [];
            try {
                for (var data_1 = tslib_1.__values(data), data_1_1 = data_1.next(); !data_1_1.done; data_1_1 = data_1.next()) {
                    var value = data_1_1.value;
                    /** @type {?} */
                    var parsedValue = this._getParsedTag(value);
                    if (parsedValue || !isNaN(parsedValue)) {
                        this.items.push(parsedValue);
                    }
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (data_1_1 && !data_1_1.done && (_a = data_1.return)) _a.call(data_1);
                }
                finally { if (e_1) throw e_1.error; }
            }
            this.onChange(this.items);
            this.isPristine = false;
            this._cd.markForCheck();
        }
    };
    /**
     * @param {?} disabled
     * @return {?}
     */
    StTagInputComponent.prototype.setDisabledState = /**
     * @param {?} disabled
     * @return {?}
     */
    function (disabled) {
        this.disabled = disabled;
        this._cd.markForCheck();
    };
    /**
     * @param {?} control
     * @return {?}
     */
    StTagInputComponent.prototype.validate = /**
     * @param {?} control
     * @return {?}
     */
    function (control) {
    };
    // Registry the change function to propagate internal model changes
    // Registry the change function to propagate internal model changes
    /**
     * @param {?} fn
     * @return {?}
     */
    StTagInputComponent.prototype.registerOnChange = 
    // Registry the change function to propagate internal model changes
    /**
     * @param {?} fn
     * @return {?}
     */
    function (fn) {
        this.onChange = fn;
    };
    /**
     * @param {?} fn
     * @return {?}
     */
    StTagInputComponent.prototype.registerOnTouched = /**
     * @param {?} fn
     * @return {?}
     */
    function (fn) {
        this.onTouched = fn;
    };
    // Input actions
    // Input actions
    /**
     * @param {?} text
     * @return {?}
     */
    StTagInputComponent.prototype.onInputText = 
    // Input actions
    /**
     * @param {?} text
     * @return {?}
     */
    function (text) {
        this.innerInputContent = text;
        this.showAutocompleteMenu();
    };
    /**
     * @param {?} event
     * @return {?}
     */
    StTagInputComponent.prototype.onInputFocusIn = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        if (!this._isDisabled) {
            this._focus = true;
            this._newElementInput.focus();
            this._forceResetAutoCompleteList();
            this.showAutocompleteMenu();
        }
        event.stopPropagation();
    };
    /**
     * @param {?} event
     * @return {?}
     */
    StTagInputComponent.prototype.onInputFocusOut = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        if (!this.expandedMenu) {
            this._focus = false;
            this.addCurrentTag();
        }
        event.stopPropagation();
    };
    /**
     * @param {?} event
     * @return {?}
     */
    StTagInputComponent.prototype.onInputKeyDown = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        switch (event.keyCode) {
            case 188: // Comma
            case 13: // Enter
                if (this.innerInputContent.length && this.isValidInput) {
                    this.addTag(this.innerInputContent);
                    this._forceResetAutoCompleteList();
                }
                event.preventDefault();
                break;
            case 9: // Tab
                if (this.innerInputContent.length && this.isValidInput) {
                    this.addTag(this.innerInputContent);
                    event.preventDefault();
                }
                else if (this.innerInputContent.length) {
                    this.clearInput();
                }
                break;
            case 46: // Delete
                if (this.innerInputContent.length) {
                    this.clearInput();
                }
                else if (this.items.length) {
                    event.target.previousElementSibling.focus();
                }
                break;
            case 8: // Backspace
            case 37: // Left
                if (this.items.length && !this.innerInputContent.length) {
                    event.target.previousElementSibling.focus();
                }
                break;
            default:
                break;
        }
    };
    // Tag actions
    // Tag actions
    /**
     * @param {?} event
     * @param {?} index
     * @return {?}
     */
    StTagInputComponent.prototype.onTagKeyDown = 
    // Tag actions
    /**
     * @param {?} event
     * @param {?} index
     * @return {?}
     */
    function (event, index) {
        switch (event.keyCode) {
            case 8: // Backspace
            case 46: // Delete
                if (this._selected !== null) {
                    this.deleteTag(this._selected);
                    this._selected = null;
                    this._newElementInput.focus();
                }
                break;
            case 37: // Left
                if (this._selected > 0) {
                    event.target.previousElementSibling.focus();
                }
                break;
            case 39: // Right
                if (this._selected < this.items.length) {
                    event.target.nextElementSibling.focus();
                }
                break;
            default:
                break;
        }
    };
    /**
     * @param {?} event
     * @param {?} index
     * @return {?}
     */
    StTagInputComponent.prototype.onTagFocusIn = /**
     * @param {?} event
     * @param {?} index
     * @return {?}
     */
    function (event, index) {
        if (!this._isDisabled) {
            this._focus = true;
            this.addCurrentTag();
            this.expandedMenu = false;
            this._selected = index;
        }
    };
    /**
     * @param {?} event
     * @param {?} index
     * @return {?}
     */
    StTagInputComponent.prototype.onTagFocusOut = /**
     * @param {?} event
     * @param {?} index
     * @return {?}
     */
    function (event, index) {
        this._focus = false;
        this._selected = null;
    };
    /**
     * @param {?} event
     * @param {?} index
     * @return {?}
     */
    StTagInputComponent.prototype.onTagClick = /**
     * @param {?} event
     * @param {?} index
     * @return {?}
     */
    function (event, index) {
        event.stopPropagation();
        event.preventDefault();
    };
    // Dropdown actions
    // Dropdown actions
    /**
     * @param {?} data
     * @return {?}
     */
    StTagInputComponent.prototype.onListSelect = 
    // Dropdown actions
    /**
     * @param {?} data
     * @return {?}
     */
    function (data) {
        this._focus = false;
        if (data.value.length && this.items.indexOf(data.value) === -1) {
            this.addTag(data.value);
        }
        else {
            this.clearInput();
        }
    };
    /**
     * @param {?} event
     * @return {?}
     */
    StTagInputComponent.prototype.onClickOutside = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        if (this.expandedMenu) {
            this._focus = false;
            this.addCurrentTag();
        }
    };
    /**
     * @return {?}
     */
    StTagInputComponent.prototype.showError = /**
     * @return {?}
     */
    function () {
        return typeof this.errorMessage === 'string' && (!this.isPristine || this.forceValidations) && !this._focus && !this.disabled;
    };
    /**
     * @param {?} index
     * @return {?}
     */
    StTagInputComponent.prototype.deleteTag = /**
     * @param {?} index
     * @return {?}
     */
    function (index) {
        this.items.splice(index, 1);
        this.onChange(this.items);
        this._newElementInput.value = '';
        this.innerInputContent = '';
        this._newElementInput.dispatchEvent(new Event('input'));
    };
    /**
     * @private
     * @param {?} tag
     * @return {?}
     */
    StTagInputComponent.prototype.addTag = /**
     * @private
     * @param {?} tag
     * @return {?}
     */
    function (tag) {
        /** @type {?} */
        var parsedValue = this._getParsedTag(tag);
        if (parsedValue || !isNaN(parsedValue)) {
            this.items.push(parsedValue);
            this.clearInput();
            this.onChange(this.items);
            this.isPristine = false;
        }
        this.showErrorValue = this.showError();
    };
    /**
     * @private
     * @return {?}
     */
    StTagInputComponent.prototype.addCurrentTag = /**
     * @private
     * @return {?}
     */
    function () {
        if (this.innerInputContent.length && this.isValidInput) {
            this.addTag(this.innerInputContent);
        }
        else {
            this.clearInput();
        }
    };
    /**
     * @private
     * @return {?}
     */
    StTagInputComponent.prototype.clearInput = /**
     * @private
     * @return {?}
     */
    function () {
        if (this.expandedMenu) {
            this.expandedMenu = false;
        }
        this.innerInputContent = '';
        this._newElementInput.innerText = '';
    };
    /**
     * @private
     * @return {?}
     */
    StTagInputComponent.prototype.showAutocompleteMenu = /**
     * @private
     * @return {?}
     */
    function () {
        if (this.withAutocomplete && !this.expandedMenu && this.charsToShowAutocompleteList <= this.innerInputContent.length) {
            this.expandedMenu = true;
        }
        if (this.innerInputContent === '' && this.charsToShowAutocompleteList) {
            this.expandedMenu = false;
        }
        this._cd.markForCheck();
    };
    /**
     * @private
     * @param {?} changes
     * @return {?}
     */
    StTagInputComponent.prototype.checkAutocompleteMenuChange = /**
     * @private
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        if (changes && changes.autocompleteList) {
            this._cd.markForCheck();
        }
    };
    /**
     * @private
     * @param {?} tag
     * @return {?}
     */
    StTagInputComponent.prototype._getParsedTag = /**
     * @private
     * @param {?} tag
     * @return {?}
     */
    function (tag) {
        switch (this.type) {
            case 'number': {
                return parseFloat(tag);
            }
            case 'integer': {
                return parseInt(tag, 0);
            }
            default:
                return tag;
        }
    };
    /**
     * @private
     * @return {?}
     */
    StTagInputComponent.prototype._forceResetAutoCompleteList = /**
     * @private
     * @return {?}
     */
    function () {
        if (!this.charsToShowAutocompleteList && this.withAutocomplete) {
            this._newElementInput.innerText = '';
            /** @type {?} */
            var event_1 = new Event('input', {
                'bubbles': true
            });
            event_1.data = '';
            this._newElementInput.dispatchEvent(event_1);
        }
    };
    StTagInputComponent.decorators = [
        { type: Component, args: [{
                    selector: 'st-tag-input',
                    template: "<!--\n\n    \u00A9 2017 Stratio Big Data Inc., Sucursal en Espa\u00F1a.\n\n    This software is licensed under the Apache License, Version 2.0.\n    This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;\n    without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.\n    See the terms of the License for more details.\n\n    SPDX-License-Identifier: Apache-2.0.\n\n-->\n<st-dropdown-menu [items]=\"autocompleteList\" [attr.id]=\"listId\" [placement]=\"'bottom-start'\" [active]=\"hasAutocomplete\"\n                  (change)=\"onListSelect($event)\">\n\n   <div class=\"st-tag-input\">\n      <label *ngIf=\"hasLabel\" class=\"st-tag-input__label st-label\" [attr.id]=\"labelId\"\n             [attr.title]=\"tooltip\"\n             [attr.disabled]=\"disableValue\"\n             [ngClass]=\"{ 'active': hasFocus, 'error': showErrorValue }\">{{label}}</label>\n\n      <div #inputElement class=\"st-tag-input__input st-input\"\n           [attr.disabled]=\"disableValue\" (clickOutside)=\"onClickOutside($event)\"\n           [ngClass]=\"{ 'active': hasFocus, 'error': showErrorValue }\" (click)=\"onInputFocusIn($event)\">\n\n         <span *ngIf=\"hasPlaceholder\" class=\"st-tag-input__placeholder\" [attr.disabled]=\"disableValue\">{{placeholder}}</span>\n         <div class=\"st-tag-input__item tag-item\" *ngFor=\"let item of items; let i = index;\"\n              [tabindex]=\"0\"\n              [attr.id]=\"tagId + i\"\n              (focusin)=\"onTagFocusIn($event, i)\"\n              (focusout)=\"onTagFocusOut($event, i)\"\n              (blur)=\"onTagFocusOut($event, i)\"\n              (click)=\"onTagClick($event, i)\"\n              (keydown)=\"onTagKeyDown($event, i)\">{{item}}<span *ngIf=\"!disabled\" class=\"icon-cross remove-tag-button\" (click)=\"deleteTag(i); $event.stopPropagation()\"></span>\n         </div>\n\n         <div #newElement type=\"type\"\n              class=\"st-tag-input__text inner-input\"\n              [attr.contenteditable]=\"(disableValue === null && allowFreeText)\"\n              [attr.id]=\"inputId\"\n              [ngClass]=\"{ 'error': !isValidInput }\"\n              (focusin)=\"onInputFocusIn($event)\"\n              (focusout)=\"onInputFocusOut($event)\"\n              (blur)=\"onInputFocusOut($event)\"\n              (input)=\"onInputText($event.target.textContent)\"\n              (keydown)=\"onInputKeyDown($event)\"></div>\n      </div>\n   </div>\n\n</st-dropdown-menu>\n<span class=\"info-message\" *ngIf=\"infoMessage && isPristine\">{{infoMessage}}</span>\n<span *ngIf=\"showErrorValue\" class=\"st-input-error-message\">{{errorMessage}}</span>\n\n\n",
                    host: {
                        'class': 'st-tag-input'
                    },
                    providers: [
                        { provide: NG_VALUE_ACCESSOR, useExisting: forwardRef((/**
                             * @return {?}
                             */
                            function () { return StTagInputComponent; })), multi: true },
                        { provide: NG_VALIDATORS, useExisting: forwardRef((/**
                             * @return {?}
                             */
                            function () { return StTagInputComponent; })), multi: true }
                    ],
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    styles: ["@charset \"UTF-8\";:host{display:block;width:100%;position:relative}.inner-input{border:0;outline:0;display:inline-block;padding:5px 10px;vertical-align:top;margin-bottom:5px;overflow:hidden;max-width:100%;min-height:30px}.tag-item{display:inline-block;margin-right:5px;overflow:hidden;max-width:100%}.tag-item:focus{outline:0}.st-tag-input__placeholder{display:inline-block;padding-top:5px}"]
                }] }
    ];
    /** @nocollapse */
    StTagInputComponent.ctorParameters = function () { return [
        { type: ElementRef },
        { type: ChangeDetectorRef }
    ]; };
    StTagInputComponent.propDecorators = {
        label: [{ type: Input }],
        tooltip: [{ type: Input }],
        placeholder: [{ type: Input }],
        errorMessage: [{ type: Input }],
        type: [{ type: Input }],
        withAutocomplete: [{ type: Input }],
        autocompleteList: [{ type: Input }],
        charsToShowAutocompleteList: [{ type: Input }],
        allowFreeText: [{ type: Input }],
        infoMessage: [{ type: Input }],
        forbiddenValues: [{ type: Input }],
        regularExpression: [{ type: Input }],
        forceValidations: [{ type: Input }],
        newElementInput: [{ type: ViewChild, args: ['newElement', { static: true },] }],
        inputElement: [{ type: ViewChild, args: ['inputElement', { static: false },] }],
        disabled: [{ type: Input }],
        hasAutocomplete: [{ type: HostBinding, args: ['class.st-tag-input--autocomplete',] }]
    };
    return StTagInputComponent;
}());
export { StTagInputComponent };
if (false) {
    /**
     * \@input {string | null} [label=null] Label to show over the input. It is empty by default
     * @type {?}
     */
    StTagInputComponent.prototype.label;
    /**
     * \@input {string | null} [tooltip=null] The tooltip to show  over the label. It is empty by default
     * @type {?}
     */
    StTagInputComponent.prototype.tooltip;
    /**
     * \@input {string | null} [placeholder=null] The text that appears as placeholder of the input. It is empty by default
     * @type {?}
     */
    StTagInputComponent.prototype.placeholder;
    /**
     * \@input {string | null} [errorMessage=null] Error message to show. It is empty by default
     * @type {?}
     */
    StTagInputComponent.prototype.errorMessage;
    /**
     * \@input {string | null} [type=null] Type of the items
     * @type {?}
     */
    StTagInputComponent.prototype.type;
    /**
     * \@input {boolean} [withAutocomplete=false] Enable autocomplete feature. It is false by default
     * @type {?}
     */
    StTagInputComponent.prototype.withAutocomplete;
    /**
     * \@input {(StDropDownMenuItem | StDropDownMenuGroup)[]} [autocompleteList=Array()] List to be used for autocomplete feature. It is empty by default
     * @type {?}
     */
    StTagInputComponent.prototype.autocompleteList;
    /**
     * \@input {boolean} [charsToShowAutocompleteList=Array()] List to be used for autocomplete feature. It is empty by default
     * @type {?}
     */
    StTagInputComponent.prototype.charsToShowAutocompleteList;
    /**
     * \@input {boolean} [allowFreeText=true] Boolean to allow user to type a free text or not
     * @type {?}
     */
    StTagInputComponent.prototype.allowFreeText;
    /**
     * \@input {string} [infoMessage=] Message used to inform user about what values he has to introduce
     * @type {?}
     */
    StTagInputComponent.prototype.infoMessage;
    /**
     * \@input {string[]} [forbiddenValues=Array()] A list of values that user can not type and if he types one of them,
     * tag input will be invalid. It is empty by default
     * @type {?}
     */
    StTagInputComponent.prototype.forbiddenValues;
    /**
     * \@input {string} [regularExpression=] Regular expression to validate values. It is null by default
     * @type {?}
     */
    StTagInputComponent.prototype.regularExpression;
    /**
     * \@Input {boolean} [forceValidations=false] If you specify it to 'true', the tag input checks the errors before being modified by user
     * @type {?}
     */
    StTagInputComponent.prototype.forceValidations;
    /** @type {?} */
    StTagInputComponent.prototype.newElementInput;
    /** @type {?} */
    StTagInputComponent.prototype.inputElement;
    /** @type {?} */
    StTagInputComponent.prototype.expandedMenu;
    /** @type {?} */
    StTagInputComponent.prototype.items;
    /** @type {?} */
    StTagInputComponent.prototype.innerInputContent;
    /** @type {?} */
    StTagInputComponent.prototype.isPristine;
    /** @type {?} */
    StTagInputComponent.prototype.showErrorValue;
    /**
     * @type {?}
     * @private
     */
    StTagInputComponent.prototype._focus;
    /**
     * @type {?}
     * @private
     */
    StTagInputComponent.prototype._isDisabled;
    /**
     * @type {?}
     * @private
     */
    StTagInputComponent.prototype._newElementInput;
    /**
     * @type {?}
     * @private
     */
    StTagInputComponent.prototype._selected;
    /**
     * @type {?}
     * @private
     */
    StTagInputComponent.prototype._regularExp;
    /** @type {?} */
    StTagInputComponent.prototype.onChange;
    /** @type {?} */
    StTagInputComponent.prototype.onTouched;
    /**
     * @type {?}
     * @private
     */
    StTagInputComponent.prototype._selectElement;
    /**
     * @type {?}
     * @private
     */
    StTagInputComponent.prototype._cd;
}
