// @stratio/egeo/esm5/lib/utils/demo-generator/demo-generator.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/utils/demo-generator/demo-generator.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/*
 * © 2017 Stratio Big Data Inc., Sucursal en España.
 *
 * This software is licensed under the Apache License, Version 2.0.
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the terms of the License for more details.
 *
 * SPDX-License-Identifier: Apache-2.0.
 */
import { ChangeDetectionStrategy, Component, ComponentFactoryResolver, ViewChild, ViewContainerRef } from '@angular/core';
import { DemoGeneratorProviders } from './demo-generator.interface';
var StDemoGenerator = /** @class */ (function () {
    function StDemoGenerator(cfr, config) {
        this.cfr = cfr;
        this.config = config;
        this.inputs = config.inputs;
        this.outputs = config.outputs;
        this.component = config.components[0];
    }
    /** DYNAMIC BODY DEMO COMPONENT LOAD */
    /**
     * DYNAMIC BODY DEMO COMPONENT LOAD
     * @return {?}
     */
    StDemoGenerator.prototype.ngOnDestroy = /**
     * DYNAMIC BODY DEMO COMPONENT LOAD
     * @return {?}
     */
    function () {
        if (this.componentRef) {
            this.componentRef.destroy();
        }
    };
    /**
     * @return {?}
     */
    StDemoGenerator.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        if (this.component) {
            this.loadDemo();
        }
    };
    /**
     * @private
     * @return {?}
     */
    StDemoGenerator.prototype.loadDemo = /**
     * @private
     * @return {?}
     */
    function () {
        if (!this.componentRef) {
            this.target.clear();
            /** @type {?} */
            var compFactory = this.cfr.resolveComponentFactory(this.component);
            this.componentRef = this.target.createComponent(compFactory);
            this.bindDemoInputsAndOutputs();
        }
    };
    // tslint:disable:
    // tslint:disable:
    /**
     * @private
     * @return {?}
     */
    StDemoGenerator.prototype.bindDemoInputsAndOutputs = 
    // tslint:disable:
    /**
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        // Inputs its not necessary to be defined
        if (this.inputs) {
            Object.keys(this.inputs).map((/**
             * @param {?} key
             * @return {?}
             */
            function (key) { return _this.componentRef.instance[key] = ((/** @type {?} */ (_this.inputs)))[key]; }));
        }
        // Inputs must be defined for subscribe to them
        if (this.outputs) {
            Object.keys(this.outputs).map((/**
             * @param {?} key
             * @return {?}
             */
            function (key) {
                if (((/** @type {?} */ (_this.componentRef.instance))).hasOwnProperty(key)) {
                    _this.componentRef.instance[key].subscribe(((/** @type {?} */ (_this.outputs)))[key]);
                }
            }));
        }
        this.componentRef.changeDetectorRef.detectChanges();
    };
    StDemoGenerator.decorators = [
        { type: Component, args: [{
                    selector: 'st-demo-body',
                    template: '<div #stDemoBody></div>',
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    styles: ["\n      :host {\n         width: 100%;\n         display: block;\n      }\n   "]
                }] }
    ];
    /** @nocollapse */
    StDemoGenerator.ctorParameters = function () { return [
        { type: ComponentFactoryResolver },
        { type: DemoGeneratorProviders }
    ]; };
    StDemoGenerator.propDecorators = {
        target: [{ type: ViewChild, args: ['stDemoBody', { read: ViewContainerRef, static: false },] }]
    };
    return StDemoGenerator;
}());
export { StDemoGenerator };
if (false) {
    /** @type {?} */
    StDemoGenerator.prototype.target;
    /** @type {?} */
    StDemoGenerator.prototype.inputs;
    /** @type {?} */
    StDemoGenerator.prototype.outputs;
    /** @type {?} */
    StDemoGenerator.prototype.component;
    /**
     * @type {?}
     * @private
     */
    StDemoGenerator.prototype.componentRef;
    /**
     * @type {?}
     * @private
     */
    StDemoGenerator.prototype.cfr;
    /**
     * @type {?}
     * @private
     */
    StDemoGenerator.prototype.config;
}
