// @stratio/egeo/esm5/lib/utils/demo-generator/demo-generator.module.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/utils/demo-generator/demo-generator.module.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/*
 * © 2017 Stratio Big Data Inc., Sucursal en España.
 *
 * This software is licensed under the Apache License, Version 2.0.
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the terms of the License for more details.
 *
 * SPDX-License-Identifier: Apache-2.0.
 */
import { CommonModule } from '@angular/common';
import { ANALYZE_FOR_ENTRY_COMPONENTS, NgModule } from '@angular/core';
import { StDemoGenerator } from './demo-generator';
import { DemoGeneratorProviders } from './demo-generator.interface';
import { routing } from './demo-generator.routing';
var StDemoGeneratorModule = /** @class */ (function () {
    function StDemoGeneratorModule() {
    }
    /**
     * @param {?} config
     * @return {?}
     */
    StDemoGeneratorModule.withComponents = /**
     * @param {?} config
     * @return {?}
     */
    function (config) {
        return {
            ngModule: StDemoGeneratorModule,
            providers: [
                { provide: ANALYZE_FOR_ENTRY_COMPONENTS, useValue: config.components, multi: true },
                { provide: DemoGeneratorProviders, useValue: config }
            ]
        };
    };
    StDemoGeneratorModule.decorators = [
        { type: NgModule, args: [{
                    id: 'Demo-factory',
                    imports: [CommonModule, routing],
                    exports: [StDemoGenerator],
                    declarations: [StDemoGenerator],
                    entryComponents: [StDemoGenerator],
                    providers: []
                },] }
    ];
    return StDemoGeneratorModule;
}());
export { StDemoGeneratorModule };
