// @stratio/egeo/esm2015/lib/st-alerts/st-alerts.model.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/st-alerts/st-alerts.model.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/*
 * © 2017 Stratio Big Data Inc., Sucursal en España.
 *
 * This software is licensed under the Apache License, Version 2.0.
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the terms of the License for more details.
 *
 * SPDX-License-Identifier: Apache-2.0.
 */
import { Subject } from 'rxjs';
/** @enum {number} */
const STALERT_SEVERITY = {
    SUCCESS: 0, WARNING: 1, ERROR: 2, INFO: 3,
};
export { STALERT_SEVERITY };
STALERT_SEVERITY[STALERT_SEVERITY.SUCCESS] = 'SUCCESS';
STALERT_SEVERITY[STALERT_SEVERITY.WARNING] = 'WARNING';
STALERT_SEVERITY[STALERT_SEVERITY.ERROR] = 'ERROR';
STALERT_SEVERITY[STALERT_SEVERITY.INFO] = 'INFO';
export class StAlertLink {
}
if (false) {
    /** @type {?} */
    StAlertLink.prototype.title;
    /** @type {?} */
    StAlertLink.prototype.link;
}
export class StAlert {
    /**
     * @param {?} id
     * @param {?} title
     * @param {?} message
     * @param {?} severity
     * @param {?} timeout
     * @param {?} extendedTimeout
     * @param {?} link
     */
    constructor(id, title, message, severity, timeout, extendedTimeout, link) {
        this.id = id;
        this.title = title;
        this.message = message;
        this.severity = severity;
        this.timeout = timeout;
        this.extendedTimeout = extendedTimeout;
        this.link = link;
        this._readed = false;
        this._opacity = 0;
        this._opacityState = new Subject();
        this._removeEvent = new Subject();
    }
    /**
     * @return {?}
     */
    get opacity() {
        return this._opacityState.asObservable();
    }
    /**
     * @return {?}
     */
    get removeAlertEvent() {
        return this._removeEvent.asObservable();
    }
    /**
     * @return {?}
     */
    notify() {
        this.setVisible(true);
    }
    /**
     * @return {?}
     */
    pauseNotify() {
        this._opacity = 1;
        this._readed = true;
        this.clearAnimation();
        this.stopLife();
        this._opacityState.next(this._opacity);
    }
    /**
     * @return {?}
     */
    continueNotify() {
        this.startLife();
    }
    /**
     * @return {?}
     */
    cancel() {
        this.setVisible(false);
    }
    /**
     * @private
     * @param {?} increase
     * @return {?}
     */
    setVisible(increase) {
        this.clearAnimation();
        this._changeVisibilityInterval = window.setInterval((/**
         * @return {?}
         */
        () => this.modifyVisibility(increase)), 50);
    }
    /**
     * @private
     * @param {?} increase
     * @return {?}
     */
    modifyVisibility(increase) {
        this._opacity += increase ? 0.1 : -0.1;
        if (this._opacity >= 1 || this._opacity <= 0) {
            this.clearAnimation();
            if (increase) {
                this.startLife();
            }
            else {
                this.notifyForRemove();
            }
        }
        this._opacityState.next(this._opacity);
    }
    /**
     * @private
     * @return {?}
     */
    notifyForRemove() {
        this._opacityState.complete();
        this._removeEvent.next(this);
        this._removeEvent.complete();
    }
    /**
     * @private
     * @return {?}
     */
    clearAnimation() {
        window.clearInterval(this._changeVisibilityInterval);
    }
    /**
     * @private
     * @return {?}
     */
    stopLife() {
        window.clearTimeout(this._lifeTimeout);
    }
    /**
     * @private
     * @return {?}
     */
    startLife() {
        /** @type {?} */
        let timeout = this._readed ? this.extendedTimeout : this.timeout;
        this._lifeTimeout = window.setTimeout((/**
         * @return {?}
         */
        () => this.setVisible(false)), timeout);
    }
}
if (false) {
    /**
     * @type {?}
     * @private
     */
    StAlert.prototype._changeVisibilityInterval;
    /**
     * @type {?}
     * @private
     */
    StAlert.prototype._lifeTimeout;
    /**
     * @type {?}
     * @private
     */
    StAlert.prototype._readed;
    /**
     * @type {?}
     * @private
     */
    StAlert.prototype._opacity;
    /**
     * @type {?}
     * @private
     */
    StAlert.prototype._opacityState;
    /**
     * @type {?}
     * @private
     */
    StAlert.prototype._removeEvent;
    /** @type {?} */
    StAlert.prototype.id;
    /** @type {?} */
    StAlert.prototype.title;
    /** @type {?} */
    StAlert.prototype.message;
    /** @type {?} */
    StAlert.prototype.severity;
    /** @type {?} */
    StAlert.prototype.timeout;
    /** @type {?} */
    StAlert.prototype.extendedTimeout;
    /** @type {?} */
    StAlert.prototype.link;
}
