// @stratio/egeo/esm5/lib/st-form-list/st-form-list.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/st-form-list/st-form-list.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/*
 * © 2017 Stratio Big Data Inc., Sucursal en España.
 *
 * This software is licensed under the Apache License, Version 2.0.
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the terms of the License for more details.
 *
 * SPDX-License-Identifier: Apache-2.0.
 */
import { Component, Input, ChangeDetectionStrategy, EventEmitter, Output, forwardRef, ChangeDetectorRef } from '@angular/core';
import { NG_VALUE_ACCESSOR, FormControl, NG_VALIDATORS, FormGroup, FormArray } from '@angular/forms';
import { cloneDeep as _cloneDeep } from 'lodash';
/**
 * \@description {Component} [Form list]
 *
 * The form list component allows to create dynamically list of items.
 *
 * \@example
 *
 * {html}
 *
 * ```
 *
 * <st-form-list [schema]="jsonSchema" [ngModel]="model1" name="list"
 * (blur)="onBlur($event)" (add)="onAddItem($event)" (remove)="onRemoveItem($event)">
 * </st-form-list>
 *
 * ```
 *
 *
 */
var StFormListComponent = /** @class */ (function () {
    function StFormListComponent(_cd) {
        this._cd = _cd;
        /**
         * \@Input {string} [buttonLabel='Add one more item'] String displayed in the button to add more items
         */
        this.buttonLabel = 'Add one more item';
        /**
         * \@Output {any[]} [valueChange] Notify any value change
         */
        this.valueChange = new EventEmitter();
        /**
         * \@Output {number} [add]  Notify the position of the added item and the modified model
         */
        this.add = new EventEmitter();
        /**
         * \@Output {number} [remove] Notify the position of the removed item and the modified model
         */
        this.remove = new EventEmitter();
        /**
         * \@Output {any[]} [blur] Notify when user leaves a field
         */
        this.blur = new EventEmitter();
        this.formArray = new FormArray([]);
        this._value = [];
        this.onTouched = (/**
         * @return {?}
         */
        function () {
        });
    }
    Object.defineProperty(StFormListComponent.prototype, "value", {
        get: /**
         * @return {?}
         */
        function () {
            return this._value;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            if (value !== this._value) {
                this._value = _cloneDeep(value);
                this.updateForm();
                this.onChange(value);
            }
        },
        enumerable: true,
        configurable: true
    });
    // Function to call when the value changes.
    // Function to call when the value changes.
    /**
     * @param {?} _
     * @return {?}
     */
    StFormListComponent.prototype.onChange = 
    // Function to call when the value changes.
    /**
     * @param {?} _
     * @return {?}
     */
    function (_) {
    };
    /**
     * @return {?}
     */
    StFormListComponent.prototype.addItem = /**
     * @return {?}
     */
    function () {
        this._value.push({});
        this.formArray.push(this.generateItemFormGroup());
        this.valueChange.emit(this._value);
        this.add.emit({ position: this._value.length - 1, model: this._value });
        this.onChange(this._value);
    };
    /**
     * @param {?} position
     * @return {?}
     */
    StFormListComponent.prototype.removeItem = /**
     * @param {?} position
     * @return {?}
     */
    function (position) {
        this.formArray.removeAt(position);
        this._value.splice(position, 1);
        this.valueChange.emit(this._value);
        this.remove.emit({ position: position, model: this._value });
        this.onChange(this._value);
    };
    /**
     * @param {?} propertyName
     * @return {?}
     */
    StFormListComponent.prototype.isRequired = /**
     * @param {?} propertyName
     * @return {?}
     */
    function (propertyName) {
        return propertyName && this.schema.required && this.schema.required.indexOf(propertyName) !== -1;
    };
    /**
     * @param {?=} position
     * @return {?}
     */
    StFormListComponent.prototype.generateItemFormGroup = /**
     * @param {?=} position
     * @return {?}
     */
    function (position) {
        /** @type {?} */
        var formGroup = new FormGroup({});
        /** @type {?} */
        var properties = Object.keys(this.schema.properties);
        for (var i = 0; i < properties.length; ++i) {
            /** @type {?} */
            var property = properties[i];
            /** @type {?} */
            var value = this.schema.properties[property].default;
            if (position !== undefined && this._value[i]) {
                value = position !== undefined ? this._value[i][property] : null;
            }
            formGroup.addControl(properties[i], new FormControl(value));
        }
        return formGroup;
    };
    /**
     * @param {?} control
     * @return {?}
     */
    StFormListComponent.prototype.validate = /**
     * @param {?} control
     * @return {?}
     */
    function (control) {
        /** @type {?} */
        var errors = null;
        if (this.formArray) {
            var _loop_1 = function (i) {
                /** @type {?} */
                var rowFormGroup = (/** @type {?} */ (this_1.formArray.controls[i]));
                Object.keys(rowFormGroup.controls).forEach((/**
                 * @param {?} propertyName
                 * @return {?}
                 */
                function (propertyName) {
                    if (rowFormGroup.controls[propertyName] && rowFormGroup.controls[propertyName].errors) {
                        if (!errors) {
                            errors = [];
                        }
                        if (!errors[i]) {
                            errors[i] = {};
                        }
                        errors[i][propertyName] = rowFormGroup.controls[propertyName].errors;
                    }
                }));
            };
            var this_1 = this;
            for (var i = 0; i < this.formArray.controls.length; ++i) {
                _loop_1(i);
            }
        }
        return errors;
    };
    /**
     * @param {?} value
     * @param {?} rowPosition
     * @param {?} property
     * @return {?}
     */
    StFormListComponent.prototype.onChangeProperty = /**
     * @param {?} value
     * @param {?} rowPosition
     * @param {?} property
     * @return {?}
     */
    function (value, rowPosition, property) {
        var _this = this;
        this._value[rowPosition][property] = value;
        this.valueChange.emit(this._value);
        setTimeout((/**
         * @return {?}
         */
        function () {
            _this.onChange(_this._value);
        }));
    };
    /**
     * @return {?}
     */
    StFormListComponent.prototype.onBlur = /**
     * @return {?}
     */
    function () {
        this.blur.emit(this._value);
    };
    // When value is received from outside
    // When value is received from outside
    /**
     * @param {?} value
     * @return {?}
     */
    StFormListComponent.prototype.writeValue = 
    // When value is received from outside
    /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        if (value && this._value !== value) {
            this._value = _cloneDeep(value);
            this.updateForm();
            this._cd.markForCheck();
        }
    };
    // Registry the change function to propagate internal model changes
    // Registry the change function to propagate internal model changes
    /**
     * @param {?} fn
     * @return {?}
     */
    StFormListComponent.prototype.registerOnChange = 
    // Registry the change function to propagate internal model changes
    /**
     * @param {?} fn
     * @return {?}
     */
    function (fn) {
        this.onChange = fn;
    };
    /**
     * @param {?} fn
     * @return {?}
     */
    StFormListComponent.prototype.registerOnTouched = /**
     * @param {?} fn
     * @return {?}
     */
    function (fn) {
        this.onTouched = fn;
    };
    // Allows Angular to disable the form.
    // Allows Angular to disable the form.
    /**
     * @param {?} disable
     * @return {?}
     */
    StFormListComponent.prototype.setDisabledState = 
    // Allows Angular to disable the form.
    /**
     * @param {?} disable
     * @return {?}
     */
    function (disable) {
        if (disable) {
            this.formArray.disable();
        }
        else {
            this.formArray.enable();
        }
    };
    /**
     * @private
     * @return {?}
     */
    StFormListComponent.prototype.updateForm = /**
     * @private
     * @return {?}
     */
    function () {
        this.formArray = new FormArray([]);
        if (this._value) {
            for (var i = 0; i < this._value.length; ++i) {
                this.formArray.push(this.generateItemFormGroup(i));
            }
        }
    };
    StFormListComponent.decorators = [
        { type: Component, args: [{
                    selector: 'st-form-list',
                    template: "<form novalidate class=\"st-form-list\">\n   <div *ngFor=\"let row of value; let i=index\" class=\"st-form-list__row\">\n         <st-form-field class=\"st-form-list__row-item\"\n                        *ngFor=\"let property of schema.properties | stObjectToArray\"\n                        [schema]=\"property\"\n                        [formControl]=\"formArray.controls[i].get(property.key)\"\n                        [ngModel]=\"row[property.key]\"\n                        (ngModelChange)=\"onChangeProperty($event, i, property.key)\"\n                        name=\"{{property.key}}\"\n                        qaTag=\"{{property.key}}-{{i}}\"\n                        (blur)=\"onBlur()\"\n                        [required]=\"isRequired(property.key)\">\n         </st-form-field>\n         <i class=\"icon-cross remove-button\" (click)=\"removeItem(i)\"></i>\n   </div>\n\n   <button *ngIf=\"formArray && !formArray.disabled\" class=\"button button-link\" (click)=\"addItem()\">\n      <i class=\"icon-circle-plus\"></i><span>{{buttonLabel}}</span>\n   </button>\n</form>\n",
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    providers: [
                        { provide: NG_VALUE_ACCESSOR, useExisting: forwardRef((/**
                             * @return {?}
                             */
                            function () { return StFormListComponent; })), multi: true },
                        { provide: NG_VALIDATORS, useExisting: forwardRef((/**
                             * @return {?}
                             */
                            function () { return StFormListComponent; })), multi: true }
                    ],
                    styles: [".st-form-list{width:100%}.st-form-list__row{display:flex;position:relative}.st-form-list__row :first-child{margin-right:10px}.st-form-list__row-item{flex:1;margin-bottom:20px}.st-form-list__row:first-of-type .remove-button{margin-top:25px}.st-form-list__row:not(first-of-type) .remove-button{margin-top:6px}.button.button-link{margin-top:4px}.button.button-link>i{padding-right:4px}.remove-button{vertical-align:middle;padding-left:15px;cursor:pointer;position:absolute;right:-30px;top:0}"]
                }] }
    ];
    /** @nocollapse */
    StFormListComponent.ctorParameters = function () { return [
        { type: ChangeDetectorRef }
    ]; };
    StFormListComponent.propDecorators = {
        schema: [{ type: Input }],
        buttonLabel: [{ type: Input }],
        valueChange: [{ type: Output }],
        add: [{ type: Output }],
        remove: [{ type: Output }],
        blur: [{ type: Output }],
        value: [{ type: Input }]
    };
    return StFormListComponent;
}());
export { StFormListComponent };
if (false) {
    /**
     * \@Input {any} [schema=''] JSON schema of items
     * @type {?}
     */
    StFormListComponent.prototype.schema;
    /**
     * \@Input {string} [buttonLabel='Add one more item'] String displayed in the button to add more items
     * @type {?}
     */
    StFormListComponent.prototype.buttonLabel;
    /**
     * \@Output {any[]} [valueChange] Notify any value change
     * @type {?}
     */
    StFormListComponent.prototype.valueChange;
    /**
     * \@Output {number} [add]  Notify the position of the added item and the modified model
     * @type {?}
     */
    StFormListComponent.prototype.add;
    /**
     * \@Output {number} [remove] Notify the position of the removed item and the modified model
     * @type {?}
     */
    StFormListComponent.prototype.remove;
    /**
     * \@Output {any[]} [blur] Notify when user leaves a field
     * @type {?}
     */
    StFormListComponent.prototype.blur;
    /** @type {?} */
    StFormListComponent.prototype.formArray;
    /**
     * @type {?}
     * @private
     */
    StFormListComponent.prototype._value;
    /** @type {?} */
    StFormListComponent.prototype.onTouched;
    /**
     * @type {?}
     * @private
     */
    StFormListComponent.prototype._cd;
}
