// @stratio/egeo/esm5/lib/directives/st-input-adjustable/st-input-adjustable.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/directives/st-input-adjustable/st-input-adjustable.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/*
 * © 2017 Stratio Big Data Inc., Sucursal en España.
 *
 * This software is licensed under the Apache License, Version 2.0.
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the terms of the License for more details.
 *
 * SPDX-License-Identifier: Apache-2.0.
 */
import { Directive, ElementRef, Input } from '@angular/core';
import { NgModel } from '@angular/forms';
var StInputAdjustable = /** @class */ (function () {
    function StInputAdjustable(el, ngModel) {
        this.el = el;
        this.ngModel = ngModel;
    }
    /**
     * @return {?}
     */
    StInputAdjustable.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.createHiddenTitle();
        this.ngModel.valueChanges.subscribe((/**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            _this.text = value || _this.placeholder;
            _this.hiddenTitle.innerText = _this.text;
            setTimeout((/**
             * @return {?}
             */
            function () {
                _this.el.nativeElement.parentNode.style.width = _this.hiddenTitle.clientWidth + 10 + 'px';
            }));
        }));
    };
    /**
     * @private
     * @return {?}
     */
    StInputAdjustable.prototype.createHiddenTitle = /**
     * @private
     * @return {?}
     */
    function () {
        if (!this.ngModel.value) {
            this.el.nativeElement.value = null;
            this.el.nativeElement.placeholder = this.placeholder;
        }
        this.hiddenTitle = this.el.nativeElement.parentNode.querySelector('.hidden-text');
        if (!this.hiddenTitle) {
            this.hiddenTitle = document.createElement('span');
            this.el.nativeElement.parentNode.appendChild(this.hiddenTitle);
        }
        this.hiddenTitle.className = 'hidden-text';
        this.hiddenTitle.style.visibility = 'hidden';
        this.hiddenTitle.style.position = 'absolute';
        this.hiddenTitle.style.textTransform = 'none';
        this.hiddenTitle.style.whiteSpace = 'nowrap';
        this.hiddenTitle.style.width = 'auto';
        this.hiddenTitle.style.height = 'auto';
    };
    StInputAdjustable.decorators = [
        { type: Directive, args: [{
                    selector: '[st-input-adjustable]'
                },] }
    ];
    /** @nocollapse */
    StInputAdjustable.ctorParameters = function () { return [
        { type: ElementRef },
        { type: NgModel }
    ]; };
    StInputAdjustable.propDecorators = {
        placeholder: [{ type: Input }]
    };
    return StInputAdjustable;
}());
export { StInputAdjustable };
if (false) {
    /** @type {?} */
    StInputAdjustable.prototype.placeholder;
    /**
     * @type {?}
     * @private
     */
    StInputAdjustable.prototype.text;
    /**
     * @type {?}
     * @private
     */
    StInputAdjustable.prototype.hiddenTitle;
    /**
     * @type {?}
     * @private
     */
    StInputAdjustable.prototype.el;
    /**
     * @type {?}
     * @private
     */
    StInputAdjustable.prototype.ngModel;
}
