// @stratio/egeo/esm5/lib/st-filter-selector/st-filter-selector.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/st-filter-selector/st-filter-selector.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/*
 * © 2017 Stratio Big Data Inc., Sucursal en España. All rights reserved.
 *
 * This software – including all its source code – contains proprietary
 * information of Stratio Big Data Inc., Sucursal en España and
 * may not be revealed, sold, transferred, modified, distributed or
 * otherwise made available, licensed or sublicensed to third parties;
 * nor reverse engineered, disassembled or decompiled, without express
 * written authorization from Stratio Big Data Inc., Sucursal en España.
 */
import { ChangeDetectionStrategy, Component, EventEmitter, Input, Output } from '@angular/core';
import { StDropDownMenuItem, StDropDownVisualMode } from '../st-dropdown-menu/st-dropdown-menu.interface';
import { StPopPlacement } from '../st-pop/st-pop.model';
/**
 * \@description {Component} [Filter selector]
 *
 * The filter selector allows display a filter with a list of options
 *
 * \@example
 *
 * {html}
 *
 * ```
 * <st-filter-selector
 *    [filterList]="filter"
 *    [openFilter]="openFilter"
 *    [selected]="selectedFilter"
 *    (clickFilter)="onFilter($event)"
 *    (changeFilterVisibility)="onChangeFilterVisibility()"
 *    (closeFilter)="onCloseFilter()">
 * </st-filter-selector>
 * ```
 *
 */
var StFilterSelectorComponent = /** @class */ (function () {
    function StFilterSelectorComponent() {
        /**
         * \@Input {boolean} [openFilter=false] This boolean is needed to specify the status of the filter (open or closed)
         */
        this.openFilter = false;
        /**
         * \@Input {boolean} [menuAlignedToRight=false] Boolean to indicate if menu has to be displayed aligned to the right
         */
        this.menuAlignedToRight = false;
        /**
         * \@Output {StDropDownMenuItem} [clickFilter=] Event emitted when an option is selected
         */
        this.clickFilter = new EventEmitter();
        /**
         * \@Output {boolean} [changeFilterVisibility=] Event emitted when visibility of options changes
         */
        this.changeFilterVisibility = new EventEmitter();
        /**
         * \@Output {boolean} [closeFilter=] Event emitted when option menu is closed
         */
        this.closeFilter = new EventEmitter();
        /**
         * \@output {StDropDownMenuItem} [itemMouseEnter] Event emitted when mouse is over an item
         */
        this.itemMouseEnter = new EventEmitter();
        /**
         * \@output {StDropDownMenuItem} [itemMouseLeave] Event emitted when mouse leaves an item
         */
        this.itemMouseLeave = new EventEmitter();
        this.sectionMenuVisualMode = StDropDownVisualMode.MENU_LIST;
    }
    /**
     * @return {?}
     */
    StFilterSelectorComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.menuPlacement = this.menuAlignedToRight ? StPopPlacement.BOTTOM_END : StPopPlacement.BOTTOM_START;
        setTimeout((/**
         * @return {?}
         */
        function () {
            if (!_this.selected && _this.filterList && _this.filterList.length) {
                _this.clickFilter.emit(_this.filterList[0]);
            }
        }));
    };
    /**
     * @param {?} selectedFilter
     * @return {?}
     */
    StFilterSelectorComponent.prototype.onClickFilter = /**
     * @param {?} selectedFilter
     * @return {?}
     */
    function (selectedFilter) {
        this.clickFilter.emit(selectedFilter);
    };
    /**
     * @return {?}
     */
    StFilterSelectorComponent.prototype.onChangeFilterVisibility = /**
     * @return {?}
     */
    function () {
        this.changeFilterVisibility.emit(!this.openFilter);
    };
    /**
     * @return {?}
     */
    StFilterSelectorComponent.prototype.onCloseFilter = /**
     * @return {?}
     */
    function () {
        this.closeFilter.emit(true);
    };
    /**
     * @param {?} item
     * @return {?}
     */
    StFilterSelectorComponent.prototype.onItemMouseEnter = /**
     * @param {?} item
     * @return {?}
     */
    function (item) {
        this.itemMouseEnter.emit(item);
    };
    /**
     * @param {?} item
     * @return {?}
     */
    StFilterSelectorComponent.prototype.onItemMouseLeave = /**
     * @param {?} item
     * @return {?}
     */
    function (item) {
        this.itemMouseLeave.emit(item);
    };
    StFilterSelectorComponent.decorators = [
        { type: Component, args: [{
                    selector: 'st-filter-selector',
                    template: "<!--\n\n    \u00A9 2017 Stratio Big Data Inc., Sucursal en Espa\u00F1a. All rights reserved.\n\n    This software \u2013 including all its source code \u2013 contains proprietary\n    information of Stratio Big Data Inc., Sucursal en Espa\u00F1a and\n    may not be revealed, sold, transferred, modified, distributed or\n    otherwise made available, licensed or sublicensed to third parties;\n    nor reverse engineered, disassembled or decompiled, without express\n    written authorization from Stratio Big Data Inc., Sucursal en Espa\u00F1a.\n\n-->\n\n<st-dropdown-menu\n   class=\"st-filter-selector\"\n   [items]=\"filterList\"\n   [active]=\"openFilter\"\n   [placement]=\"menuPlacement\"\n   [openToLeft]=\"menuAlignedToRight\"\n   [visualMode]=\"sectionMenuVisualMode\"\n   (change)=\"onClickFilter($event)\"\n   (clickOutside)=\"onCloseFilter()\"\n   (itemMouseEnter)=\"onItemMouseEnter($event)\"\n   (itemMouseLeave)=\"onItemMouseLeave($event)\">\n   <div (click)=\"onChangeFilterVisibility()\" class=\"filter-title\">\n      <span>{{ selected?.label }}</span>\n      <i class=\"icon-arrow2_down arrow\"></i>\n   </div>\n</st-dropdown-menu>\n",
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    styles: ["@charset \"UTF-8\";.st-filter-selector{position:relative;display:inline-block;cursor:pointer;text-align:left}.st-filter-selector .filter-title{display:inline-flex;padding-bottom:10px}.st-filter-selector .filter-title>span{padding-right:10px}.st-filter-selector .filter-title>i{vertical-align:middle}"]
                }] }
    ];
    StFilterSelectorComponent.propDecorators = {
        openFilter: [{ type: Input }],
        filterList: [{ type: Input }],
        selected: [{ type: Input }],
        menuAlignedToRight: [{ type: Input }],
        clickFilter: [{ type: Output }],
        changeFilterVisibility: [{ type: Output }],
        closeFilter: [{ type: Output }],
        itemMouseEnter: [{ type: Output }],
        itemMouseLeave: [{ type: Output }]
    };
    return StFilterSelectorComponent;
}());
export { StFilterSelectorComponent };
if (false) {
    /**
     * \@Input {boolean} [openFilter=false] This boolean is needed to specify the status of the filter (open or closed)
     * @type {?}
     */
    StFilterSelectorComponent.prototype.openFilter;
    /**
     * \@Input {StDropDownMenuItem[]} [filterList=] List of options
     * @type {?}
     */
    StFilterSelectorComponent.prototype.filterList;
    /**
     * \@Input {StDropDownMenuItem} [selected=] Selected option
     * @type {?}
     */
    StFilterSelectorComponent.prototype.selected;
    /**
     * \@Input {boolean} [menuAlignedToRight=false] Boolean to indicate if menu has to be displayed aligned to the right
     * @type {?}
     */
    StFilterSelectorComponent.prototype.menuAlignedToRight;
    /**
     * \@Output {StDropDownMenuItem} [clickFilter=] Event emitted when an option is selected
     * @type {?}
     */
    StFilterSelectorComponent.prototype.clickFilter;
    /**
     * \@Output {boolean} [changeFilterVisibility=] Event emitted when visibility of options changes
     * @type {?}
     */
    StFilterSelectorComponent.prototype.changeFilterVisibility;
    /**
     * \@Output {boolean} [closeFilter=] Event emitted when option menu is closed
     * @type {?}
     */
    StFilterSelectorComponent.prototype.closeFilter;
    /**
     * \@output {StDropDownMenuItem} [itemMouseEnter] Event emitted when mouse is over an item
     * @type {?}
     */
    StFilterSelectorComponent.prototype.itemMouseEnter;
    /**
     * \@output {StDropDownMenuItem} [itemMouseLeave] Event emitted when mouse leaves an item
     * @type {?}
     */
    StFilterSelectorComponent.prototype.itemMouseLeave;
    /** @type {?} */
    StFilterSelectorComponent.prototype.sectionMenuVisualMode;
    /** @type {?} */
    StFilterSelectorComponent.prototype.menuPlacement;
}
