// @stratio/egeo/esm5/lib/utils/event-window-manager.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/utils/event-window-manager.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @abstract
 */
var /**
 * @abstract
 */
EventWindowManager = /** @class */ (function () {
    function EventWindowManager(_renderer, _cd) {
        this._renderer = _renderer;
        this._cd = _cd;
    }
    /**
     * @protected
     * @return {?}
     */
    EventWindowManager.prototype.openElement = /**
     * @protected
     * @return {?}
     */
    function () {
        var _this = this;
        if (this.isActive) {
            this.closeElement();
        }
        else {
            this.isActive = !this.isActive;
            setTimeout((/**
             * @return {?}
             */
            function () {
                _this.globalListener = _this._renderer.listen('document', 'click', _this.onClickOutside.bind(_this));
                _this.forceClose = _this._renderer.listen('document', 'stForceClose', _this.onForceClose.bind(_this));
            }));
        }
    };
    /**
     * @protected
     * @return {?}
     */
    EventWindowManager.prototype.closeElement = /**
     * @protected
     * @return {?}
     */
    function () {
        if (this.isActive) {
            this.isActive = !this.isActive;
            this._cd.markForCheck();
            if (this.globalListener !== undefined && typeof (this.globalListener) === 'function') {
                this.globalListener();
                this.forceClose();
            }
        }
    };
    /**
     * @private
     * @param {?} event
     * @return {?}
     */
    EventWindowManager.prototype.onClickOutside = /**
     * @private
     * @param {?} event
     * @return {?}
     */
    function (event) {
        if (!this._eventElementRef.nativeElement.contains(event.target)) {
            this.closeElement();
        }
    };
    /**
     * @private
     * @param {?} event
     * @return {?}
     */
    EventWindowManager.prototype.onForceClose = /**
     * @private
     * @param {?} event
     * @return {?}
     */
    function (event) {
        this.closeElement();
    };
    return EventWindowManager;
}());
/**
 * @abstract
 */
export { EventWindowManager };
if (false) {
    /** @type {?} */
    EventWindowManager.prototype.isActive;
    /** @type {?} */
    EventWindowManager.prototype._eventElementRef;
    /**
     * @type {?}
     * @private
     */
    EventWindowManager.prototype.globalListener;
    /**
     * @type {?}
     * @private
     */
    EventWindowManager.prototype.forceClose;
    /**
     * @type {?}
     * @private
     */
    EventWindowManager.prototype._renderer;
    /**
     * @type {?}
     * @private
     */
    EventWindowManager.prototype._cd;
    /**
     * @abstract
     * @return {?}
     */
    EventWindowManager.prototype.ngOnDestroy = function () { };
}
