// @stratio/egeo/esm2015/lib/st-select/st-check-validations.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/st-select/st-check-validations.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/*
 * © 2017 Stratio Big Data Inc., Sucursal en España.
 *
 * This software is licensed under the Apache License, Version 2.0.
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the terms of the License for more details.
 *
 * SPDX-License-Identifier: Apache-2.0.
 */
import { Directive, Input } from '@angular/core';
import { NgControl } from '@angular/forms';
export class StCheckValidationsDirective {
    /**
     * @param {?} control
     */
    constructor(control) {
        this.control = control;
    }
    /**
     * @param {?} fn
     * @return {?}
     */
    registerOnChange(fn) {
        this.onChange = fn;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.sub = this.control.statusChanges.subscribe((/**
         * @return {?}
         */
        () => this.notifyErrorsIfExists()));
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        if (this.sub) {
            this.sub.unsubscribe();
        }
    }
    // TODO implement error check and notify correct error
    /**
     * @private
     * @return {?}
     */
    notifyErrorsIfExists() {
        if (this.control.valid && this.onChange) {
            this.onChange(null);
        }
        else if (this.onChange) {
            this.onChange(this.errorMessage);
        }
    }
}
StCheckValidationsDirective.decorators = [
    { type: Directive, args: [{ selector: '[stCheckValidations]' },] }
];
/** @nocollapse */
StCheckValidationsDirective.ctorParameters = () => [
    { type: NgControl }
];
StCheckValidationsDirective.propDecorators = {
    errorMessage: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    StCheckValidationsDirective.prototype.errorMessage;
    /** @type {?} */
    StCheckValidationsDirective.prototype.onChange;
    /**
     * @type {?}
     * @private
     */
    StCheckValidationsDirective.prototype.sub;
    /**
     * @type {?}
     * @private
     */
    StCheckValidationsDirective.prototype.control;
}
