// @stratio/egeo/esm5/lib/st-table/shared/st-table-row/st-table-row.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/st-table/shared/st-table-row/st-table-row.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/*
 * © 2017 Stratio Big Data Inc., Sucursal en España.
 *
 * This software is licensed under the Apache License, Version 2.0.
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the terms of the License for more details.
 *
 * SPDX-License-Identifier: Apache-2.0.
 */
import { ChangeDetectionStrategy, Component, HostListener, Input, ViewChild, HostBinding, ElementRef } from '@angular/core';
/**
 * \@description {Component} [Table Row]
 *
 * This component is designed to be added to the table component
 *
 * \@example
 *
 * {html}
 *
 * ```
 *  <tr st-table-row [selected] ="true" [standUpSelected] = "false">
 *      <!-- CELL LIST WILL BE HERE -->
 *  </tr>
 * ```
 *
 */
var StTableRowComponent = /** @class */ (function () {
    function StTableRowComponent(elementRef) {
        this.elementRef = elementRef;
        /**
         * \@Input {boolean} [standUpSelected=true] It indicates if when row is selected, it has to be displayed stood up
         */
        this.standUpSelected = true;
        this.rootClass = true;
        this.showHoverMenu = false;
    }
    Object.defineProperty(StTableRowComponent.prototype, "hasSelectedClass", {
        get: /**
         * @return {?}
         */
        function () {
            return this.selected && this.standUpSelected;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    StTableRowComponent.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        if (this.hoverMenu && this.hoverMenu.nativeElement.children.length === 0) {
            this.elementRef.nativeElement.removeChild(this.hoverMenu.nativeElement);
        }
    };
    /**
     * @return {?}
     */
    StTableRowComponent.prototype.onShowHoverMenu = /**
     * @return {?}
     */
    function () {
        this.showHoverMenu = true;
    };
    /**
     * @return {?}
     */
    StTableRowComponent.prototype.onHideHoverMenu = /**
     * @return {?}
     */
    function () {
        this.showHoverMenu = false;
    };
    StTableRowComponent.decorators = [
        { type: Component, args: [{
                    selector: '[st-table-row]',
                    template: "<!--\n\n    \u00A9 2017 Stratio Big Data Inc., Sucursal en Espa\u00F1a.\n\n    This software is licensed under the Apache License, Version 2.0.\n    This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;\n    without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.\n    See the terms of the License for more details.\n\n    SPDX-License-Identifier: Apache-2.0.\n\n-->\n<ng-content\n   select='[st-table-row-content]'>\n</ng-content>\n\n<td class='hover-menu' #hoverMenu [ngClass]=\"{'hover-menu--show': showHoverMenu}\">\n   <ng-content select='[st-table-row-hover]'></ng-content>\n</td>\n",
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    styles: ["@charset \"UTF-8\";:host{position:relative;width:100%}.hover-menu{opacity:0;vertical-align:top;position:relative}.hover-menu--show{opacity:1!important}"]
                }] }
    ];
    /** @nocollapse */
    StTableRowComponent.ctorParameters = function () { return [
        { type: ElementRef }
    ]; };
    StTableRowComponent.propDecorators = {
        selected: [{ type: Input }],
        standUpSelected: [{ type: Input }],
        hoverMenu: [{ type: ViewChild, args: ['hoverMenu', { static: false },] }],
        rootClass: [{ type: HostBinding, args: ['class.st-table-row',] }],
        hasSelectedClass: [{ type: HostBinding, args: ['class.selected',] }],
        onShowHoverMenu: [{ type: HostListener, args: ['mouseover',] }],
        onHideHoverMenu: [{ type: HostListener, args: ['mouseout',] }]
    };
    return StTableRowComponent;
}());
export { StTableRowComponent };
if (false) {
    /**
     * \@Input {boolean} [selected=''] It indicates if row is selected or not
     * @type {?}
     */
    StTableRowComponent.prototype.selected;
    /**
     * \@Input {boolean} [standUpSelected=true] It indicates if when row is selected, it has to be displayed stood up
     * @type {?}
     */
    StTableRowComponent.prototype.standUpSelected;
    /** @type {?} */
    StTableRowComponent.prototype.hoverMenu;
    /** @type {?} */
    StTableRowComponent.prototype.rootClass;
    /** @type {?} */
    StTableRowComponent.prototype.showHoverMenu;
    /**
     * @type {?}
     * @private
     */
    StTableRowComponent.prototype.elementRef;
}
