// @stratio/egeo/esm2015/lib/st-dynamic-table/utils/st-dynamic-table.utils.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/st-dynamic-table/utils/st-dynamic-table.utils.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/*
 * © 2017 Stratio Big Data Inc., Sucursal en España.
 *
 * This software is licensed under the Apache License, Version 2.0.
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the terms of the License for more details.
 *
 * SPDX-License-Identifier: Apache-2.0.
 */
export class StDynamicTableUtils {
    /**
     * @param {?} jsonSchema
     * @param {?=} uiDefinitions
     * @return {?}
     */
    static getHeaderFieldsFromJsonSchema(jsonSchema, uiDefinitions) {
        /** @type {?} */
        const fields = [];
        if (jsonSchema && jsonSchema.properties) {
            /** @type {?} */
            const properties = Object.keys(jsonSchema.properties);
            if (properties) {
                properties.forEach((/**
                 * @param {?} _propertyKey
                 * @return {?}
                 */
                _propertyKey => {
                    /** @type {?} */
                    const _property = jsonSchema.properties[_propertyKey];
                    if (_property && _property.type !== 'object') {
                        /** @type {?} */
                        const uiDefinition = uiDefinitions && uiDefinitions[_propertyKey];
                        /** @type {?} */
                        const filters = StDynamicTableUtils.getHeaderFiltersFromJsonSchema(_property, uiDefinition);
                        /** @type {?} */
                        const isSortable = uiDefinition && uiDefinitions[_propertyKey].sortable;
                        fields.push({
                            id: _propertyKey,
                            label: _property.title || _propertyKey,
                            reference: _property.$ref,
                            fk: uiDefinition && uiDefinition.fk,
                            group: uiDefinition && uiDefinition.group_field ? uiDefinition.group_field.name : null,
                            filters: filters,
                            filterable: filters && (filters.templateRef !== undefined || (filters.filterConfig && filters.filterConfig.length > 0)),
                            sortable: isSortable
                        });
                    }
                }));
            }
        }
        return fields;
    }
    /**
     * @param {?} propertyDefinition
     * @param {?} uiDefinition
     * @return {?}
     */
    static getHeaderFiltersFromJsonSchema(propertyDefinition, uiDefinition) {
        /** @type {?} */
        let filters;
        if (propertyDefinition.enum && propertyDefinition.enum.length) {
            filters = {
                title: propertyDefinition.title,
                filterConfig: []
            };
            propertyDefinition.enum.forEach((/**
             * @param {?} _value
             * @return {?}
             */
            _value => {
                filters.filterConfig.push({
                    id: (/** @type {?} */ (_value)),
                    name: (/** @type {?} */ (_value))
                });
            }));
        }
        else if (uiDefinition && uiDefinition.templateRef) {
            filters = {
                title: propertyDefinition.title,
                templateRef: uiDefinition.templateRef
            };
        }
        return filters;
    }
}
