// @stratio/egeo/esm5/lib/st-alerts/alert-box/st-alert-box.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/st-alerts/alert-box/st-alert-box.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/*
 * © 2017 Stratio Big Data Inc., Sucursal en España.
 *
 * This software is licensed under the Apache License, Version 2.0.
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the terms of the License for more details.
 *
 * SPDX-License-Identifier: Apache-2.0.
 */
import { Component, Input, ChangeDetectorRef, ChangeDetectionStrategy, Output, EventEmitter } from '@angular/core';
import { STALERT_SEVERITY, StAlert } from '../st-alerts.model';
var StAlertBoxComponent = /** @class */ (function () {
    function StAlertBoxComponent(_cd) {
        this._cd = _cd;
        /**
         * \@Input {boolean} [showInConsole=false] Display logs in console
         */
        this.showInConsole = false;
        /**
         * \@Output {StAlertLink} [clickLink=] Event emitted when link is clicked
         */
        this.clickLink = new EventEmitter();
        /**
         * \@Output {boolean} [close=] Event emitted when alert is closed
         */
        this.close = new EventEmitter();
        this.opacity = 0;
    }
    /**
     * @return {?}
     */
    StAlertBoxComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.alert.opacity.subscribe((/**
         * @param {?} opacity
         * @return {?}
         */
        function (opacity) { return _this.changeOpacity(opacity); }));
        this.alert.notify();
        if (this.showInConsole) {
            this.notifyConsole();
        }
        this.severityColorValue = this.getSeverityColor();
        this.iconValue = this.getIcon();
    };
    /**
     * @return {?}
     */
    StAlertBoxComponent.prototype.closeAlert = /**
     * @return {?}
     */
    function () {
        this.alert.cancel();
        this.close.emit(true);
    };
    /**
     * @return {?}
     */
    StAlertBoxComponent.prototype.getIcon = /**
     * @return {?}
     */
    function () {
        switch (this.alert.severity) {
            case STALERT_SEVERITY.ERROR: return 'icon-circle-alert';
            case STALERT_SEVERITY.WARNING: return 'icon-alert';
            case STALERT_SEVERITY.SUCCESS: return 'icon-circle-check';
            case STALERT_SEVERITY.INFO: return 'icon-info';
            default: return '';
        }
    };
    /**
     * @return {?}
     */
    StAlertBoxComponent.prototype.getSeverityColor = /**
     * @return {?}
     */
    function () {
        switch (this.alert.severity) {
            case STALERT_SEVERITY.ERROR: return 'sth-alert-box-error';
            case STALERT_SEVERITY.WARNING: return 'sth-alert-box-warning';
            case STALERT_SEVERITY.SUCCESS: return 'sth-alert-box-success';
            case STALERT_SEVERITY.INFO: return 'sth-alert-box-info';
            default: return '';
        }
    };
    /**
     * @return {?}
     */
    StAlertBoxComponent.prototype.goTo = /**
     * @return {?}
     */
    function () {
        this.clickLink.emit(this.alert.link);
    };
    /**
     * @param {?} opacity
     * @return {?}
     */
    StAlertBoxComponent.prototype.changeOpacity = /**
     * @param {?} opacity
     * @return {?}
     */
    function (opacity) {
        this.opacity = opacity;
        this._cd.markForCheck();
    };
    /**
     * @private
     * @return {?}
     */
    StAlertBoxComponent.prototype.notifyConsole = /**
     * @private
     * @return {?}
     */
    function () {
        switch (this.alert.severity) {
            case STALERT_SEVERITY.ERROR:
                console.error("ERROR-" + this.alert.title + ": " + this.alert.message);
                break;
            case STALERT_SEVERITY.WARNING:
                console.warn("WARNING-" + this.alert.title + ": " + this.alert.message);
                break;
            case STALERT_SEVERITY.SUCCESS:
                console.log("SUCCESS-" + this.alert.title + ": " + this.alert.message);
                break;
            default:
                console.error("ERROR: severity not found for " + this.alert.title + ": " + this.alert.message);
                break;
        }
    };
    StAlertBoxComponent.decorators = [
        { type: Component, args: [{
                    selector: 'st-alert-box',
                    template: "<!--\n\n    \u00A9 2017 Stratio Big Data Inc., Sucursal en Espa\u00F1a.\n\n    This software is licensed under the Apache License, Version 2.0.\n    This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;\n    without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.\n    See the terms of the License for more details.\n\n    SPDX-License-Identifier: Apache-2.0.\n\n-->\n<div class=\"sth-alert-box\" [ngClass]=\"severityColorValue\" [ngStyle]=\"{opacity: opacity}\">\n   <div class=\"sth-alert-box-content\">\n      <div class=\"sth-alert-box-close-button\">\n         <i (click)=\"closeAlert()\" class=\"icon-cross\"></i>\n      </div>\n      <header class=\"sth-alert-box-header\">\n         <i class=\"status-icon\" [ngClass]=\"iconValue\"></i>\n         <span>{{alert.title}}</span>\n      </header>\n      <div class=\"sth-alert-box-body\">\n         <p>{{alert.message}}</p>\n      </div>\n   </div>\n   <footer *ngIf=\"alert.link\" class=\"sth-alert-box-footer\">\n      <a (click)=\"goTo()\">{{alert.link.title}}</a>\n   </footer>\n</div>\n",
                    changeDetection: ChangeDetectionStrategy.OnPush
                }] }
    ];
    /** @nocollapse */
    StAlertBoxComponent.ctorParameters = function () { return [
        { type: ChangeDetectorRef }
    ]; };
    StAlertBoxComponent.propDecorators = {
        alert: [{ type: Input }],
        showInConsole: [{ type: Input }],
        clickLink: [{ type: Output }],
        close: [{ type: Output }]
    };
    return StAlertBoxComponent;
}());
export { StAlertBoxComponent };
if (false) {
    /**
     * \@Input {StAlert} [alert=] Data of the alert
     * @type {?}
     */
    StAlertBoxComponent.prototype.alert;
    /**
     * \@Input {boolean} [showInConsole=false] Display logs in console
     * @type {?}
     */
    StAlertBoxComponent.prototype.showInConsole;
    /**
     * \@Output {StAlertLink} [clickLink=] Event emitted when link is clicked
     * @type {?}
     */
    StAlertBoxComponent.prototype.clickLink;
    /**
     * \@Output {boolean} [close=] Event emitted when alert is closed
     * @type {?}
     */
    StAlertBoxComponent.prototype.close;
    /** @type {?} */
    StAlertBoxComponent.prototype.iconValue;
    /** @type {?} */
    StAlertBoxComponent.prototype.opacity;
    /** @type {?} */
    StAlertBoxComponent.prototype.severityColorValue;
    /**
     * @type {?}
     * @private
     */
    StAlertBoxComponent.prototype._cd;
}
