// @stratio/egeo/esm5/lib/st-item-list/item-list-item/item-list-item.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/st-item-list/item-list-item/item-list-item.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
/*
 * © 2017 Stratio Big Data Inc., Sucursal en España.
 *
 * This software is licensed under the Apache License, Version 2.0.
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the terms of the License for more details.
 *
 * SPDX-License-Identifier: Apache-2.0.
 */
import { ChangeDetectionStrategy, Component, EventEmitter, Input, Output } from '@angular/core';
import { StEgeo, StRequired } from '../../decorators/require-decorators';
var ItemListItemComponent = /** @class */ (function () {
    function ItemListItemComponent() {
        this.qaTag = '';
        this.align = 'left';
        this.selectItem = new EventEmitter();
    }
    Object.defineProperty(ItemListItemComponent.prototype, "itemName", {
        get: /**
         * @return {?}
         */
        function () {
            return this.item.name;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ItemListItemComponent.prototype, "itemIcon", {
        get: /**
         * @return {?}
         */
        function () {
            return this.item.icon;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ItemListItemComponent.prototype, "itemSelected", {
        get: /**
         * @return {?}
         */
        function () {
            return this.item.selected ? true : false;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ItemListItemComponent.prototype, "itemQaTag", {
        get: /**
         * @return {?}
         */
        function () {
            return this.qaTag && this.qaTag + '-item-' + this.item.id || '';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ItemListItemComponent.prototype, "itemStyle", {
        get: /**
         * @return {?}
         */
        function () {
            /** @type {?} */
            var style = "sth-item-list__item item-list-item item-list-item--" + this.align;
            style += this.itemSelected ? ' sth-item-list__item--selected' : '';
            return style;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} event
     * @return {?}
     */
    ItemListItemComponent.prototype.emitOnSelect = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        event.preventDefault();
        event.stopImmediatePropagation();
        this.selectItem.emit(this.item);
    };
    ItemListItemComponent.decorators = [
        { type: Component, args: [{
                    selector: 'item-list-item',
                    template: "<!--\n\n    \u00A9 2017 Stratio Big Data Inc., Sucursal en Espa\u00F1a.\n\n    This software is licensed under the Apache License, Version 2.0.\n    This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;\n    without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.\n    See the terms of the License for more details.\n\n    SPDX-License-Identifier: Apache-2.0.\n\n-->\n<div [ngClass]=\"itemStyle\" [attr.id]=\"itemQaTag\" [title]=\"itemName\" (click)=\"emitOnSelect($event)\">\n   <i *ngIf=\"align === 'left'\" class=\"item-list-item__icon item-list-item__icon--left {{itemIcon}}\"></i>\n   <span class=\"item-list-item__name\">{{itemName}}</span>\n   <i *ngIf=\"align === 'right'\" class=\"item-list-item__icon item-list-item__icon--right {{itemIcon}}\"></i>\n</div>\n",
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    styles: ["@charset \"UTF-8\";.item-list-item{display:inline-block;min-width:100%;width:100%;height:100%;padding:5px 0;overflow:hidden;white-space:nowrap;text-overflow:ellipsis}.item-list-item--left{text-align:left}.item-list-item--right{text-align:right}.item-list-item__icon{position:relative;top:2px}.item-list-item__icon--left{margin-right:10px}.item-list-item__icon--right{margin-left:10px}.item-list-item__name{position:relative;top:1px}:host{display:block;padding-right:20px}"]
                }] }
    ];
    /** @nocollapse */
    ItemListItemComponent.ctorParameters = function () { return []; };
    ItemListItemComponent.propDecorators = {
        item: [{ type: Input }],
        qaTag: [{ type: Input }],
        align: [{ type: Input }],
        selectItem: [{ type: Output }]
    };
    tslib_1.__decorate([
        StRequired(),
        tslib_1.__metadata("design:type", Object)
    ], ItemListItemComponent.prototype, "item", void 0);
    ItemListItemComponent = tslib_1.__decorate([
        StEgeo(),
        tslib_1.__metadata("design:paramtypes", [])
    ], ItemListItemComponent);
    return ItemListItemComponent;
}());
export { ItemListItemComponent };
if (false) {
    /** @type {?} */
    ItemListItemComponent.prototype.item;
    /** @type {?} */
    ItemListItemComponent.prototype.qaTag;
    /** @type {?} */
    ItemListItemComponent.prototype.align;
    /** @type {?} */
    ItemListItemComponent.prototype.selectItem;
}
