// @stratio/egeo/esm2015/lib/st-breadcrumbs/st-breadcrumbs.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/st-breadcrumbs/st-breadcrumbs.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/*
 * © 2017 Stratio Big Data Inc., Sucursal en España.
 *
 * This software is licensed under the Apache License, Version 2.0.
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the terms of the License for more details.
 *
 * SPDX-License-Identifier: Apache-2.0.
 */
import { ChangeDetectionStrategy, Component, EventEmitter, Input, Output } from '@angular/core';
import { range as _range } from 'lodash';
import { StBreadCrumbMode } from './st-breadcrumbs.interface';
/**
 * \@description {Component} [Breadcrumbs]
 *
 * The breadcrumb is s a type of secondary navigation
 * element that shows a trail for allowing users to keep track of their location.
 *
 * \@model
 *
 *   [Id, optional label and icon of an option] {./st-breadcrumbs.interface.ts#StBreadCrumbItem}
 *
 * \@example
 *
 * {html}
 *
 * ```
 * <st-breadcrumbs
 *    [options]="options"
 *    (select)="onSelect($event)"
 *    qaTag="breadcrumbs">
 * </st-breadcrumbs>
 * ```
 *
 */
export class StBreadCrumbsComponent {
    constructor() {
        /**
         * \@Input {StBreadCrumbItem[]} [options=\[\]] List of navigation parts for show in breadcrumb
         */
        this.options = [];
        /**
         * \@Input {number} [elementsToShow=5] Max number of elements to show.
         */
        this.elementsToShow = 5;
        /**
         * \@Input {StBreadCrumbMode} [mode=StBreadCrumbMode.DEFAULT] Visualization mode
         */
        this.mode = StBreadCrumbMode.DEFAULT;
        /**
         * \@Output {number} [select] Notify any item selection
         */
        this.select = new EventEmitter();
        this.indexArray = [];
        this._initPos = 0;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.calculateStart(this.options, this.elementsToShow);
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    ngOnChanges(changes) {
        if (changes.options || changes.elementsToShow) {
            /** @type {?} */
            const currElsToShow = changes.elementsToShow ? changes.elementsToShow.currentValue : this.elementsToShow;
            /** @type {?} */
            const currOptions = changes.options ? changes.options.currentValue : this.options;
            this.calculateStart(currOptions, currElsToShow);
        }
    }
    /**
     * @param {?} index
     * @return {?}
     */
    onSelect(index) {
        /** @type {?} */
        const toEmit = index > -1 ? index : this._initPos - 1;
        if (toEmit < (this.options.length - 1)) {
            this.select.emit(toEmit);
        }
    }
    /**
     * @param {?} index
     * @return {?}
     */
    getId(index) {
        return this.options[index].id;
    }
    /**
     * @param {?} index
     * @return {?}
     */
    getLabel(index) {
        return index > -1 ? this.options[index].label : '...';
    }
    /**
     * @param {?} index
     * @return {?}
     */
    getIcon(index) {
        return index > -1 ? this.options[index].icon : '';
    }
    /**
     * @param {?} index
     * @return {?}
     */
    hasIcon(index) {
        return index > -1 && Boolean(this.options[index].icon);
    }
    /**
     * @private
     * @param {?} options
     * @param {?} elementsToShow
     * @return {?}
     */
    calculateStart(options, elementsToShow) {
        if (options.length > elementsToShow) {
            this._initPos = options.length - elementsToShow + 1; // use +1 because first element is fixed
            this.indexArray = _range(this._initPos, this._initPos + elementsToShow - 1);
            this.indexArray.unshift(-1); // Add dots fixed element
            this.indexArray.unshift(0); // Add first fixed element
        }
        else {
            this._initPos = 0;
            this.indexArray = _range(0, options.length);
        }
    }
}
StBreadCrumbsComponent.decorators = [
    { type: Component, args: [{
                selector: 'st-breadcrumbs',
                template: "<!--\n\n    \u00A9 2017 Stratio Big Data Inc., Sucursal en Espa\u00F1a.\n\n    This software is licensed under the Apache License, Version 2.0.\n    This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;\n    without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.\n    See the terms of the License for more details.\n\n    SPDX-License-Identifier: Apache-2.0.\n\n-->\n<ul class=\"st-breadcrumbs {{mode}}-mode\" [attr.id]=\"qaTag\">\n   <ng-content select=\"st-breadcrumbs-item\"></ng-content>\n   <ng-container *ngIf=\"options.length\">\n      <st-breadcrumbs-item\n         class=\"sth-breadcrumbs__item-container\"\n         *ngFor=\"let idx of indexArray; let last=last\"\n         (click)=\"onSelect(idx)\"\n         [active]=\"last\"\n         [qaTag]=\"qaTag + '-' + idx\">\n         <i *ngIf=\"hasIcon(idx)\" [ngClass]=\"getIcon(idx)\"></i>\n         <span>{{getLabel(idx)}}</span>\n      </st-breadcrumbs-item>\n   </ng-container>\n</ul>\n",
                changeDetection: ChangeDetectionStrategy.OnPush,
                styles: ["@charset \"UTF-8\";ul{text-overflow:ellipsis;overflow:hidden;list-style:none;margin:0;padding:0}"]
            }] }
];
StBreadCrumbsComponent.propDecorators = {
    options: [{ type: Input }],
    elementsToShow: [{ type: Input }],
    mode: [{ type: Input }],
    qaTag: [{ type: Input }],
    select: [{ type: Output }]
};
if (false) {
    /**
     * \@Input {StBreadCrumbItem[]} [options=\[\]] List of navigation parts for show in breadcrumb
     * @type {?}
     */
    StBreadCrumbsComponent.prototype.options;
    /**
     * \@Input {number} [elementsToShow=5] Max number of elements to show.
     * @type {?}
     */
    StBreadCrumbsComponent.prototype.elementsToShow;
    /**
     * \@Input {StBreadCrumbMode} [mode=StBreadCrumbMode.DEFAULT] Visualization mode
     * @type {?}
     */
    StBreadCrumbsComponent.prototype.mode;
    /**
     * \@Input {string} [qaTag=''] Id value for qa test
     * @type {?}
     */
    StBreadCrumbsComponent.prototype.qaTag;
    /**
     * \@Output {number} [select] Notify any item selection
     * @type {?}
     */
    StBreadCrumbsComponent.prototype.select;
    /** @type {?} */
    StBreadCrumbsComponent.prototype.indexArray;
    /**
     * @type {?}
     * @private
     */
    StBreadCrumbsComponent.prototype._initPos;
}
