// @stratio/egeo/esm5/lib/st-toggle-buttons/st-toggle-buttons.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/st-toggle-buttons/st-toggle-buttons.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
/*
 * © 2017 Stratio Big Data Inc., Sucursal en España.
 *
 * This software is licensed under the Apache License, Version 2.0.
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the terms of the License for more details.
 *
 * SPDX-License-Identifier: Apache-2.0.
 */
import { ChangeDetectionStrategy, Component, EventEmitter, Input, Output } from '@angular/core';
var StToggleButtonsComponent = /** @class */ (function () {
    function StToggleButtonsComponent() {
        this.select = new EventEmitter();
    }
    Object.defineProperty(StToggleButtonsComponent.prototype, "tabs", {
        set: /**
         * @param {?} tabs
         * @return {?}
         */
        function (tabs) {
            this.parsedTabs = tabs
                .map((/**
             * @param {?} __0
             * @return {?}
             */
            function (_a) {
                var label = _a.label, props = tslib_1.__rest(_a, ["label"]);
                return (tslib_1.__assign({}, props, { label: label, trimLabel: label.replace(/[^a-zA-Z0-9]/g, '') }));
            }));
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} selectedTab
     * @return {?}
     */
    StToggleButtonsComponent.prototype.onClick = /**
     * @param {?} selectedTab
     * @return {?}
     */
    function (selectedTab) {
        var e_1, _a;
        try {
            for (var _b = tslib_1.__values(this.parsedTabs), _c = _b.next(); !_c.done; _c = _b.next()) {
                var tab = _c.value;
                tab.active = false;
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_1) throw e_1.error; }
        }
        selectedTab.active = true;
        this.select.emit(selectedTab);
    };
    StToggleButtonsComponent.decorators = [
        { type: Component, args: [{
                    selector: 'st-toggle-buttons',
                    template: "<!--\n\n    \u00A9 2017 Stratio Big Data Inc., Sucursal en Espa\u00F1a.\n\n    This software is licensed under the Apache License, Version 2.0.\n    This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;\n    without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.\n    See the terms of the License for more details.\n\n    SPDX-License-Identifier: Apache-2.0.\n\n-->\n<section class=\"st-toggle-buttons\">\n   <div *ngIf=\"description\" class=\"st-toggle-buttons__description sth-toggle-buttons__description\">\n      {{description}}\n   </div>\n   <ul class=\"st-toggle-buttons__buttons sth-toggle-buttons__buttons\">\n      <li *ngFor=\"let tab of parsedTabs; let i = index;\"\n         (click)=\"onClick(tab)\">\n         <button [attr.id]=\"qaTag + '-' + tab.trimLabel\" class=\"st-toggle-buttons__button sth-toggle-buttons__button\"\n         [ngClass]=\"{'st-toggle-buttons__button--active': tab.active, 'sth-toggle-buttons__button--active': tab.active}\">\n            {{tab.label}}\n            <span *ngIf=\"tab.number?.toFixed\">({{tab.number}})</span>\n         </button>\n      </li>\n   </ul>\n</section>\n",
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    styles: ["@charset \"UTF-8\";.st-toggle-buttons{list-style:none}.st-toggle-buttons li{display:inline-block}.st-toggle-buttons .st-toggle-buttons__description{display:inline-block;height:35px;line-height:35px;padding-right:10px}.st-toggle-buttons .st-toggle-buttons__buttons{display:inline-block;border-width:1px;border-style:solid;border-radius:4px}.st-toggle-buttons .st-toggle-buttons__button{display:inline-block;padding:11px 21px;cursor:pointer;box-shadow:none}.st-toggle-buttons .st-toggle-buttons__button:focus{outline:0}"]
                }] }
    ];
    StToggleButtonsComponent.propDecorators = {
        description: [{ type: Input }],
        qaTag: [{ type: Input }],
        select: [{ type: Output }],
        tabs: [{ type: Input }]
    };
    return StToggleButtonsComponent;
}());
export { StToggleButtonsComponent };
if (false) {
    /** @type {?} */
    StToggleButtonsComponent.prototype.description;
    /** @type {?} */
    StToggleButtonsComponent.prototype.qaTag;
    /** @type {?} */
    StToggleButtonsComponent.prototype.select;
    /** @type {?} */
    StToggleButtonsComponent.prototype.parsedTabs;
}
