// @stratio/egeo/esm5/lib/st-alerts/st-alerts.model.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/st-alerts/st-alerts.model.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/*
 * © 2017 Stratio Big Data Inc., Sucursal en España.
 *
 * This software is licensed under the Apache License, Version 2.0.
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the terms of the License for more details.
 *
 * SPDX-License-Identifier: Apache-2.0.
 */
import { Subject } from 'rxjs';
/** @enum {number} */
var STALERT_SEVERITY = {
    SUCCESS: 0, WARNING: 1, ERROR: 2, INFO: 3,
};
export { STALERT_SEVERITY };
STALERT_SEVERITY[STALERT_SEVERITY.SUCCESS] = 'SUCCESS';
STALERT_SEVERITY[STALERT_SEVERITY.WARNING] = 'WARNING';
STALERT_SEVERITY[STALERT_SEVERITY.ERROR] = 'ERROR';
STALERT_SEVERITY[STALERT_SEVERITY.INFO] = 'INFO';
var StAlertLink = /** @class */ (function () {
    function StAlertLink() {
    }
    return StAlertLink;
}());
export { StAlertLink };
if (false) {
    /** @type {?} */
    StAlertLink.prototype.title;
    /** @type {?} */
    StAlertLink.prototype.link;
}
var StAlert = /** @class */ (function () {
    function StAlert(id, title, message, severity, timeout, extendedTimeout, link) {
        this.id = id;
        this.title = title;
        this.message = message;
        this.severity = severity;
        this.timeout = timeout;
        this.extendedTimeout = extendedTimeout;
        this.link = link;
        this._readed = false;
        this._opacity = 0;
        this._opacityState = new Subject();
        this._removeEvent = new Subject();
    }
    Object.defineProperty(StAlert.prototype, "opacity", {
        get: /**
         * @return {?}
         */
        function () {
            return this._opacityState.asObservable();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(StAlert.prototype, "removeAlertEvent", {
        get: /**
         * @return {?}
         */
        function () {
            return this._removeEvent.asObservable();
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    StAlert.prototype.notify = /**
     * @return {?}
     */
    function () {
        this.setVisible(true);
    };
    /**
     * @return {?}
     */
    StAlert.prototype.pauseNotify = /**
     * @return {?}
     */
    function () {
        this._opacity = 1;
        this._readed = true;
        this.clearAnimation();
        this.stopLife();
        this._opacityState.next(this._opacity);
    };
    /**
     * @return {?}
     */
    StAlert.prototype.continueNotify = /**
     * @return {?}
     */
    function () {
        this.startLife();
    };
    /**
     * @return {?}
     */
    StAlert.prototype.cancel = /**
     * @return {?}
     */
    function () {
        this.setVisible(false);
    };
    /**
     * @private
     * @param {?} increase
     * @return {?}
     */
    StAlert.prototype.setVisible = /**
     * @private
     * @param {?} increase
     * @return {?}
     */
    function (increase) {
        var _this = this;
        this.clearAnimation();
        this._changeVisibilityInterval = window.setInterval((/**
         * @return {?}
         */
        function () { return _this.modifyVisibility(increase); }), 50);
    };
    /**
     * @private
     * @param {?} increase
     * @return {?}
     */
    StAlert.prototype.modifyVisibility = /**
     * @private
     * @param {?} increase
     * @return {?}
     */
    function (increase) {
        this._opacity += increase ? 0.1 : -0.1;
        if (this._opacity >= 1 || this._opacity <= 0) {
            this.clearAnimation();
            if (increase) {
                this.startLife();
            }
            else {
                this.notifyForRemove();
            }
        }
        this._opacityState.next(this._opacity);
    };
    /**
     * @private
     * @return {?}
     */
    StAlert.prototype.notifyForRemove = /**
     * @private
     * @return {?}
     */
    function () {
        this._opacityState.complete();
        this._removeEvent.next(this);
        this._removeEvent.complete();
    };
    /**
     * @private
     * @return {?}
     */
    StAlert.prototype.clearAnimation = /**
     * @private
     * @return {?}
     */
    function () {
        window.clearInterval(this._changeVisibilityInterval);
    };
    /**
     * @private
     * @return {?}
     */
    StAlert.prototype.stopLife = /**
     * @private
     * @return {?}
     */
    function () {
        window.clearTimeout(this._lifeTimeout);
    };
    /**
     * @private
     * @return {?}
     */
    StAlert.prototype.startLife = /**
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        /** @type {?} */
        var timeout = this._readed ? this.extendedTimeout : this.timeout;
        this._lifeTimeout = window.setTimeout((/**
         * @return {?}
         */
        function () { return _this.setVisible(false); }), timeout);
    };
    return StAlert;
}());
export { StAlert };
if (false) {
    /**
     * @type {?}
     * @private
     */
    StAlert.prototype._changeVisibilityInterval;
    /**
     * @type {?}
     * @private
     */
    StAlert.prototype._lifeTimeout;
    /**
     * @type {?}
     * @private
     */
    StAlert.prototype._readed;
    /**
     * @type {?}
     * @private
     */
    StAlert.prototype._opacity;
    /**
     * @type {?}
     * @private
     */
    StAlert.prototype._opacityState;
    /**
     * @type {?}
     * @private
     */
    StAlert.prototype._removeEvent;
    /** @type {?} */
    StAlert.prototype.id;
    /** @type {?} */
    StAlert.prototype.title;
    /** @type {?} */
    StAlert.prototype.message;
    /** @type {?} */
    StAlert.prototype.severity;
    /** @type {?} */
    StAlert.prototype.timeout;
    /** @type {?} */
    StAlert.prototype.extendedTimeout;
    /** @type {?} */
    StAlert.prototype.link;
}
