// @stratio/egeo/esm5/lib/st-two-list-selection/list-selection/list-selection.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/st-two-list-selection/list-selection/list-selection.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
/*
 * © 2017 Stratio Big Data Inc., Sucursal en España.
 *
 * This software is licensed under the Apache License, Version 2.0.
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the terms of the License for more details.
 *
 * SPDX-License-Identifier: Apache-2.0.
 */
import { ChangeDetectionStrategy, Component, EventEmitter, Input, Output } from '@angular/core';
import { StRequired } from '../../decorators/require-decorators';
var ListSelectionComponent = /** @class */ (function () {
    function ListSelectionComponent() {
        this.editable = false;
        this.important = false;
        this.hasAllList = false;
        this.hasSearch = true;
        this.isLoading = false;
        this.orderOptions = [];
        this.mode = 'normal';
        this.scrollBottom = new EventEmitter();
        this.selectItem = new EventEmitter();
        this.selectItemNonEditable = new EventEmitter();
        this.selectExtraLabel = new EventEmitter();
        this.search = new EventEmitter();
        this.changeOrder = new EventEmitter();
        this.scrollItems = [];
    }
    Object.defineProperty(ListSelectionComponent.prototype, "searchQaTag", {
        get: /**
         * @return {?}
         */
        function () {
            return this.qaTag + '-search';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ListSelectionComponent.prototype, "listCheckAllQaTag", {
        get: /**
         * @return {?}
         */
        function () {
            return this.qaTag + '-all-checkbox';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ListSelectionComponent.prototype, "listQaTag", {
        get: /**
         * @return {?}
         */
        function () {
            return this.qaTag + '-list';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ListSelectionComponent.prototype, "hasOrder", {
        get: /**
         * @return {?}
         */
        function () {
            return this.orderOptions && this.orderOptions.length > 0;
        },
        enumerable: true,
        configurable: true
    });
    ListSelectionComponent.decorators = [
        { type: Component, args: [{
                    selector: 'list-selection',
                    template: "<!--\n\n    \u00A9 2017 Stratio Big Data Inc., Sucursal en Espa\u00F1a.\n\n    This software is licensed under the Apache License, Version 2.0.\n    This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;\n    without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.\n    See the terms of the License for more details.\n\n    SPDX-License-Identifier: Apache-2.0.\n\n-->\n<div class=\"list-selection\">\n   <div class=\"list-selection__rectangle\">\n      <p class=\"title sth-two-list__list-title\" [ngClass]=\"{'sth-two-list__important-title': important}\">{{title}}</p>\n      <p class=\"subtitle sth-two-list__list-subtitle\">{{subtitle}}</p>\n      <st-search class=\"search\" *ngIf=\"hasSearch && ( list && list.length >= showSearchNumber) || (hasSearch && !showSearchNumber)\" [placeholder]=\"searchPlaceholder\" [qaTag]=\"searchQaTag\" (search)=\"search.emit($event)\"></st-search>\n   </div>\n   <list-item [disabled]=\"list.length === 0\" class=\"list-selection__list-item-all\" *ngIf=\"hasAllList && scrollItems\" [item]=\"itemAll\"\n      (selectItem)=\"selectItem.emit(itemAll);\" [editable]=\"editable\" [qaTag]=\"listCheckAllQaTag\"></list-item>\n\n   <list-scroll [list]=\"list\" [itemAll]=\"itemAll\" [mode]=\"mode\" (selectItem)=\"selectItem.emit($event)\" (selectItemNonEditable)=\"selectItemNonEditable.emit($event)\"  [hasAllList]=\"hasAllList\" [qaTag]=\"listQaTag\" [editable]=\"editable\" [qaTag]=\"listQaTag\" [editable]=\"editable\"\n             (selectExtraLabel)=\"selectExtraLabel.emit($event)\" (scrollBottom)=\"scrollBottom.emit($event)\" [isLoading]=\"isLoading\" [fetchingDataText]=\"fetchingDataText\" [menuOptionList]=\"menuOptionList\"></list-scroll>\n   <div class=\"list-selection__rectangle bottom\" *ngIf=\"hasOrder\" >\n         <st-select [disabled]=\"list.length === 0\" class=\"order-select\" button=\"Order\" [options]=\"orderOptions\" (select)=\"changeOrder.emit($event)\" [placeholder]=\"orderPlaceholder\"></st-select>\n   </div>\n</div>\n",
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    styles: ["@charset \"UTF-8\";.list-selection{display:flex;flex-direction:column;flex-wrap:nowrap;width:100%;height:100%}.list-selection__rectangle{background-color:#fafafa;padding:22px;border-bottom-width:1px;border-bottom-color:#dbdbdb}.list-selection__rectangle.bottom{border-bottom-width:0;border-top-width:1px;border-top-color:#dbdbdb}.list-selection__list-item-all{padding:5px 22px;margin-bottom:5px;border-bottom-width:1px;border-bottom-color:#dbdbdb}.title{width:100%;margin-bottom:5px;padding:0;font-weight:400;font-size:.938rem;color:#585858}.subtitle{font-size:1rem;color:#585858;font-weight:400;font-style:normal;font-stretch:normal}.list{list-style:none;padding:0;width:100%;height:80%;overflow:auto}.list.with-order{height:71%}.search{margin-top:18px;display:block}.order-select{display:block}virtual-scroll{display:block;width:100%;height:100%}"]
                }] }
    ];
    ListSelectionComponent.propDecorators = {
        list: [{ type: Input }],
        editable: [{ type: Input }],
        title: [{ type: Input }],
        subtitle: [{ type: Input }],
        searchPlaceholder: [{ type: Input }],
        qaTag: [{ type: Input }],
        important: [{ type: Input }],
        hasAllList: [{ type: Input }],
        hasSearch: [{ type: Input }],
        isLoading: [{ type: Input }],
        itemAll: [{ type: Input }],
        menuOptionList: [{ type: Input }],
        orderOptions: [{ type: Input }],
        orderPlaceholder: [{ type: Input }],
        mode: [{ type: Input }],
        showSearchNumber: [{ type: Input }],
        fetchingDataText: [{ type: Input }],
        scrollBottom: [{ type: Output }],
        selectItem: [{ type: Output }],
        selectItemNonEditable: [{ type: Output }],
        selectExtraLabel: [{ type: Output }],
        search: [{ type: Output }],
        changeOrder: [{ type: Output }]
    };
    tslib_1.__decorate([
        StRequired(),
        tslib_1.__metadata("design:type", Array)
    ], ListSelectionComponent.prototype, "list", void 0);
    tslib_1.__decorate([
        StRequired(),
        tslib_1.__metadata("design:type", String)
    ], ListSelectionComponent.prototype, "title", void 0);
    tslib_1.__decorate([
        StRequired(),
        tslib_1.__metadata("design:type", String)
    ], ListSelectionComponent.prototype, "searchPlaceholder", void 0);
    tslib_1.__decorate([
        StRequired(),
        tslib_1.__metadata("design:type", String)
    ], ListSelectionComponent.prototype, "qaTag", void 0);
    return ListSelectionComponent;
}());
export { ListSelectionComponent };
if (false) {
    /** @type {?} */
    ListSelectionComponent.prototype.list;
    /** @type {?} */
    ListSelectionComponent.prototype.editable;
    /** @type {?} */
    ListSelectionComponent.prototype.title;
    /** @type {?} */
    ListSelectionComponent.prototype.subtitle;
    /** @type {?} */
    ListSelectionComponent.prototype.searchPlaceholder;
    /** @type {?} */
    ListSelectionComponent.prototype.qaTag;
    /** @type {?} */
    ListSelectionComponent.prototype.important;
    /** @type {?} */
    ListSelectionComponent.prototype.hasAllList;
    /** @type {?} */
    ListSelectionComponent.prototype.hasSearch;
    /** @type {?} */
    ListSelectionComponent.prototype.isLoading;
    /** @type {?} */
    ListSelectionComponent.prototype.itemAll;
    /** @type {?} */
    ListSelectionComponent.prototype.menuOptionList;
    /** @type {?} */
    ListSelectionComponent.prototype.orderOptions;
    /** @type {?} */
    ListSelectionComponent.prototype.orderPlaceholder;
    /** @type {?} */
    ListSelectionComponent.prototype.mode;
    /** @type {?} */
    ListSelectionComponent.prototype.showSearchNumber;
    /** @type {?} */
    ListSelectionComponent.prototype.fetchingDataText;
    /** @type {?} */
    ListSelectionComponent.prototype.scrollBottom;
    /** @type {?} */
    ListSelectionComponent.prototype.selectItem;
    /** @type {?} */
    ListSelectionComponent.prototype.selectItemNonEditable;
    /** @type {?} */
    ListSelectionComponent.prototype.selectExtraLabel;
    /** @type {?} */
    ListSelectionComponent.prototype.search;
    /** @type {?} */
    ListSelectionComponent.prototype.changeOrder;
    /** @type {?} */
    ListSelectionComponent.prototype.scrollItems;
}
