// @stratio/egeo/esm5/lib/st-dropdown-menu/st-dropdown-menu.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/st-dropdown-menu/st-dropdown-menu.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
/*
 * © 2017 Stratio Big Data Inc., Sucursal en España.
 *
 * This software is licensed under the Apache License, Version 2.0.
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the terms of the License for more details.
 *
 * SPDX-License-Identifier: Apache-2.0.
 */
import { ChangeDetectionStrategy, ChangeDetectorRef, Component, ElementRef, EventEmitter, HostListener, Input, Output, Renderer2, ViewChild } from '@angular/core';
import { StPopOffset, StPopPlacement } from '../st-pop/st-pop.model';
import { ARROW_KEY_CODE, StDropDownMenuItem, StDropDownVisualMode } from './st-dropdown-menu.interface';
/**
 * \@description {Component} [Dropdown Menu]
 * This directive show a dropdown menu list in element that you attach
 *
 *
 * \@model
 *
 *   [Menu items] {./st-dropdown-menu.interface.ts#StDropDownMenuItem}
 *
 * \@example
 *
 * {html}
 *
 * ```
 * <st-dropdown-menu [items]="list" [active]="show" (change)="onChange(event)">
 *    <button class="button button-primary" (click)="show = !show">Show menu</button>
 * </st-dropdown-menu>
 * ```
 */
var StDropdownMenuComponent = /** @class */ (function () {
    function StDropdownMenuComponent(el, cd, renderer) {
        this.el = el;
        this.cd = cd;
        this.renderer = renderer;
        /**
         * \@Input {boolean} [active=false] Show or hide list
         */
        this.active = false;
        /**
         * \@Input {StDropDownMenuItem[] | StDropDownMenuGroup[]} [items=\[\]] List of items or groups of them to show in menu
         */
        this.items = [];
        /* tslint:disable-next-line:max-line-length */
        /**
         * \@Input {StPopPlacement} [placement=StPopPlacement.BOTTOM_START] Possible positions of menu with respect element to attach
         */
        this.placement = StPopPlacement.BOTTOM_START;
        /**
         * \@Input {string} [emptyListMessage=''] Message to show in case of empty list
         */
        this.emptyListMessage = '';
        /**
         * \@Input {StDropDownMenuItem | undefined} [selectedItem=undefined] Define selected item without passing as property
         */
        this.selectedItem = undefined;
        /**
         * \@Input {StDropDownMenuItem | undefined} [itemsBeforeScroll=undefined] Define selected item without passing as property
         */
        this.itemsBeforeScroll = 8;
        /**
         * \@Input {boolean} [moveSelected=true] If true, move selected item to top in menu when open
         */
        this.moveSelected = true;
        /**
         * \@Input {boolean} [styleSelected=true] If true, apply class selected to selected item
         */
        this.styleSelected = true;
        /**
         * \@Input {boolean} [isLoading=true] If true, show loader at the end of the list
         */
        this.isLoading = false;
        /**
         * \@Input {StPopOffset} [offset={x: 0 , y: 0}] For position with offset in x o y axis
         */
        this.offset = { x: 0, y: 0 };
        /**
         * \@Input {boolean} [openToLeft=false] For calculating all positions from the right corner
         */
        this.openToLeft = false;
        /**
         * \@Input {StDropdownVisualMode} [visualMode=StDropDownVisualMode.OPTION_LIST] It is needed to specify the styles applied to the list.
         *  By default is displayed as a normal option list
         */
        this.visualMode = StDropDownVisualMode.OPTION_LIST;
        /**
         * \@Input {boolean} [keyBoardMove=false] It is needed to activate navigation through options using the keyboard
         */
        this.keyBoardMove = false;
        /**
         * \@output {StDropDownMenuItem} [change] Event emitted when user select an item
         */
        this.change = new EventEmitter();
        /**
         * \@output {any} [scrollAtBottom] Event emitted when scroll reach the end of the list
         */
        this.scrollAtBottom = new EventEmitter();
        /**
         * \@output {StDropDownMenuItem} [itemMouseEnter] Event emitted when mouse is over an item
         */
        this.itemMouseEnter = new EventEmitter();
        /**
         * \@output {StDropDownMenuItem} [itemMouseLeave] Event emitted when mouse leaves an item
         */
        this.itemMouseLeave = new EventEmitter();
        this.widthMenu = '0px';
        this._itemHeight = 42;
        this._focusedOptionPos = -1;
    }
    /**
     * @return {?}
     */
    StDropdownMenuComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        if (this.keyBoardMove) {
            this._focusListenerFn = this.renderer.listen('document', 'keydown', this.arrowKeyListener.bind(this));
        }
    };
    Object.defineProperty(StDropdownMenuComponent.prototype, "componentId", {
        get: /**
         * @return {?}
         */
        function () {
            /** @type {?} */
            var id = ((/** @type {?} */ (this.el.nativeElement))).getAttribute('id');
            return id !== undefined && id !== null ? id : null;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(StDropdownMenuComponent.prototype, "menuId", {
        get: /**
         * @return {?}
         */
        function () {
            return this.componentId !== null ? this.componentId + "-menu" : null;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(StDropdownMenuComponent.prototype, "isItemGroup", {
        get: /**
         * @return {?}
         */
        function () {
            return this.isDropDownGroup(this.items);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(StDropdownMenuComponent.prototype, "menuMaxHeight", {
        get: /**
         * @return {?}
         */
        function () {
            return this.itemsBeforeScroll ? this._itemHeight * this.itemsBeforeScroll + "px" : null;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(StDropdownMenuComponent.prototype, "listClasses", {
        get: /**
         * @return {?}
         */
        function () {
            return { 'st-dropdown-menu': true, 'active': this.active, 'menu-list': this.displayAsMenuList() };
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} value
     * @return {?}
     */
    StDropdownMenuComponent.prototype.getItemId = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        return this.componentId !== null && value !== undefined ? this.componentId + "-option-" + this.getItemValueMerged(value) : null;
    };
    /**
     * @param {?} value
     * @return {?}
     */
    StDropdownMenuComponent.prototype.isDropDownGroup = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        return value && value.length > 0 && ((/** @type {?} */ (value[0]))).title !== undefined;
    };
    /**
     * @return {?}
     */
    StDropdownMenuComponent.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        this.updateWidth();
    };
    /**
     * @param {?} changes
     * @return {?}
     */
    StDropdownMenuComponent.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        var _this = this;
        if (changes && changes.active && changes.active.currentValue && this.selectedItem && this.moveSelected) {
            // Only can do this functionality with timeout because we need to wait for angular to load new DOM
            // with items before move scroll
            setTimeout((/**
             * @return {?}
             */
            function () {
                if (_this.itemListElement) {
                    /** @type {?} */
                    var parent_1 = _this.itemListElement.nativeElement;
                    /** @type {?} */
                    var listOfElements = parent_1.getElementsByClassName('selected');
                    if (listOfElements && listOfElements.length > 0) {
                        /** @type {?} */
                        var target = ((/** @type {?} */ (listOfElements.item(0))));
                        parent_1.scrollTop = target.offsetTop - parent_1.offsetTop;
                    }
                }
            }));
        }
        else {
            if (changes && changes.active && !changes.active.currentValue) {
                this._focusedOptionPos = -1;
            }
        }
    };
    /**
     * @param {?} value
     * @return {?}
     */
    StDropdownMenuComponent.prototype.onChange = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        this.change.emit(value);
    };
    /**
     * @return {?}
     */
    StDropdownMenuComponent.prototype.onHandleScroll = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var element = this.itemListElement.nativeElement;
        if (element.scrollHeight - element.scrollTop === element.clientHeight && !this.isLoading) {
            this.scrollAtBottom.emit();
        }
    };
    /**
     * @param {?} item
     * @return {?}
     */
    StDropdownMenuComponent.prototype.onMouseEnter = /**
     * @param {?} item
     * @return {?}
     */
    function (item) {
        this.itemMouseEnter.emit(item);
    };
    /**
     * @param {?} item
     * @return {?}
     */
    StDropdownMenuComponent.prototype.onMouseLeave = /**
     * @param {?} item
     * @return {?}
     */
    function (item) {
        this.itemMouseLeave.emit(item);
    };
    /**
     * @return {?}
     */
    StDropdownMenuComponent.prototype.displayAsMenuList = /**
     * @return {?}
     */
    function () {
        return this.visualMode === StDropDownVisualMode.MENU_LIST;
    };
    /**
     * @return {?}
     */
    StDropdownMenuComponent.prototype.updateWidth = /**
     * @return {?}
     */
    function () {
        var _this = this;
        /** @type {?} */
        var button = this.buttonElement.nativeElement;
        setTimeout((/**
         * @return {?}
         */
        function () {
            if (button.children && button.children.length > 0) {
                _this.widthMenu = button.children[0].getBoundingClientRect().width + 'px';
            }
            else {
                _this.widthMenu = button.getBoundingClientRect().width + 'px';
            }
            _this.cd.markForCheck();
        }));
    };
    /**
     * @return {?}
     */
    StDropdownMenuComponent.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        if (this._focusListenerFn) {
            this._focusListenerFn();
        }
    };
    /**
     * @private
     * @param {?} value
     * @return {?}
     */
    StDropdownMenuComponent.prototype.getItemValueMerged = /**
     * @private
     * @param {?} value
     * @return {?}
     */
    function (value) {
        return value.toString().replace(/\s+/g, '_');
    };
    /**
     * @private
     * @return {?}
     */
    StDropdownMenuComponent.prototype.getSelectedItemPosition = /**
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        if (this.selectedItem && this.items) {
            /** @type {?} */
            var _items_1 = [];
            if (this.isDropDownGroup(this.items)) {
                this.items.forEach((/**
                 * @param {?} item
                 * @return {?}
                 */
                function (item) {
                    if (((/** @type {?} */ (item))).items) {
                        _items_1.push.apply(_items_1, tslib_1.__spread(((/** @type {?} */ (item))).items));
                    }
                    else {
                        _items_1.push(((/** @type {?} */ (item))));
                    }
                }));
            }
            else {
                _items_1 = this.items;
            }
            return _items_1.findIndex((/**
             * @param {?} item
             * @return {?}
             */
            function (item) { return item.value === _this.selectedItem.value; }));
        }
        else {
            return -1;
        }
    };
    /**
     * @private
     * @param {?} e
     * @return {?}
     */
    StDropdownMenuComponent.prototype.arrowKeyListener = /**
     * @private
     * @param {?} e
     * @return {?}
     */
    function (e) {
        /** @type {?} */
        var selectedItemPosition = this.getSelectedItemPosition();
        if (selectedItemPosition > -1) {
            this._focusedOptionPos = selectedItemPosition;
        }
        /** @type {?} */
        var nextFocus;
        if (e.code === ARROW_KEY_CODE.ARROW_DOWN || e.code === ARROW_KEY_CODE.ARROW_UP) {
            event.preventDefault();
            /** @type {?} */
            var options = this.el.nativeElement.querySelectorAll('.st-dropdown-menu-item');
            nextFocus = e.code === ARROW_KEY_CODE.ARROW_DOWN || this._focusedOptionPos === -1 ? 1 : -1;
            this._focusedOptionPos = this._focusedOptionPos + nextFocus;
            if (this._focusedOptionPos < 0) {
                this._focusedOptionPos = options.length - 1;
            }
            else if (this._focusedOptionPos > options.length - 1) {
                this._focusedOptionPos = 0;
            }
            if (options[this._focusedOptionPos]) {
                options[this._focusedOptionPos].focus();
            }
        }
        this.cd.markForCheck();
    };
    StDropdownMenuComponent.decorators = [
        { type: Component, args: [{
                    selector: 'st-dropdown-menu',
                    template: "<!--\n\n    \u00A9 2017 Stratio Big Data Inc., Sucursal en Espa\u00F1a.\n\n    This software is licensed under the Apache License, Version 2.0.\n    This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;\n    without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.\n    See the terms of the License for more details.\n\n    SPDX-License-Identifier: Apache-2.0.\n\n-->\n<st-pop [hidden]=\"!active\" [placement]=\"placement\" [offset]=\"offset\" [openToLeft]=\"openToLeft\">\n\n   <div pop-button #buttonId>\n      <ng-content></ng-content>\n   </div>\n\n   <div pop-content [style.z-index]=\"9\" class=\"dropdown-content\">\n\n      <div *ngIf=\"active\" [ngClass]=\"listClasses\" [style.minWidth]='widthMenu'>\n         <ng-content select=[dropdown-header]></ng-content>\n\n         <ul *ngIf=\"!isItemGroup && ((items && items.length) || emptyListMessage)\"\n             #itemList\n             aria-label=\"submenu\"\n             [ngClass]=\"{'st-custom-scrollbar': displayAsMenuList()}\"\n             [attr.aria-hidden]=\"!active\"\n             [style.maxHeight]=\"menuMaxHeight\"\n             [attr.id]=\"menuId\"\n             (scroll)=\"onHandleScroll()\">\n            <st-dropdown-menu-item\n               *ngFor=\"let item of items; let i = index\"\n               [attr.id]=\"getItemId(item.value)\"\n               [index]=\"i\"\n               [item]=\"item\"\n               [selectedItem]=\"selectedItem\"\n               [styleSelected]=\"styleSelected\"\n               (change)=\"onChange($event)\"\n               (mouseenter)=\"onMouseEnter(item)\"\n               (mouseleave)=\"onMouseLeave(item)\">\n            </st-dropdown-menu-item>\n            <div *ngIf=\"isLoading\" class=\"spinner-container\">\n               <st-spinner class=\"spinner\"></st-spinner>\n            </div>\n            <div *ngIf=\"!items || !items.length\" class=\"without-results st-dropdown-menu-without-results\"\n                 [style.maxWidth]='widthMenu'>\n               {{emptyListMessage}}\n            </div>\n         </ul>\n\n         <div *ngIf=\"isItemGroup\">\n            <ul [attr.aria-hidden]=\"!active\" aria-label=\"submenu\" [style.maxHeight]=\"menuMaxHeight\"\n                [attr.id]=\"menuId\" #itemList>\n               <li class=\"st-dropdown-menu-parent\" *ngFor=\"let group of items\">\n                  <h3>\n                     <span>{{ group.title }}</span>\n                  </h3>\n                  <ul>\n                     <st-dropdown-menu-item\n                        *ngFor=\"let item of group.items || []; let i = index\"\n                        [attr.id]=\"getItemId(item.value)\"\n                        [index]=\"i\"\n                        [item]=\"item\"\n                        [selectedItem]=\"selectedItem\"\n                        [styleSelected]=\"styleSelected\"\n                        (change)=\"onChange($event)\"\n                        (mouseenter)=\"onMouseEnter(item)\"\n                        (mouseleave)=\"onMouseLeave(item)\">\n                     </st-dropdown-menu-item>\n\n                     <div class=\"without-results st-dropdown-menu-without-results\" [style.maxWidth]='widthMenu'\n                          *ngIf=\"!group || !group.items || !group.items.length\">\n                        {{emptyListMessage}}\n                     </div>\n                  </ul>\n               </li>\n               <div *ngIf=\"isLoading\" class=\"spinner-container\">\n                  <st-spinner class=\"spinner\"></st-spinner>\n               </div>\n            </ul>\n         </div>\n         <ng-content select=[dropdown-footer]></ng-content>\n      </div>\n   </div>\n</st-pop>\n",
                    changeDetection: ChangeDetectionStrategy.OnPush
                }] }
    ];
    /** @nocollapse */
    StDropdownMenuComponent.ctorParameters = function () { return [
        { type: ElementRef },
        { type: ChangeDetectorRef },
        { type: Renderer2 }
    ]; };
    StDropdownMenuComponent.propDecorators = {
        active: [{ type: Input }],
        items: [{ type: Input }],
        placement: [{ type: Input }],
        emptyListMessage: [{ type: Input }],
        selectedItem: [{ type: Input }],
        itemsBeforeScroll: [{ type: Input }],
        moveSelected: [{ type: Input }],
        styleSelected: [{ type: Input }],
        isLoading: [{ type: Input }],
        offset: [{ type: Input }],
        openToLeft: [{ type: Input }],
        visualMode: [{ type: Input }],
        keyBoardMove: [{ type: Input }],
        change: [{ type: Output }],
        scrollAtBottom: [{ type: Output }],
        itemMouseEnter: [{ type: Output }],
        itemMouseLeave: [{ type: Output }],
        buttonElement: [{ type: ViewChild, args: ['buttonId', { static: false },] }],
        itemListElement: [{ type: ViewChild, args: ['itemList', { static: false },] }],
        updateWidth: [{ type: HostListener, args: ['window:resize',] }, { type: HostListener, args: ['window:load',] }]
    };
    return StDropdownMenuComponent;
}());
export { StDropdownMenuComponent };
if (false) {
    /**
     * \@Input {boolean} [active=false] Show or hide list
     * @type {?}
     */
    StDropdownMenuComponent.prototype.active;
    /**
     * \@Input {StDropDownMenuItem[] | StDropDownMenuGroup[]} [items=\[\]] List of items or groups of them to show in menu
     * @type {?}
     */
    StDropdownMenuComponent.prototype.items;
    /**
     * \@Input {StPopPlacement} [placement=StPopPlacement.BOTTOM_START] Possible positions of menu with respect element to attach
     * @type {?}
     */
    StDropdownMenuComponent.prototype.placement;
    /**
     * \@Input {string} [emptyListMessage=''] Message to show in case of empty list
     * @type {?}
     */
    StDropdownMenuComponent.prototype.emptyListMessage;
    /**
     * \@Input {StDropDownMenuItem | undefined} [selectedItem=undefined] Define selected item without passing as property
     * @type {?}
     */
    StDropdownMenuComponent.prototype.selectedItem;
    /**
     * \@Input {StDropDownMenuItem | undefined} [itemsBeforeScroll=undefined] Define selected item without passing as property
     * @type {?}
     */
    StDropdownMenuComponent.prototype.itemsBeforeScroll;
    /**
     * \@Input {boolean} [moveSelected=true] If true, move selected item to top in menu when open
     * @type {?}
     */
    StDropdownMenuComponent.prototype.moveSelected;
    /**
     * \@Input {boolean} [styleSelected=true] If true, apply class selected to selected item
     * @type {?}
     */
    StDropdownMenuComponent.prototype.styleSelected;
    /**
     * \@Input {boolean} [isLoading=true] If true, show loader at the end of the list
     * @type {?}
     */
    StDropdownMenuComponent.prototype.isLoading;
    /**
     * \@Input {StPopOffset} [offset={x: 0 , y: 0}] For position with offset in x o y axis
     * @type {?}
     */
    StDropdownMenuComponent.prototype.offset;
    /**
     * \@Input {boolean} [openToLeft=false] For calculating all positions from the right corner
     * @type {?}
     */
    StDropdownMenuComponent.prototype.openToLeft;
    /**
     * \@Input {StDropdownVisualMode} [visualMode=StDropDownVisualMode.OPTION_LIST] It is needed to specify the styles applied to the list.
     *  By default is displayed as a normal option list
     * @type {?}
     */
    StDropdownMenuComponent.prototype.visualMode;
    /**
     * \@Input {boolean} [keyBoardMove=false] It is needed to activate navigation through options using the keyboard
     * @type {?}
     */
    StDropdownMenuComponent.prototype.keyBoardMove;
    /**
     * \@output {StDropDownMenuItem} [change] Event emitted when user select an item
     * @type {?}
     */
    StDropdownMenuComponent.prototype.change;
    /**
     * \@output {any} [scrollAtBottom] Event emitted when scroll reach the end of the list
     * @type {?}
     */
    StDropdownMenuComponent.prototype.scrollAtBottom;
    /**
     * \@output {StDropDownMenuItem} [itemMouseEnter] Event emitted when mouse is over an item
     * @type {?}
     */
    StDropdownMenuComponent.prototype.itemMouseEnter;
    /**
     * \@output {StDropDownMenuItem} [itemMouseLeave] Event emitted when mouse leaves an item
     * @type {?}
     */
    StDropdownMenuComponent.prototype.itemMouseLeave;
    /** @type {?} */
    StDropdownMenuComponent.prototype.buttonElement;
    /** @type {?} */
    StDropdownMenuComponent.prototype.itemListElement;
    /** @type {?} */
    StDropdownMenuComponent.prototype.widthMenu;
    /**
     * @type {?}
     * @private
     */
    StDropdownMenuComponent.prototype._itemHeight;
    /**
     * @type {?}
     * @private
     */
    StDropdownMenuComponent.prototype._focusedOptionPos;
    /**
     * @type {?}
     * @private
     */
    StDropdownMenuComponent.prototype._focusListenerFn;
    /**
     * @type {?}
     * @private
     */
    StDropdownMenuComponent.prototype.el;
    /**
     * @type {?}
     * @private
     */
    StDropdownMenuComponent.prototype.cd;
    /**
     * @type {?}
     * @private
     */
    StDropdownMenuComponent.prototype.renderer;
}
