// @stratio/egeo/esm2015/lib/st-two-list-selection/st-two-list-selection.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/st-two-list-selection/st-two-list-selection.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
/*
 * © 2017 Stratio Big Data Inc., Sucursal en España.
 *
 * This software is licensed under the Apache License, Version 2.0.
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the terms of the License for more details.
 *
 * SPDX-License-Identifier: Apache-2.0.
 */
import { ChangeDetectionStrategy, ChangeDetectorRef, Component, EventEmitter, Input, OnChanges, OnInit, Output, SimpleChanges } from '@angular/core';
import { StEgeo, StRequired } from '../decorators/require-decorators';
import { StTwoListSelection } from './st-two-list-selection';
let StTwoListSelectionComponent = class StTwoListSelectionComponent extends StTwoListSelection {
    /**
     * @param {?} cd
     */
    constructor(cd) {
        super(cd);
        this.cd = cd;
        this.editable = false;
        this.hasAllListAll = false;
        this.hasAllListSelected = false;
        this.hasSearch = true;
        this.isLoading = false;
        this.mode = 'normal';
        this.moveAllToSelectedButton = true;
        this.moveAllToAllButton = true;
        this.moveToSelectedButton = true;
        this.moveToAllButton = true;
        this.orderAllOptions = [];
        this.orderSelectedOptions = [];
        this.sortBy = 'id';
        this.change = new EventEmitter();
        this.changeOrderAll = new EventEmitter();
        this.changeOrderSelected = new EventEmitter();
        this.numItemsSelectedAll = new EventEmitter();
        this.numItemsSelectedSelected = new EventEmitter();
        this.scrollBottomAll = new EventEmitter();
        this.searchOverAll = new EventEmitter();
        this.searchOverSelected = new EventEmitter();
        this.selectedElementsChange = new EventEmitter();
        this.selectExtraLabelAll = new EventEmitter();
        this.selectExtraLabelSelected = new EventEmitter();
        this.selectItemNonEditable = new EventEmitter();
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.init(this.allElements, this.selectedElements, this.selectedElementsChange, this.sortBy, this.hasAllListAll, this.hasAllListSelected, this.itemAll);
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    ngOnChanges(changes) {
        this.checkChanges(changes, 'allElements', 'selectedElements');
        this.change.emit();
    }
    /**
     * @return {?}
     */
    get allList() {
        return this.copyAllElement;
    }
    /**
     * @return {?}
     */
    get selectedList() {
        return this.copySelectedElements;
    }
};
StTwoListSelectionComponent.ctorParameters = () => [
    { type: ChangeDetectorRef }
];
StTwoListSelectionComponent.decorators = [
    { type: Component, args: [{
                selector: 'st-two-list-selection',
                template: `
      <st-two-list-selection-view
         [allElements]="allList | stFilterList:searchBy:allSearch"
         [selectedElements]="selectedList | stFilterList:searchBy:selectedSearch"
         [config]="config"
         [editable]="editable"
         [hasAllListAll]="hasAllListAll"
         [hasAllListSelected]="hasAllListSelected"
         [itemAll]="itemAll"
         [qaTag]="qaTag"
         [isLoading]="isLoading"
         [hasSearch]="hasSearch"
         [orderAllOptions]="orderAllOptions"
         [orderSelectedOptions]="orderSelectedOptions"
         [mode]="mode"
         [moveAllToSelectedButton]="moveAllToSelectedButton"
         [moveAllToAllButton]="moveAllToAllButton"
         [moveToAllButton]="moveToAllButton"
         [moveToSelectedButton]="moveToSelectedButton"
         [showSearchNumber]="showSearchNumber"

         (selectAllElement)="onSelectAllElement($event)"
         (selectSelectedElement)="onSelectSelectedElement($event)"
         (searchOverAll)="onSearchOverAll($event)"
         (searchOverSelected)="onSearchOverSelected($event)"
         (moveToSelected)="onMoveToSelected($event)"
         (moveToAll)="onMoveToAll($event)"
         (moveAllToSelected)="onMoveAllToSelected($event)"
         (moveAllToAll)="onMoveAllToAll($event)"
         (numItemsSelectedAll)="numItemsSelectedAll.emit($event)"
         (numItemsSelectedSelected)="numItemsSelectedSelected.emit($event)"
         (selectExtraLabelAll)="selectExtraLabelAll.emit($event)"
         (selectExtraLabelSelected)="selectExtraLabelAll.emit($event)"
         (selectItemNonEditable)="selectItemNonEditable.emit($event)"
         (changeOrderAll)="changeOrderAll.emit($event)"
         (changeOrderSelected)="changeOrderSelected.emit($event)"
         (scrollBottomAll)="scrollBottomAll.emit($event)"
      ></st-two-list-selection-view>
   `,
                changeDetection: ChangeDetectionStrategy.OnPush
            }] }
];
/** @nocollapse */
StTwoListSelectionComponent.ctorParameters = () => [
    { type: ChangeDetectorRef }
];
StTwoListSelectionComponent.propDecorators = {
    allElements: [{ type: Input }],
    config: [{ type: Input }],
    editable: [{ type: Input }],
    hasAllListAll: [{ type: Input }],
    hasAllListSelected: [{ type: Input }],
    hasSearch: [{ type: Input }],
    isLoading: [{ type: Input }],
    itemAll: [{ type: Input }],
    mode: [{ type: Input }],
    moveAllToSelectedButton: [{ type: Input }],
    moveAllToAllButton: [{ type: Input }],
    moveToSelectedButton: [{ type: Input }],
    moveToAllButton: [{ type: Input }],
    orderAllOptions: [{ type: Input }],
    orderSelectedOptions: [{ type: Input }],
    qaTag: [{ type: Input }],
    selectedElements: [{ type: Input }],
    showSearchNumber: [{ type: Input }],
    sortBy: [{ type: Input }],
    change: [{ type: Output }],
    changeOrderAll: [{ type: Output }],
    changeOrderSelected: [{ type: Output }],
    numItemsSelectedAll: [{ type: Output }],
    numItemsSelectedSelected: [{ type: Output }],
    scrollBottomAll: [{ type: Output }],
    searchOverAll: [{ type: Output }],
    searchOverSelected: [{ type: Output }],
    selectedElementsChange: [{ type: Output }],
    selectExtraLabelAll: [{ type: Output }],
    selectExtraLabelSelected: [{ type: Output }],
    selectItemNonEditable: [{ type: Output }]
};
tslib_1.__decorate([
    StRequired('editable'),
    tslib_1.__metadata("design:type", Array)
], StTwoListSelectionComponent.prototype, "allElements", void 0);
tslib_1.__decorate([
    StRequired(),
    tslib_1.__metadata("design:type", String)
], StTwoListSelectionComponent.prototype, "qaTag", void 0);
tslib_1.__decorate([
    StRequired(),
    tslib_1.__metadata("design:type", Array)
], StTwoListSelectionComponent.prototype, "selectedElements", void 0);
StTwoListSelectionComponent = tslib_1.__decorate([
    StEgeo(),
    tslib_1.__metadata("design:paramtypes", [ChangeDetectorRef])
], StTwoListSelectionComponent);
export { StTwoListSelectionComponent };
if (false) {
    /** @type {?} */
    StTwoListSelectionComponent.prototype.allElements;
    /** @type {?} */
    StTwoListSelectionComponent.prototype.config;
    /** @type {?} */
    StTwoListSelectionComponent.prototype.editable;
    /** @type {?} */
    StTwoListSelectionComponent.prototype.hasAllListAll;
    /** @type {?} */
    StTwoListSelectionComponent.prototype.hasAllListSelected;
    /** @type {?} */
    StTwoListSelectionComponent.prototype.hasSearch;
    /** @type {?} */
    StTwoListSelectionComponent.prototype.isLoading;
    /** @type {?} */
    StTwoListSelectionComponent.prototype.itemAll;
    /** @type {?} */
    StTwoListSelectionComponent.prototype.mode;
    /** @type {?} */
    StTwoListSelectionComponent.prototype.moveAllToSelectedButton;
    /** @type {?} */
    StTwoListSelectionComponent.prototype.moveAllToAllButton;
    /** @type {?} */
    StTwoListSelectionComponent.prototype.moveToSelectedButton;
    /** @type {?} */
    StTwoListSelectionComponent.prototype.moveToAllButton;
    /** @type {?} */
    StTwoListSelectionComponent.prototype.orderAllOptions;
    /** @type {?} */
    StTwoListSelectionComponent.prototype.orderSelectedOptions;
    /** @type {?} */
    StTwoListSelectionComponent.prototype.qaTag;
    /** @type {?} */
    StTwoListSelectionComponent.prototype.selectedElements;
    /** @type {?} */
    StTwoListSelectionComponent.prototype.showSearchNumber;
    /** @type {?} */
    StTwoListSelectionComponent.prototype.sortBy;
    /** @type {?} */
    StTwoListSelectionComponent.prototype.change;
    /** @type {?} */
    StTwoListSelectionComponent.prototype.changeOrderAll;
    /** @type {?} */
    StTwoListSelectionComponent.prototype.changeOrderSelected;
    /** @type {?} */
    StTwoListSelectionComponent.prototype.numItemsSelectedAll;
    /** @type {?} */
    StTwoListSelectionComponent.prototype.numItemsSelectedSelected;
    /** @type {?} */
    StTwoListSelectionComponent.prototype.scrollBottomAll;
    /** @type {?} */
    StTwoListSelectionComponent.prototype.searchOverAll;
    /** @type {?} */
    StTwoListSelectionComponent.prototype.searchOverSelected;
    /** @type {?} */
    StTwoListSelectionComponent.prototype.selectedElementsChange;
    /** @type {?} */
    StTwoListSelectionComponent.prototype.selectExtraLabelAll;
    /** @type {?} */
    StTwoListSelectionComponent.prototype.selectExtraLabelSelected;
    /** @type {?} */
    StTwoListSelectionComponent.prototype.selectItemNonEditable;
    /**
     * @type {?}
     * @private
     */
    StTwoListSelectionComponent.prototype.cd;
}
