// @stratio/egeo/esm2015/lib/st-two-list-selection/st-two-list-selection.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/st-two-list-selection/st-two-list-selection.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/*
 * © 2017 Stratio Big Data Inc., Sucursal en España.
 *
 * This software is licensed under the Apache License, Version 2.0.
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the terms of the License for more details.
 *
 * SPDX-License-Identifier: Apache-2.0.
 */
import { EventEmitter } from '@angular/core';
import * as _ from 'lodash';
import { StFilterList } from '../pipes/search-filter/search-filter.pipe';
export class StTwoListSelection {
    /**
     * @param {?} _cd
     */
    constructor(_cd) {
        this._cd = _cd;
        // Work Lists
        this.copyAllElement = [];
        this.copySelectedElements = [];
        this.allSearch = '';
        this.hasCheckboxAllList = false;
        this.hasCheckboxSelectedList = false;
        this.numItemsSelectedAll = new EventEmitter();
        this.numItemsSelectedSelected = new EventEmitter();
        this.searchBy = 'name';
        this.searchOverAll = new EventEmitter();
        this.searchOverSelected = new EventEmitter();
        this.selectedSearch = '';
    }
    // Check selected element
    /**
     * @param {?} selection
     * @return {?}
     */
    onSelectAllElement(selection) {
        if (this.hasCheckboxAllList && selection.itemAll) {
            this.copyAllElement = this.applySelectAll(this.copyAllElement, this.allSearch, selection.selected);
            selection.selected = !selection.selected;
        }
        else {
            if (this.canSelect(selection, this.copyAllElement)) {
                selection.selected = !selection.selected;
            }
        }
        this.copyAllElement = _.cloneDeep(this.copyAllElement);
        if (this.copyAllElement) {
            this.numItemsSelectedAll.emit(this.getNumItemsSelected(this.copyAllElement));
        }
    }
    // Check selected element
    /**
     * @param {?} selection
     * @return {?}
     */
    onSelectSelectedElement(selection) {
        if (this.hasCheckboxSelectedList && selection.itemAll) {
            this.copySelectedElements = this.applySelectAll(this.copySelectedElements, this.selectedSearch, selection.selected);
            selection.selected = !selection.selected;
        }
        else {
            if (this.canSelect(selection, this.copySelectedElements)) {
                selection.selected = !selection.selected;
            }
        }
        this.copySelectedElements = _.cloneDeep(this.copySelectedElements);
        if (this.copySelectedElements) {
            this.numItemsSelectedSelected.emit(this.getNumItemsSelected(this.copySelectedElements));
        }
    }
    // Update search filter
    /**
     * @param {?} search
     * @return {?}
     */
    onSearchOverAll(search) {
        this.allSearch = search.text;
        this.searchOverAll.emit(this.allSearch);
    }
    // Update search filter
    /**
     * @param {?} search
     * @return {?}
     */
    onSearchOverSelected(search) {
        this.selectedSearch = search.text;
        this.searchOverSelected.emit(this.selectedSearch);
    }
    // Move from all to selected
    /**
     * @param {?} event
     * @return {?}
     */
    onMoveToSelected(event) {
        /** @type {?} */
        let ids = this.getIdsToMove(this.copyAllElement);
        /** @type {?} */
        let result = this.moveIdsFromAllToSelected(this.originalAll, this.originalSelected, ids);
        this.emitter.emit(result);
        if (this.itemAll)
            this.itemAll.selected = false;
    }
    // Remove from selected
    /**
     * @param {?} event
     * @return {?}
     */
    onMoveToAll(event) {
        /** @type {?} */
        let ids = this.getIdsToMove(this.copySelectedElements);
        /** @type {?} */
        let result = this.removeIdsFromSelected(this.originalSelected, ids);
        this.emitter.emit(result);
        if (this.itemAll)
            this.itemAll.selected = false;
    }
    // Move all to selected
    /**
     * @param {?} event
     * @return {?}
     */
    onMoveAllToSelected(event) {
        this.emitter.emit(_.cloneDeep(this.originalAll.filter((/**
         * @param {?} elem
         * @return {?}
         */
        (elem) => !elem.disabled))));
    }
    // Remove All from selected
    /**
     * @param {?} event
     * @return {?}
     */
    onMoveAllToAll(event) {
        this.emitter.emit([]);
    }
    /**
     * @param {?} all
     * @param {?} selected
     * @param {?} changeEmitter
     * @param {?} sorted
     * @param {?=} hasCheckboxAllList
     * @param {?=} hasCheckboxSelectedList
     * @param {?=} itemAll
     * @return {?}
     */
    init(all, selected, changeEmitter, sorted, hasCheckboxAllList = false, hasCheckboxSelectedList = false, itemAll = (/** @type {?} */ (null))) {
        this.emitter = changeEmitter;
        this.sortLists = sorted;
        this.originalAll = all;
        this.originalSelected = selected;
        this.generateWorkLists();
        this.hasCheckboxAllList = hasCheckboxAllList;
        this.hasCheckboxSelectedList = hasCheckboxSelectedList;
        this.itemAll = itemAll;
    }
    /**
     * @param {?} changes
     * @param {?} allList
     * @param {?} selectedList
     * @return {?}
     */
    checkChanges(changes, allList, selectedList) {
        if (changes[allList] !== undefined) {
            this.originalAll = changes[allList].currentValue;
        }
        if (changes[selectedList] !== undefined) {
            this.originalSelected = changes[selectedList].currentValue;
        }
        if (changes[allList] !== undefined || changes[selectedList] !== undefined) {
            this.generateWorkLists();
        }
    }
    /**
     * @private
     * @param {?} twoList
     * @param {?} elemSearched
     * @param {?} allSelected
     * @return {?}
     */
    applySelectAll(twoList, elemSearched, allSelected) {
        /** @type {?} */
        let filterPipe = new StFilterList();
        /** @type {?} */
        let filteredBySearch = filterPipe.transform(twoList, 'name', elemSearched);
        if (elemSearched) {
            twoList.forEach((/**
             * @param {?} elem
             * @return {?}
             */
            (elem) => {
                if (!elem.disabled) {
                    elem.selected = false;
                    if (filteredBySearch.find((/**
                     * @param {?} filtered
                     * @return {?}
                     */
                    filtered => filtered.name === elem.name)) !== undefined) {
                        elem.selected = !allSelected;
                    }
                }
            }));
            this._cd.markForCheck();
        }
        else {
            twoList = _.cloneDeep(this.changeSelectedItemList(twoList, !allSelected));
        }
        return twoList;
    }
    /**
     * @private
     * @param {?} list
     * @param {?} selected
     * @return {?}
     */
    changeSelectedItemList(list, selected) {
        return _.cloneDeep(list).map((/**
         * @param {?} elem
         * @return {?}
         */
        (elem) => {
            elem.selected = (!elem.disabled) ? selected : elem.selected;
            return elem;
        }));
    }
    /**
     * @private
     * @param {?} list
     * @return {?}
     */
    getNumItemsSelected(list) {
        /** @type {?} */
        const reducer = (/**
         * @param {?} accumulator
         * @param {?} currentValue
         * @return {?}
         */
        (accumulator, currentValue) => {
            return (currentValue.selected) ? accumulator + 1 : accumulator + 0;
        });
        return list.reduce(reducer, 0);
    }
    /**
     * @private
     * @return {?}
     */
    generateWorkLists() {
        this.copyAllElement = this.copyLists(this.originalAll, this.copyAllElement, this.originalSelected);
        this.copySelectedElements = this.copyLists(this.originalSelected, this.copySelectedElements);
        if (this.sortLists !== undefined && (this.sortLists === 'id' || this.sortLists === 'name')) {
            this.copyAllElement = this.orderListBy(this.copyAllElement, this.sortLists);
            this.copySelectedElements = this.orderListBy(this.copySelectedElements, this.sortLists);
        }
    }
    /**
     * @private
     * @param {?} newValue
     * @param {?} prevState
     * @param {?=} without
     * @return {?}
     */
    copyLists(newValue, prevState, without) {
        /** @type {?} */
        let list = _.cloneDeep(newValue);
        if (without && without.length > 0) {
            list = _.differenceBy(list, without, 'id');
        }
        /** @type {?} */
        let selected = this.getIdsToMove(prevState);
        this.searchAndSelect(list, selected);
        return list;
    }
    /**
     * @private
     * @param {?} list
     * @param {?} selected
     * @return {?}
     */
    searchAndSelect(list, selected) {
        /** @type {?} */
        let i = 0;
        /** @type {?} */
        let actualIndex = -1;
        while (selected.length > 0 && i < list.length) {
            actualIndex = selected.indexOf(list[i].id);
            if (actualIndex > -1) {
                list[i].selected = true;
                selected.splice(actualIndex, 1);
                actualIndex = -1;
            }
            i++;
        }
    }
    /**
     * @private
     * @param {?} list
     * @return {?}
     */
    getIdsToMove(list) {
        return list.reduce((/**
         * @param {?} prev
         * @param {?} curr
         * @return {?}
         */
        (prev, curr) => {
            if (curr.selected) {
                curr.selected = false;
                prev.push(curr.id);
            }
            return prev;
        }), []);
    }
    /**
     * @private
     * @param {?} all
     * @param {?} selected
     * @param {?} ids
     * @return {?}
     */
    moveIdsFromAllToSelected(all, selected, ids) {
        /** @type {?} */
        let items = all.filter((/**
         * @param {?} item
         * @return {?}
         */
        (item) => _.includes(ids, item.id)));
        return _.xorBy(selected, _.cloneDeep(items), 'id');
    }
    /**
     * @private
     * @param {?} selected
     * @param {?} ids
     * @return {?}
     */
    removeIdsFromSelected(selected, ids) {
        return selected.reduce((/**
         * @param {?} prev
         * @param {?} curr
         * @return {?}
         */
        (prev, curr) => {
            if (!_.includes(ids, curr.id)) {
                prev.push(_.clone(curr));
            }
            return prev;
        }), []);
    }
    /**
     * @private
     * @param {?} list
     * @param {?} by
     * @return {?}
     */
    orderListBy(list, by) {
        return _.sortBy(list, by);
    }
    /**
     * @private
     * @param {?} selection
     * @param {?} list
     * @return {?}
     */
    canSelect(selection, list) {
        return selection && !selection.disabled && list !== undefined && list.length > 0;
    }
}
if (false) {
    /** @type {?} */
    StTwoListSelection.prototype.originalAll;
    /** @type {?} */
    StTwoListSelection.prototype.originalSelected;
    /** @type {?} */
    StTwoListSelection.prototype.copyAllElement;
    /** @type {?} */
    StTwoListSelection.prototype.copySelectedElements;
    /** @type {?} */
    StTwoListSelection.prototype.allSearch;
    /** @type {?} */
    StTwoListSelection.prototype.hasCheckboxAllList;
    /** @type {?} */
    StTwoListSelection.prototype.hasCheckboxSelectedList;
    /** @type {?} */
    StTwoListSelection.prototype.itemAll;
    /** @type {?} */
    StTwoListSelection.prototype.numItemsSelectedAll;
    /** @type {?} */
    StTwoListSelection.prototype.numItemsSelectedSelected;
    /** @type {?} */
    StTwoListSelection.prototype.searchBy;
    /** @type {?} */
    StTwoListSelection.prototype.searchOverAll;
    /** @type {?} */
    StTwoListSelection.prototype.searchOverSelected;
    /** @type {?} */
    StTwoListSelection.prototype.selectedSearch;
    /**
     * @type {?}
     * @private
     */
    StTwoListSelection.prototype.emitter;
    /**
     * @type {?}
     * @private
     */
    StTwoListSelection.prototype.sortLists;
    /**
     * @type {?}
     * @private
     */
    StTwoListSelection.prototype._cd;
}
