// @stratio/egeo/esm5/lib/st-sidebar/st-sidebar-item-list/st-sidebar-item-list.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/st-sidebar/st-sidebar-item-list/st-sidebar-item-list.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/*
 * © 2017 Stratio Big Data Inc., Sucursal en España.
 *
 * This software is licensed under the Apache License, Version 2.0.
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the terms of the License for more details.
 *
 * SPDX-License-Identifier: Apache-2.0.
 */
import { ChangeDetectionStrategy, Component, Input, EventEmitter, Output } from '@angular/core';
import { StSidebarVisualMode } from '../st-sidebar-visual-mode';
/**
 * \@description {Component} [SidebarItemList]
 *
 * The sidebar item list component has been designed to display a list of items in a sidebar.
 *
 * \@model
 *
 *   [Sidebar items] {../st-sidebar-item.interface.ts#StSidebarItem}
 *
 * \@example
 *
 * {html}
 *
 * ```
 * <st-sidebar-item-list [items]="items" (change)="onChange($event)" [active]="active" [deep]="deep">
 * </st-sidebar-item-list>
 * ```
 *
 */
var StSidebarItemListComponent = /** @class */ (function () {
    function StSidebarItemListComponent() {
        /**
         * \@Input {StSidebarItem[]} [items=''] List of items displayed on the menu
         */
        this.items = [];
        /**
         * \@Input {number} [deep=0] Deep of the item list in the sidebar
         */
        this.deep = 0;
        /**
         * \@Input {StSidebarVisualMode} [visualMode='StSidebarVisualMode.normal'] Visual mode used to display the item list
         */
        this.visualMode = StSidebarVisualMode.normal;
        /**
         * \@Output {StSidebarItem} [change=''] Event emitted when the active item is changed. This emit the active item
         */
        this.change = new EventEmitter();
        this.displayAsComplexModeValue = false;
        this.expanded = [];
    }
    Object.defineProperty(StSidebarItemListComponent.prototype, "active", {
        /** @Input {string} [active=''] The id of the current active item */
        get: /**
         * \@Input {string} [active=''] The id of the current active item
         * @return {?}
         */
        function () {
            return this._active;
        },
        set: /**
         * @param {?} active
         * @return {?}
         */
        function (active) {
            if (!this._isActive(active)) {
                this._active = active;
                this._updateStatus();
            }
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    StSidebarItemListComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        if (!this._active && this.items && this.items.length) {
            this._active = this.items[0];
        }
        this._updateStatus();
        this.displayAsComplexModeValue = this.displayAsComplexMode();
    };
    /**
     * @param {?} item
     * @param {?} index
     * @return {?}
     */
    StSidebarItemListComponent.prototype.getItemClasses = /**
     * @param {?} item
     * @param {?} index
     * @return {?}
     */
    function (item, index) {
        /** @type {?} */
        var classes = {};
        /** @type {?} */
        var active = this._isActive(item);
        /** @type {?} */
        var activeChild = this.hasActiveChild(item);
        classes[item.class] = item.class;
        classes['item--complex'] = item.items && item.items.length;
        classes['item--active'] = active;
        classes['item--expanded'] = this.expanded[index];
        classes['item--has-active'] = activeChild;
        classes['item--disabled'] = item.disabled;
        return classes;
    };
    /**
     * @param {?} item
     * @param {?} position
     * @return {?}
     */
    StSidebarItemListComponent.prototype.onSelectItem = /**
     * @param {?} item
     * @param {?} position
     * @return {?}
     */
    function (item, position) {
        if (!item.disabled) {
            if (item.items && item.items.length) {
                this.expanded[position] = !this.expanded[position];
            }
            else {
                if (!this._active || this._active.id !== item.id) {
                    this.change.emit(item);
                    this._closeRestOfItems(item.id);
                }
            }
        }
    };
    /**
     * @param {?} item
     * @return {?}
     */
    StSidebarItemListComponent.prototype.hasActiveChild = /**
     * @param {?} item
     * @return {?}
     */
    function (item) {
        /** @type {?} */
        var found = false;
        /** @type {?} */
        var i = 0;
        if (item.items && item.items.length) {
            while (!found && i < item.items.length) {
                if (this._isActive(item.items[i])) {
                    found = true;
                }
                ++i;
            }
            if (!found && item.items && item.items.length) {
                return this._hasActiveChildInChildren(item.items);
            }
        }
        return found;
    };
    /**
     * @param {?} item
     * @return {?}
     */
    StSidebarItemListComponent.prototype.onChange = /**
     * @param {?} item
     * @return {?}
     */
    function (item) {
        if (!this._isActive(item)) {
            this._closeRestOfItems(item.id);
            this.change.emit(item);
        }
    };
    /**
     * @return {?}
     */
    StSidebarItemListComponent.prototype.displayAsComplexMode = /**
     * @return {?}
     */
    function () {
        return this.visualMode === StSidebarVisualMode.complex;
    };
    /**
     * @private
     * @param {?} item
     * @return {?}
     */
    StSidebarItemListComponent.prototype._isActive = /**
     * @private
     * @param {?} item
     * @return {?}
     */
    function (item) {
        return item && this._active && this._active.id === item.id;
    };
    /**
     * @private
     * @return {?}
     */
    StSidebarItemListComponent.prototype._updateStatus = /**
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        this.items.forEach((/**
         * @param {?} item
         * @param {?} i
         * @return {?}
         */
        function (item, i) {
            if (_this.hasActiveChild(item)) {
                _this.expanded[i] = true;
            }
        }));
    };
    /**
     * @private
     * @param {?} items
     * @return {?}
     */
    StSidebarItemListComponent.prototype._hasActiveChildInChildren = /**
     * @private
     * @param {?} items
     * @return {?}
     */
    function (items) {
        /** @type {?} */
        var found = false;
        /** @type {?} */
        var i = 0;
        while (!found && i < items.length) {
            if (this.hasActiveChild(items[i])) {
                found = true;
            }
            ++i;
        }
        return found;
    };
    /**
     * @private
     * @param {?} itemId
     * @return {?}
     */
    StSidebarItemListComponent.prototype._closeRestOfItems = /**
     * @private
     * @param {?} itemId
     * @return {?}
     */
    function (itemId) {
        var _this = this;
        this.expanded.forEach((/**
         * @param {?} open
         * @param {?} i
         * @return {?}
         */
        function (open, i) {
            if (open && _this.items[i].id !== itemId) {
                _this.expanded[i] = false;
            }
        }));
    };
    StSidebarItemListComponent.decorators = [
        { type: Component, args: [{
                    selector: 'st-sidebar-item-list',
                    template: "<!--\n\n    \u00A9 2017 Stratio Big Data Inc., Sucursal en Espa\u00F1a.\n\n    This software is licensed under the Apache License, Version 2.0.\n    This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;\n    without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.\n    See the terms of the License for more details.\n\n    SPDX-License-Identifier: Apache-2.0.\n\n-->\n\n<ul class=\"item-list\" [ngClass]=\"{'item-list--inner': deep > 0, 'complex-mode': displayAsComplexModeValue}\">\n   <li [id]=\"item.id\" class=\"item\" [ngClass]=\"getItemClasses(item, index)\" *ngFor=\"let item of items; let index = index\">\n      <span class=\"item__label\" [ngClass]=\"{'item__label--with-results': item.result}\" (click)=\"onSelectItem(item, index)\">{{item.label}}</span>\n      <span *ngIf=\"item.result\" class=\"result__box\"><span class=\"result__text\">{{item.result}}</span></span>\n      <st-sidebar-item-list *ngIf=\"item.items && item.items.length && expanded[index]\"\n                            [items]=\"item.items\"\n                            [active]=\"active\"\n                            [visualMode]=\"visualMode\"\n                            [deep]=\"deep + 1\"\n                            (change)=\"onChange($event)\">\n      </st-sidebar-item-list>\n   </li>\n</ul>\n",
                    providers: [],
                    host: { class: 'st-sidebar-item-list' },
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    styles: ["@charset \"UTF-8\";:host{width:100%;height:100%;display:block}.item-list{list-style:none}.item{cursor:pointer}"]
                }] }
    ];
    StSidebarItemListComponent.propDecorators = {
        items: [{ type: Input }],
        deep: [{ type: Input }],
        visualMode: [{ type: Input }],
        change: [{ type: Output }],
        active: [{ type: Input }]
    };
    return StSidebarItemListComponent;
}());
export { StSidebarItemListComponent };
if (false) {
    /**
     * \@Input {StSidebarItem[]} [items=''] List of items displayed on the menu
     * @type {?}
     */
    StSidebarItemListComponent.prototype.items;
    /**
     * \@Input {number} [deep=0] Deep of the item list in the sidebar
     * @type {?}
     */
    StSidebarItemListComponent.prototype.deep;
    /**
     * \@Input {StSidebarVisualMode} [visualMode='StSidebarVisualMode.normal'] Visual mode used to display the item list
     * @type {?}
     */
    StSidebarItemListComponent.prototype.visualMode;
    /**
     * \@Output {StSidebarItem} [change=''] Event emitted when the active item is changed. This emit the active item
     * @type {?}
     */
    StSidebarItemListComponent.prototype.change;
    /** @type {?} */
    StSidebarItemListComponent.prototype.displayAsComplexModeValue;
    /** @type {?} */
    StSidebarItemListComponent.prototype.expanded;
    /**
     * @type {?}
     * @private
     */
    StSidebarItemListComponent.prototype._active;
}
