// @stratio/egeo/esm5/lib/st-horizontal-tabs/st-horizontal-tabs.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/st-horizontal-tabs/st-horizontal-tabs.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
/*
 * © 2017 Stratio Big Data Inc., Sucursal en España.
 *
 * This software is licensed under the Apache License, Version 2.0.
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the terms of the License for more details.
 *
 * SPDX-License-Identifier: Apache-2.0.
 */
import { ChangeDetectionStrategy, Component, EventEmitter, Input, Output } from '@angular/core';
import { StHorizontalTabStatus } from './st-horizontal-tabs.model';
import { StEgeo, StRequired } from '../decorators/require-decorators';
/**
 * \@description {Component} [Horizontal tabs]
 *
 * The tabs are a navigation component that divides content into separate views hiding the ones that the user is not focused in.
 *
 * \@model
 *
 *   [Id, text and optional status of a tab] {./st-horizontal-tabs.model.ts#StStHorizontalTab}
 *
 * \@example
 *
 * {html}
 *
 * ```
 * <st-horizontal-tabs [options]="options" [qaTag]="qaTag">
 * </st-horizontal-tabs>
 * ```
 *
 */
var StHorizontalTabsComponent = /** @class */ (function () {
    function StHorizontalTabsComponent() {
        /**
         * \@Output {StHorizontalTab} [changedOption=''] This event is emitted when active option has changed. It has the
         * active option name as param
         */
        this.changedOption = new EventEmitter();
    }
    /**
     * @return {?}
     */
    StHorizontalTabsComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        if (this.options && this.options.length > 0) {
            this.activeOption = this.activeOption || this.options[0];
            this.activateOption(this.activeOption);
        }
    };
    /**
     * @param {?} option
     * @return {?}
     */
    StHorizontalTabsComponent.prototype.isActive = /**
     * @param {?} option
     * @return {?}
     */
    function (option) {
        return this.activeOption === option;
    };
    /**
     * @param {?} option
     * @return {?}
     */
    StHorizontalTabsComponent.prototype.activateOption = /**
     * @param {?} option
     * @return {?}
     */
    function (option) {
        this.activeOption = option;
        this.changedOption.emit(option);
    };
    /**
     * @param {?} tab
     * @return {?}
     */
    StHorizontalTabsComponent.prototype.getTabStatus = /**
     * @param {?} tab
     * @return {?}
     */
    function (tab) {
        return tab.status ? StHorizontalTabStatus[tab.status] : '';
    };
    StHorizontalTabsComponent.decorators = [
        { type: Component, args: [{
                    selector: 'st-horizontal-tabs',
                    template: "<!--\n\n    \u00A9 2017 Stratio Big Data Inc., Sucursal en Espa\u00F1a.\n\n    This software is licensed under the Apache License, Version 2.0.\n    This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;\n    without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.\n    See the terms of the License for more details.\n\n    SPDX-License-Identifier: Apache-2.0.\n\n-->\n<nav class=\"st-horizontal-tabs\">\n   <a *ngFor=\"let option of options; let i = index;\"\n      class=\"tab {{getTabStatus(option)}}\"\n      (click)=\"activateOption(option)\"\n      href=\"javascript:void(0)\"\n      [ngClass]=\"{'tab--active': isActive(option)}\"\n      [attr.id]=\"qaTag + '-tab-' + i\">{{option.text}}</a>\n   <ng-content></ng-content>\n</nav>\n",
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    styles: ["@charset \"UTF-8\";.st-horizontal-tabs{position:relative;display:flex;width:100%;list-style:none}.tab{display:inline-block;text-align:center;box-sizing:border-box;line-height:1;padding:0 30px 7px;margin-top:auto;cursor:pointer}.tab--active{padding-bottom:5px}.line{display:block;transition:.4s ease-in-out}"]
                }] }
    ];
    StHorizontalTabsComponent.propDecorators = {
        activeOption: [{ type: Input }],
        options: [{ type: Input }],
        qaTag: [{ type: Input }],
        changedOption: [{ type: Output }]
    };
    tslib_1.__decorate([
        StRequired(),
        tslib_1.__metadata("design:type", Array)
    ], StHorizontalTabsComponent.prototype, "options", void 0);
    /**
     * \@description {Component} [Horizontal tabs]
     *
     * The tabs are a navigation component that divides content into separate views hiding the ones that the user is not focused in.
     *
     * \@model
     *
     *   [Id, text and optional status of a tab] {./st-horizontal-tabs.model.ts#StStHorizontalTab}
     *
     * \@example
     *
     * {html}
     *
     * ```
     * <st-horizontal-tabs [options]="options" [qaTag]="qaTag">
     * </st-horizontal-tabs>
     * ```
     *
     */
    StHorizontalTabsComponent = tslib_1.__decorate([
        StEgeo()
    ], StHorizontalTabsComponent);
    return StHorizontalTabsComponent;
}());
export { StHorizontalTabsComponent };
if (false) {
    /**
     * \@Input {StHorizontalTab} [activeOption=''] Current active option
     * @type {?}
     */
    StHorizontalTabsComponent.prototype.activeOption;
    /**
     * \@Input {StHorizontalTab[]} [^options=''] An array of StHorizontalTab objects (see below) that defines the links
     * that will appear and that will be disabled
     * @type {?}
     */
    StHorizontalTabsComponent.prototype.options;
    /**
     * \@Input {string} [qaTag=''] Prefix used to generate the id values for qa tests
     * @type {?}
     */
    StHorizontalTabsComponent.prototype.qaTag;
    /**
     * \@Output {StHorizontalTab} [changedOption=''] This event is emitted when active option has changed. It has the
     * active option name as param
     * @type {?}
     */
    StHorizontalTabsComponent.prototype.changedOption;
}
