// @stratio/egeo/esm5/lib/st-header/st-header.model.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/st-header/st-header.model.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
/*
 * © 2017 Stratio Big Data Inc., Sucursal en España.
 *
 * This software is licensed under the Apache License, Version 2.0.
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the terms of the License for more details.
 *
 * SPDX-License-Identifier: Apache-2.0.
 */
import { StDropDownMenuItem } from '../st-dropdown-menu/st-dropdown-menu.interface';
var StHeaderMenuOption = /** @class */ (function () {
    function StHeaderMenuOption() {
    }
    return StHeaderMenuOption;
}());
export { StHeaderMenuOption };
if (false) {
    /** @type {?} */
    StHeaderMenuOption.prototype.icon;
    /** @type {?} */
    StHeaderMenuOption.prototype.label;
    /** @type {?} */
    StHeaderMenuOption.prototype.subMenus;
    /** @type {?} */
    StHeaderMenuOption.prototype.link;
    /** @type {?} */
    StHeaderMenuOption.prototype.external;
    /** @type {?} */
    StHeaderMenuOption.prototype.openInNewPage;
}
var StHeaderSubMenuOption = /** @class */ (function () {
    function StHeaderSubMenuOption() {
    }
    return StHeaderSubMenuOption;
}());
export { StHeaderSubMenuOption };
if (false) {
    /** @type {?} */
    StHeaderSubMenuOption.prototype.label;
    /** @type {?} */
    StHeaderSubMenuOption.prototype.link;
    /** @type {?} */
    StHeaderSubMenuOption.prototype.external;
    /** @type {?} */
    StHeaderSubMenuOption.prototype.openInNewPage;
}
var StHeaderSelection = /** @class */ (function () {
    function StHeaderSelection() {
    }
    return StHeaderSelection;
}());
export { StHeaderSelection };
if (false) {
    /** @type {?} */
    StHeaderSelection.prototype.link;
    /** @type {?} */
    StHeaderSelection.prototype.external;
    /** @type {?} */
    StHeaderSelection.prototype.openInNewPage;
}
var StHeaderMenuItem = /** @class */ (function (_super) {
    tslib_1.__extends(StHeaderMenuItem, _super);
    function StHeaderMenuItem() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return StHeaderMenuItem;
}(StDropDownMenuItem));
export { StHeaderMenuItem };
if (false) {
    /** @type {?} */
    StHeaderMenuItem.prototype.selection;
}
var StHeaderMenuOptionSchema = /** @class */ (function () {
    function StHeaderMenuOptionSchema() {
    }
    return StHeaderMenuOptionSchema;
}());
export { StHeaderMenuOptionSchema };
if (false) {
    /** @type {?} */
    StHeaderMenuOptionSchema.prototype.icon;
    /** @type {?} */
    StHeaderMenuOptionSchema.prototype.label;
    /** @type {?} */
    StHeaderMenuOptionSchema.prototype.subMenus;
    /** @type {?} */
    StHeaderMenuOptionSchema.prototype.link;
    /** @type {?} */
    StHeaderMenuOptionSchema.prototype.external;
    /** @type {?} */
    StHeaderMenuOptionSchema.prototype.openInNewPage;
}
var StHeaderSubMenuOptionSchema = /** @class */ (function () {
    function StHeaderSubMenuOptionSchema() {
    }
    return StHeaderSubMenuOptionSchema;
}());
export { StHeaderSubMenuOptionSchema };
if (false) {
    /** @type {?} */
    StHeaderSubMenuOptionSchema.prototype.label;
    /** @type {?} */
    StHeaderSubMenuOptionSchema.prototype.link;
    /** @type {?} */
    StHeaderSubMenuOptionSchema.prototype.external;
    /** @type {?} */
    StHeaderSubMenuOptionSchema.prototype.openInNewPage;
}
