// @stratio/egeo/esm5/lib/st-docs/st-prism/st-prism.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/st-docs/st-prism/st-prism.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/*
 * © 2017 Stratio Big Data Inc., Sucursal en España.
 *
 * This software is licensed under the Apache License, Version 2.0.
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the terms of the License for more details.
 *
 * SPDX-License-Identifier: Apache-2.0.
 */
import { Component, Input, ViewEncapsulation } from '@angular/core';
import 'prismjs';
import 'prismjs/components/prism-typescript';
var StPrismComponent = /** @class */ (function () {
    function StPrismComponent() {
        this.async = false;
        this.callback = undefined;
    }
    /**
     * @return {?}
     */
    StPrismComponent.prototype.ngAfterViewChecked = /**
     * @return {?}
     */
    function () {
        this.highlightAll(this.async, this.callback);
    };
    /**
     * Low-level function, only use if you know what you’re doing. It accepts a string of text as input and the
     * language definitions to use, and returns a string with the HTML produced.
     */
    /**
     * Low-level function, only use if you know what you’re doing. It accepts a string of text as input and the
     * language definitions to use, and returns a string with the HTML produced.
     * @param {?} element
     * @param {?} async
     * @param {?=} callback
     * @return {?}
     */
    StPrismComponent.prototype.highlightElement = /**
     * Low-level function, only use if you know what you’re doing. It accepts a string of text as input and the
     * language definitions to use, and returns a string with the HTML produced.
     * @param {?} element
     * @param {?} async
     * @param {?=} callback
     * @return {?}
     */
    function (element, async, callback) {
        Prism.highlightElement(element, async, callback);
    };
    /**
     * This is the most high-level function in Prism’s API. It fetches all the elements that have a .language-xxxx
     * class and then calls Prism.highlightElement() on each one of them.
     */
    /**
     * This is the most high-level function in Prism’s API. It fetches all the elements that have a .language-xxxx
     * class and then calls Prism.highlightElement() on each one of them.
     * @param {?} async
     * @param {?=} callback
     * @return {?}
     */
    StPrismComponent.prototype.highlightAll = /**
     * This is the most high-level function in Prism’s API. It fetches all the elements that have a .language-xxxx
     * class and then calls Prism.highlightElement() on each one of them.
     * @param {?} async
     * @param {?=} callback
     * @return {?}
     */
    function (async, callback) {
        Prism.highlightAll(async, callback);
    };
    StPrismComponent.decorators = [
        { type: Component, args: [{
                    encapsulation: ViewEncapsulation.None,
                    selector: 'st-prism-highlight',
                    template: "<pre class=\"language-{{language}}\" *ngIf=\"language\"><code class=\"language-{{language}}\"><ng-content></ng-content></code></pre>"
                }] }
    ];
    StPrismComponent.propDecorators = {
        language: [{ type: Input, args: ['language',] }],
        async: [{ type: Input, args: ['async',] }],
        callback: [{ type: Input, args: ['callback',] }]
    };
    return StPrismComponent;
}());
export { StPrismComponent };
if (false) {
    /** @type {?} */
    StPrismComponent.prototype.language;
    /**
     * @type {?}
     * @private
     */
    StPrismComponent.prototype.async;
    /**
     * @type {?}
     * @private
     */
    StPrismComponent.prototype.callback;
}
