// @stratio/egeo/esm5/lib/st-table/st-table.interface.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/st-table/st-table.interface.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/*
 * © 2017 Stratio Big Data Inc., Sucursal en España.
 *
 * This software is licensed under the Apache License, Version 2.0.
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the terms of the License for more details.
 *
 * SPDX-License-Identifier: Apache-2.0.
 */
var StTableFilterIconClasses = /** @class */ (function () {
    function StTableFilterIconClasses() {
        this.selected = 'icon-facets-2';
        this.enabled = 'icon-arrow4_down';
    }
    return StTableFilterIconClasses;
}());
export { StTableFilterIconClasses };
if (false) {
    /** @type {?} */
    StTableFilterIconClasses.prototype.selected;
    /** @type {?} */
    StTableFilterIconClasses.prototype.enabled;
}
var StTableIconClasses = /** @class */ (function () {
    function StTableIconClasses() {
        this.filter = new StTableFilterIconClasses();
        this.sort = {
            asc: 'icon-arrow-up',
            desc: 'icon-arrow-down'
        };
    }
    return StTableIconClasses;
}());
export { StTableIconClasses };
if (false) {
    /** @type {?} */
    StTableIconClasses.prototype.filter;
    /** @type {?} */
    StTableIconClasses.prototype.sort;
}
