// @stratio/egeo/esm5/lib/st-header/menu-option/menu-option.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/st-header/menu-option/menu-option.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/*
 * © 2017 Stratio Big Data Inc., Sucursal en España.
 *
 * This software is licensed under the Apache License, Version 2.0.
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the terms of the License for more details.
 *
 * SPDX-License-Identifier: Apache-2.0.
 */
import { ChangeDetectionStrategy, ChangeDetectorRef, Component, ElementRef, EventEmitter, HostBinding, HostListener, Input, Output, ViewChild } from '@angular/core';
import { NavigationEnd, Router } from '@angular/router';
import { StHeaderMenuOption } from '../st-header.model';
import { StDropDownVisualMode } from './../../st-dropdown-menu/st-dropdown-menu.interface';
import { StHeaderUtils } from '../st-header.utils';
var StHeaderMenuOptionComponent = /** @class */ (function () {
    function StHeaderMenuOptionComponent(elementRef, router, cd) {
        var _this = this;
        this.elementRef = elementRef;
        this.router = router;
        this.cd = cd;
        this.selectMenu = new EventEmitter();
        this.visualMode = StDropDownVisualMode.MENU_LIST;
        this._actualPath = '';
        this._subscription = this.router.events.subscribe((/**
         * @param {?} event
         * @return {?}
         */
        function (event) { return _this._onRouterEvent(event); }));
        this._actualPath = this.router.url;
    }
    Object.defineProperty(StHeaderMenuOptionComponent.prototype, "isCurrentRoute", {
        get: /**
         * @return {?}
         */
        function () {
            return StHeaderUtils.isRouteActive(this._option, this.router.url);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(StHeaderMenuOptionComponent.prototype, "option", {
        get: /**
         * @return {?}
         */
        function () {
            return this._option;
        },
        set: /**
         * @param {?} _option
         * @return {?}
         */
        function (_option) {
            this._option = _option;
            this.hasSubmenu = _option.subMenus && _option.subMenus.length > 0;
            this.qaId = this._getQaId();
            this.submenuList = this._getSubmenuList();
            this.cd.markForCheck();
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    StHeaderMenuOptionComponent.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this.isActive = false;
        this._subscription.unsubscribe();
    };
    /**
     * @return {?}
     */
    StHeaderMenuOptionComponent.prototype.onMenuClick = /**
     * @return {?}
     */
    function () {
        if (this.hasSubmenu) {
            this.isActive = !this.isActive;
            this.cd.markForCheck();
        }
        else {
            this.selectMenu.emit({
                link: this._option.link,
                external: this._option.external,
                openInNewPage: this._option.openInNewPage
            });
        }
    };
    /**
     * @param {?} selected
     * @return {?}
     */
    StHeaderMenuOptionComponent.prototype.changeOption = /**
     * @param {?} selected
     * @return {?}
     */
    function (selected) {
        this.isActive = false;
        this.cd.markForCheck();
        this.selectMenu.emit(selected.selection);
    };
    /**
     * @param {?} event
     * @return {?}
     */
    StHeaderMenuOptionComponent.prototype.onClickOutside = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        /** @type {?} */
        var isMyComponent = this.isActive && this.menu.nativeElement.contains(event.target);
        if (!isMyComponent && this.isActive) {
            this.isActive = false;
            this.cd.markForCheck();
        }
    };
    /**
     * @private
     * @param {?} event
     * @return {?}
     */
    StHeaderMenuOptionComponent.prototype._onRouterEvent = /**
     * @private
     * @param {?} event
     * @return {?}
     */
    function (event) {
        if (event instanceof NavigationEnd) {
            this._actualPath = event.urlAfterRedirects;
            this.submenuList = this._getSubmenuList();
            this.cd.markForCheck();
        }
    };
    /**
     * @private
     * @return {?}
     */
    StHeaderMenuOptionComponent.prototype._getQaId = /**
     * @private
     * @return {?}
     */
    function () {
        if (!this._option) {
            return null;
        }
        /** @type {?} */
        var id = this.elementRef.nativeElement.id + "-" + this._option.label.toLowerCase();
        id.replace(/\s+/ig, '_');
        return id;
    };
    /**
     * @private
     * @return {?}
     */
    StHeaderMenuOptionComponent.prototype._getSubmenuList = /**
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        return this._option && this.hasSubmenu ? this._option.subMenus.map((/**
         * @param {?} _
         * @return {?}
         */
        function (_) { return ({
            label: _.label,
            value: _.link,
            selected: _this._actualPath === _.link,
            selection: (/** @type {?} */ ({
                link: _.link,
                external: _.external,
                openInNewPage: _.openInNewPage
            }))
        }); })) : [];
    };
    StHeaderMenuOptionComponent.decorators = [
        { type: Component, args: [{
                    selector: 'st-header-menu-option',
                    template: "<!--\n\n    \u00A9 2017 Stratio Big Data Inc., Sucursal en Espa\u00F1a.\n\n    This software is licensed under the Apache License, Version 2.0.\n    This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;\n    without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.\n    See the terms of the License for more details.\n\n    SPDX-License-Identifier: Apache-2.0.\n\n-->\n\n<st-dropdown-menu [items]=\"submenuList\"\n                  [visualMode]=\"visualMode\"\n                  [active]=\"isActive && hasSubmenu\"\n                  [placement]=\"'bottom-start'\"\n                  [offset]=\"{x: 0, y: 10}\"\n                  (change)=\"changeOption($event)\">\n   <div class=\"st-header-menu-option\" #menu\n        [attr.id]=\"qaId\"\n        (click)=\"onMenuClick()\">\n      <a class=\"st-header-menu-option-name\">\n         <i *ngIf=\"option.icon\"\n            class=\"st-header-menu-option-icon\"\n            [ngClass]=\"option.icon\"\n            [title]=\"option.label\"></i>\n         <div *ngIf=\"showMenuName\" class=\"st-header-menu-option-label\">{{option.label}}</div>\n      </a>\n      <i *ngIf=\"hasSubmenu\" class=\"icon-arrow2_down st-header-menu-option-arrow\"></i>\n   </div>\n</st-dropdown-menu>\n",
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    styles: ["@charset \"UTF-8\";:host{display:flex;justify-content:center;flex-direction:column;height:100%;padding:0 10px}:host.active{background:#566574}:host.active .st-header-menu-option,:host.active .st-header-menu-option:hover{color:#fff}"]
                }] }
    ];
    /** @nocollapse */
    StHeaderMenuOptionComponent.ctorParameters = function () { return [
        { type: ElementRef },
        { type: Router },
        { type: ChangeDetectorRef }
    ]; };
    StHeaderMenuOptionComponent.propDecorators = {
        showMenuName: [{ type: Input }],
        selectMenu: [{ type: Output }],
        menu: [{ type: ViewChild, args: ['menu', { static: false },] }],
        isCurrentRoute: [{ type: HostBinding, args: ['class.active',] }],
        option: [{ type: Input }],
        onClickOutside: [{ type: HostListener, args: ['document:click', ['$event'],] }]
    };
    return StHeaderMenuOptionComponent;
}());
export { StHeaderMenuOptionComponent };
if (false) {
    /** @type {?} */
    StHeaderMenuOptionComponent.prototype.showMenuName;
    /** @type {?} */
    StHeaderMenuOptionComponent.prototype.selectMenu;
    /** @type {?} */
    StHeaderMenuOptionComponent.prototype.menu;
    /** @type {?} */
    StHeaderMenuOptionComponent.prototype.isActive;
    /** @type {?} */
    StHeaderMenuOptionComponent.prototype.visualMode;
    /** @type {?} */
    StHeaderMenuOptionComponent.prototype.hasSubmenu;
    /** @type {?} */
    StHeaderMenuOptionComponent.prototype.qaId;
    /** @type {?} */
    StHeaderMenuOptionComponent.prototype.submenuList;
    /**
     * @type {?}
     * @private
     */
    StHeaderMenuOptionComponent.prototype._subscription;
    /**
     * @type {?}
     * @private
     */
    StHeaderMenuOptionComponent.prototype._actualPath;
    /**
     * @type {?}
     * @private
     */
    StHeaderMenuOptionComponent.prototype._option;
    /**
     * @type {?}
     * @private
     */
    StHeaderMenuOptionComponent.prototype.elementRef;
    /**
     * @type {?}
     * @private
     */
    StHeaderMenuOptionComponent.prototype.router;
    /**
     * @type {?}
     * @private
     */
    StHeaderMenuOptionComponent.prototype.cd;
}
