// @stratio/egeo/esm5/lib/st-pop/st-pop.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/st-pop/st-pop.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/*
 * © 2017 Stratio Big Data Inc., Sucursal en España.
 *
 * This software is licensed under the Apache License, Version 2.0.
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the terms of the License for more details.
 *
 * SPDX-License-Identifier: Apache-2.0.
 */
import { ChangeDetectionStrategy, ChangeDetectorRef, Component, ElementRef, HostListener, Input } from '@angular/core';
import { StPopOffset, StPopPlacement } from './st-pop.model';
/**
 * \@description {Component} [Pop]
 *
 * The pop is a component for manage floating elements like popups or dropdown-menu. This element need two element inside,
 * a button element that launch popper and a content element that whose position will be relativo to button element.
 *
 * \@example
 *
 * {html}
 *
 * ```
 * <st-pop [hidden]="false" placement="bottom">
 *    <div pop-button>Button</div>
 *    <div pop-content>Content</div>
 * </st-pop>
 * ```
 */
var StPopComponent = /** @class */ (function () {
    function StPopComponent(_el, _cd) {
        this._el = _el;
        this._cd = _cd;
        /**
         * \@Input {StPopPlacement} [placement=StPopPlacement.BOTOM_START] Define position of content relative to button
         */
        this.placement = StPopPlacement.BOTTOM_START;
        /**
         * \@Input {StPopOffset} [offset={x: 0 , y: 0}] For position with offset in x o y axis
         */
        this.offset = { x: 0, y: 0 };
        /**
         * \@Input {boolean} [openToLeft=false] For calculating all positions from the right corner
         */
        this.openToLeft = false;
        this._hidden = true;
    }
    Object.defineProperty(StPopComponent.prototype, "hidden", {
        get: /**
         * @return {?}
         */
        function () {
            return this._hidden;
        },
        /** @Input {boleean} [hidden=true]  TRUE: show pop content, FALSE: hide pop content  */
        set: /**
         * \@Input {boleean} [hidden=true]  TRUE: show pop content, FALSE: hide pop content
         * @param {?} value
         * @return {?}
         */
        function (value) {
            var _this = this;
            setTimeout((/**
             * @return {?}
             */
            function () {
                if (!value) {
                    _this.calculatePosition();
                }
                _this._hidden = value;
                _this._cd.markForCheck();
            }));
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    StPopComponent.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        this.calculatePosition();
    };
    /**
     * @param {?} changes
     * @return {?}
     */
    StPopComponent.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        this.calculatePosition();
    };
    /**
     * @return {?}
     */
    StPopComponent.prototype.updateWidth = /**
     * @return {?}
     */
    function () {
        this.calculatePosition();
    };
    /**
     * @private
     * @return {?}
     */
    StPopComponent.prototype.getContentElement = /**
     * @private
     * @return {?}
     */
    function () {
        return this._el.nativeElement.querySelector('[pop-content]');
    };
    /**
     * @private
     * @return {?}
     */
    StPopComponent.prototype.calculatePosition = /**
     * @private
     * @return {?}
     */
    function () {
        /** @type {?} */
        var buttonParentEl = this._el.nativeElement.querySelector('[pop-button]');
        /** @type {?} */
        var contentEl = this.getContentElement();
        /** @type {?} */
        var buttonEl = buttonParentEl && buttonParentEl.firstElementChild ?
            buttonParentEl.firstElementChild : undefined;
        if (buttonEl) {
            /** @type {?} */
            var coords = this.getCoords(buttonEl);
            contentEl.style.position = 'absolute';
            contentEl.style.transform = "translate3d(" + coords.x + "px, " + coords.y + "px, " + coords.z + "px)";
        }
    };
    /**
     * @private
     * @param {?} buttonEl
     * @return {?}
     */
    StPopComponent.prototype.getCoords = /**
     * @private
     * @param {?} buttonEl
     * @return {?}
     */
    function (buttonEl) {
        /** @type {?} */
        var coords = { x: 0, y: 0, z: 0 };
        /** @type {?} */
        var direction = this.openToLeft ? this.getContentElement().getBoundingClientRect().width : 0;
        /** @type {?} */
        var clientRect = buttonEl.getBoundingClientRect();
        switch (this.placement) {
            case StPopPlacement.BOTTOM:
                coords.x = clientRect.width / 2;
                break;
            case StPopPlacement.BOTTOM_END:
                coords.x = clientRect.width;
                break;
            case StPopPlacement.TOP:
                coords.y = clientRect.height * -1;
                coords.x = clientRect.width / 2;
                break;
            case StPopPlacement.TOP_START:
                coords.y = clientRect.height * -1;
                break;
            case StPopPlacement.TOP_END:
                coords.y = clientRect.height * -1;
                coords.x = clientRect.width;
                break;
            default:
                break;
        }
        coords.x = Math.ceil(coords.x + this.offset.x - direction);
        coords.y = Math.ceil(coords.y + this.offset.y);
        return coords;
    };
    StPopComponent.decorators = [
        { type: Component, args: [{
                    selector: 'st-pop',
                    template: "<!--\n\n    \u00A9 2017 Stratio Big Data Inc., Sucursal en Espa\u00F1a.\n\n    This software is licensed under the Apache License, Version 2.0.\n    This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;\n    without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.\n    See the terms of the License for more details.\n\n    SPDX-License-Identifier: Apache-2.0.\n\n-->\n<ng-content select=[pop-button]></ng-content>\n<div [style.visibility]=\"hidden ? 'hidden' : 'visible'\" class=\"content\">\n   <ng-content select=[pop-content]></ng-content>\n</div>\n",
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    styles: ["@charset \"UTF-8\";.content{height:0}"]
                }] }
    ];
    /** @nocollapse */
    StPopComponent.ctorParameters = function () { return [
        { type: ElementRef },
        { type: ChangeDetectorRef }
    ]; };
    StPopComponent.propDecorators = {
        placement: [{ type: Input }],
        offset: [{ type: Input }],
        openToLeft: [{ type: Input }],
        hidden: [{ type: Input }],
        updateWidth: [{ type: HostListener, args: ['window:load',] }]
    };
    return StPopComponent;
}());
export { StPopComponent };
if (false) {
    /**
     * \@Input {StPopPlacement} [placement=StPopPlacement.BOTOM_START] Define position of content relative to button
     * @type {?}
     */
    StPopComponent.prototype.placement;
    /**
     * \@Input {StPopOffset} [offset={x: 0 , y: 0}] For position with offset in x o y axis
     * @type {?}
     */
    StPopComponent.prototype.offset;
    /**
     * \@Input {boolean} [openToLeft=false] For calculating all positions from the right corner
     * @type {?}
     */
    StPopComponent.prototype.openToLeft;
    /**
     * @type {?}
     * @private
     */
    StPopComponent.prototype._hidden;
    /**
     * @type {?}
     * @private
     */
    StPopComponent.prototype._el;
    /**
     * @type {?}
     * @private
     */
    StPopComponent.prototype._cd;
}
