// @stratio/egeo/esm5/lib/st-two-list-selection/st-two-list-selection.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/st-two-list-selection/st-two-list-selection.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/*
 * © 2017 Stratio Big Data Inc., Sucursal en España.
 *
 * This software is licensed under the Apache License, Version 2.0.
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the terms of the License for more details.
 *
 * SPDX-License-Identifier: Apache-2.0.
 */
import { EventEmitter } from '@angular/core';
import * as _ from 'lodash';
import { StFilterList } from '../pipes/search-filter/search-filter.pipe';
var StTwoListSelection = /** @class */ (function () {
    function StTwoListSelection(_cd) {
        this._cd = _cd;
        // Work Lists
        this.copyAllElement = [];
        this.copySelectedElements = [];
        this.allSearch = '';
        this.hasCheckboxAllList = false;
        this.hasCheckboxSelectedList = false;
        this.numItemsSelectedAll = new EventEmitter();
        this.numItemsSelectedSelected = new EventEmitter();
        this.searchBy = 'name';
        this.searchOverAll = new EventEmitter();
        this.searchOverSelected = new EventEmitter();
        this.selectedSearch = '';
    }
    // Check selected element
    // Check selected element
    /**
     * @param {?} selection
     * @return {?}
     */
    StTwoListSelection.prototype.onSelectAllElement = 
    // Check selected element
    /**
     * @param {?} selection
     * @return {?}
     */
    function (selection) {
        if (this.hasCheckboxAllList && selection.itemAll) {
            this.copyAllElement = this.applySelectAll(this.copyAllElement, this.allSearch, selection.selected);
            selection.selected = !selection.selected;
        }
        else {
            if (this.canSelect(selection, this.copyAllElement)) {
                selection.selected = !selection.selected;
            }
        }
        this.copyAllElement = _.cloneDeep(this.copyAllElement);
        if (this.copyAllElement) {
            this.numItemsSelectedAll.emit(this.getNumItemsSelected(this.copyAllElement));
        }
    };
    // Check selected element
    // Check selected element
    /**
     * @param {?} selection
     * @return {?}
     */
    StTwoListSelection.prototype.onSelectSelectedElement = 
    // Check selected element
    /**
     * @param {?} selection
     * @return {?}
     */
    function (selection) {
        if (this.hasCheckboxSelectedList && selection.itemAll) {
            this.copySelectedElements = this.applySelectAll(this.copySelectedElements, this.selectedSearch, selection.selected);
            selection.selected = !selection.selected;
        }
        else {
            if (this.canSelect(selection, this.copySelectedElements)) {
                selection.selected = !selection.selected;
            }
        }
        this.copySelectedElements = _.cloneDeep(this.copySelectedElements);
        if (this.copySelectedElements) {
            this.numItemsSelectedSelected.emit(this.getNumItemsSelected(this.copySelectedElements));
        }
    };
    // Update search filter
    // Update search filter
    /**
     * @param {?} search
     * @return {?}
     */
    StTwoListSelection.prototype.onSearchOverAll = 
    // Update search filter
    /**
     * @param {?} search
     * @return {?}
     */
    function (search) {
        this.allSearch = search.text;
        this.searchOverAll.emit(this.allSearch);
    };
    // Update search filter
    // Update search filter
    /**
     * @param {?} search
     * @return {?}
     */
    StTwoListSelection.prototype.onSearchOverSelected = 
    // Update search filter
    /**
     * @param {?} search
     * @return {?}
     */
    function (search) {
        this.selectedSearch = search.text;
        this.searchOverSelected.emit(this.selectedSearch);
    };
    // Move from all to selected
    // Move from all to selected
    /**
     * @param {?} event
     * @return {?}
     */
    StTwoListSelection.prototype.onMoveToSelected = 
    // Move from all to selected
    /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        /** @type {?} */
        var ids = this.getIdsToMove(this.copyAllElement);
        /** @type {?} */
        var result = this.moveIdsFromAllToSelected(this.originalAll, this.originalSelected, ids);
        this.emitter.emit(result);
        if (this.itemAll)
            this.itemAll.selected = false;
    };
    // Remove from selected
    // Remove from selected
    /**
     * @param {?} event
     * @return {?}
     */
    StTwoListSelection.prototype.onMoveToAll = 
    // Remove from selected
    /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        /** @type {?} */
        var ids = this.getIdsToMove(this.copySelectedElements);
        /** @type {?} */
        var result = this.removeIdsFromSelected(this.originalSelected, ids);
        this.emitter.emit(result);
        if (this.itemAll)
            this.itemAll.selected = false;
    };
    // Move all to selected
    // Move all to selected
    /**
     * @param {?} event
     * @return {?}
     */
    StTwoListSelection.prototype.onMoveAllToSelected = 
    // Move all to selected
    /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        this.emitter.emit(_.cloneDeep(this.originalAll.filter((/**
         * @param {?} elem
         * @return {?}
         */
        function (elem) { return !elem.disabled; }))));
    };
    // Remove All from selected
    // Remove All from selected
    /**
     * @param {?} event
     * @return {?}
     */
    StTwoListSelection.prototype.onMoveAllToAll = 
    // Remove All from selected
    /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        this.emitter.emit([]);
    };
    /**
     * @param {?} all
     * @param {?} selected
     * @param {?} changeEmitter
     * @param {?} sorted
     * @param {?=} hasCheckboxAllList
     * @param {?=} hasCheckboxSelectedList
     * @param {?=} itemAll
     * @return {?}
     */
    StTwoListSelection.prototype.init = /**
     * @param {?} all
     * @param {?} selected
     * @param {?} changeEmitter
     * @param {?} sorted
     * @param {?=} hasCheckboxAllList
     * @param {?=} hasCheckboxSelectedList
     * @param {?=} itemAll
     * @return {?}
     */
    function (all, selected, changeEmitter, sorted, hasCheckboxAllList, hasCheckboxSelectedList, itemAll) {
        if (hasCheckboxAllList === void 0) { hasCheckboxAllList = false; }
        if (hasCheckboxSelectedList === void 0) { hasCheckboxSelectedList = false; }
        if (itemAll === void 0) { itemAll = (/** @type {?} */ (null)); }
        this.emitter = changeEmitter;
        this.sortLists = sorted;
        this.originalAll = all;
        this.originalSelected = selected;
        this.generateWorkLists();
        this.hasCheckboxAllList = hasCheckboxAllList;
        this.hasCheckboxSelectedList = hasCheckboxSelectedList;
        this.itemAll = itemAll;
    };
    /**
     * @param {?} changes
     * @param {?} allList
     * @param {?} selectedList
     * @return {?}
     */
    StTwoListSelection.prototype.checkChanges = /**
     * @param {?} changes
     * @param {?} allList
     * @param {?} selectedList
     * @return {?}
     */
    function (changes, allList, selectedList) {
        if (changes[allList] !== undefined) {
            this.originalAll = changes[allList].currentValue;
        }
        if (changes[selectedList] !== undefined) {
            this.originalSelected = changes[selectedList].currentValue;
        }
        if (changes[allList] !== undefined || changes[selectedList] !== undefined) {
            this.generateWorkLists();
        }
    };
    /**
     * @private
     * @param {?} twoList
     * @param {?} elemSearched
     * @param {?} allSelected
     * @return {?}
     */
    StTwoListSelection.prototype.applySelectAll = /**
     * @private
     * @param {?} twoList
     * @param {?} elemSearched
     * @param {?} allSelected
     * @return {?}
     */
    function (twoList, elemSearched, allSelected) {
        /** @type {?} */
        var filterPipe = new StFilterList();
        /** @type {?} */
        var filteredBySearch = filterPipe.transform(twoList, 'name', elemSearched);
        if (elemSearched) {
            twoList.forEach((/**
             * @param {?} elem
             * @return {?}
             */
            function (elem) {
                if (!elem.disabled) {
                    elem.selected = false;
                    if (filteredBySearch.find((/**
                     * @param {?} filtered
                     * @return {?}
                     */
                    function (filtered) { return filtered.name === elem.name; })) !== undefined) {
                        elem.selected = !allSelected;
                    }
                }
            }));
            this._cd.markForCheck();
        }
        else {
            twoList = _.cloneDeep(this.changeSelectedItemList(twoList, !allSelected));
        }
        return twoList;
    };
    /**
     * @private
     * @param {?} list
     * @param {?} selected
     * @return {?}
     */
    StTwoListSelection.prototype.changeSelectedItemList = /**
     * @private
     * @param {?} list
     * @param {?} selected
     * @return {?}
     */
    function (list, selected) {
        return _.cloneDeep(list).map((/**
         * @param {?} elem
         * @return {?}
         */
        function (elem) {
            elem.selected = (!elem.disabled) ? selected : elem.selected;
            return elem;
        }));
    };
    /**
     * @private
     * @param {?} list
     * @return {?}
     */
    StTwoListSelection.prototype.getNumItemsSelected = /**
     * @private
     * @param {?} list
     * @return {?}
     */
    function (list) {
        /** @type {?} */
        var reducer = (/**
         * @param {?} accumulator
         * @param {?} currentValue
         * @return {?}
         */
        function (accumulator, currentValue) {
            return (currentValue.selected) ? accumulator + 1 : accumulator + 0;
        });
        return list.reduce(reducer, 0);
    };
    /**
     * @private
     * @return {?}
     */
    StTwoListSelection.prototype.generateWorkLists = /**
     * @private
     * @return {?}
     */
    function () {
        this.copyAllElement = this.copyLists(this.originalAll, this.copyAllElement, this.originalSelected);
        this.copySelectedElements = this.copyLists(this.originalSelected, this.copySelectedElements);
        if (this.sortLists !== undefined && (this.sortLists === 'id' || this.sortLists === 'name')) {
            this.copyAllElement = this.orderListBy(this.copyAllElement, this.sortLists);
            this.copySelectedElements = this.orderListBy(this.copySelectedElements, this.sortLists);
        }
    };
    /**
     * @private
     * @param {?} newValue
     * @param {?} prevState
     * @param {?=} without
     * @return {?}
     */
    StTwoListSelection.prototype.copyLists = /**
     * @private
     * @param {?} newValue
     * @param {?} prevState
     * @param {?=} without
     * @return {?}
     */
    function (newValue, prevState, without) {
        /** @type {?} */
        var list = _.cloneDeep(newValue);
        if (without && without.length > 0) {
            list = _.differenceBy(list, without, 'id');
        }
        /** @type {?} */
        var selected = this.getIdsToMove(prevState);
        this.searchAndSelect(list, selected);
        return list;
    };
    /**
     * @private
     * @param {?} list
     * @param {?} selected
     * @return {?}
     */
    StTwoListSelection.prototype.searchAndSelect = /**
     * @private
     * @param {?} list
     * @param {?} selected
     * @return {?}
     */
    function (list, selected) {
        /** @type {?} */
        var i = 0;
        /** @type {?} */
        var actualIndex = -1;
        while (selected.length > 0 && i < list.length) {
            actualIndex = selected.indexOf(list[i].id);
            if (actualIndex > -1) {
                list[i].selected = true;
                selected.splice(actualIndex, 1);
                actualIndex = -1;
            }
            i++;
        }
    };
    /**
     * @private
     * @param {?} list
     * @return {?}
     */
    StTwoListSelection.prototype.getIdsToMove = /**
     * @private
     * @param {?} list
     * @return {?}
     */
    function (list) {
        return list.reduce((/**
         * @param {?} prev
         * @param {?} curr
         * @return {?}
         */
        function (prev, curr) {
            if (curr.selected) {
                curr.selected = false;
                prev.push(curr.id);
            }
            return prev;
        }), []);
    };
    /**
     * @private
     * @param {?} all
     * @param {?} selected
     * @param {?} ids
     * @return {?}
     */
    StTwoListSelection.prototype.moveIdsFromAllToSelected = /**
     * @private
     * @param {?} all
     * @param {?} selected
     * @param {?} ids
     * @return {?}
     */
    function (all, selected, ids) {
        /** @type {?} */
        var items = all.filter((/**
         * @param {?} item
         * @return {?}
         */
        function (item) { return _.includes(ids, item.id); }));
        return _.xorBy(selected, _.cloneDeep(items), 'id');
    };
    /**
     * @private
     * @param {?} selected
     * @param {?} ids
     * @return {?}
     */
    StTwoListSelection.prototype.removeIdsFromSelected = /**
     * @private
     * @param {?} selected
     * @param {?} ids
     * @return {?}
     */
    function (selected, ids) {
        return selected.reduce((/**
         * @param {?} prev
         * @param {?} curr
         * @return {?}
         */
        function (prev, curr) {
            if (!_.includes(ids, curr.id)) {
                prev.push(_.clone(curr));
            }
            return prev;
        }), []);
    };
    /**
     * @private
     * @param {?} list
     * @param {?} by
     * @return {?}
     */
    StTwoListSelection.prototype.orderListBy = /**
     * @private
     * @param {?} list
     * @param {?} by
     * @return {?}
     */
    function (list, by) {
        return _.sortBy(list, by);
    };
    /**
     * @private
     * @param {?} selection
     * @param {?} list
     * @return {?}
     */
    StTwoListSelection.prototype.canSelect = /**
     * @private
     * @param {?} selection
     * @param {?} list
     * @return {?}
     */
    function (selection, list) {
        return selection && !selection.disabled && list !== undefined && list.length > 0;
    };
    return StTwoListSelection;
}());
export { StTwoListSelection };
if (false) {
    /** @type {?} */
    StTwoListSelection.prototype.originalAll;
    /** @type {?} */
    StTwoListSelection.prototype.originalSelected;
    /** @type {?} */
    StTwoListSelection.prototype.copyAllElement;
    /** @type {?} */
    StTwoListSelection.prototype.copySelectedElements;
    /** @type {?} */
    StTwoListSelection.prototype.allSearch;
    /** @type {?} */
    StTwoListSelection.prototype.hasCheckboxAllList;
    /** @type {?} */
    StTwoListSelection.prototype.hasCheckboxSelectedList;
    /** @type {?} */
    StTwoListSelection.prototype.itemAll;
    /** @type {?} */
    StTwoListSelection.prototype.numItemsSelectedAll;
    /** @type {?} */
    StTwoListSelection.prototype.numItemsSelectedSelected;
    /** @type {?} */
    StTwoListSelection.prototype.searchBy;
    /** @type {?} */
    StTwoListSelection.prototype.searchOverAll;
    /** @type {?} */
    StTwoListSelection.prototype.searchOverSelected;
    /** @type {?} */
    StTwoListSelection.prototype.selectedSearch;
    /**
     * @type {?}
     * @private
     */
    StTwoListSelection.prototype.emitter;
    /**
     * @type {?}
     * @private
     */
    StTwoListSelection.prototype.sortLists;
    /**
     * @type {?}
     * @private
     */
    StTwoListSelection.prototype._cd;
}
