// @stratio/egeo/esm5/lib/st-two-list-selection/st-two-list-selection.view.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/st-two-list-selection/st-two-list-selection.view.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
/*
 * © 2017 Stratio Big Data Inc., Sucursal en España.
 *
 * This software is licensed under the Apache License, Version 2.0.
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the terms of the License for more details.
 *
 * SPDX-License-Identifier: Apache-2.0.
 */
import { ChangeDetectorRef, ChangeDetectionStrategy, Component, EventEmitter, Input, Output, OnInit, OnChanges, SimpleChanges } from '@angular/core';
import { StEgeo, StRequired } from '../decorators/require-decorators';
var StTwoListSelectionViewComponent = /** @class */ (function () {
    function StTwoListSelectionViewComponent(_cd) {
        this._cd = _cd;
        this.editable = false;
        this.hasAllListAll = false;
        this.hasAllListSelected = false;
        this.hasSearch = true;
        this.isLoading = false;
        this.mode = 'normal';
        this.moveAllToSelectedButton = false;
        this.moveAllToAllButton = false;
        this.moveToSelectedButton = false;
        this.moveToAllButton = false;
        this.orderSelectedOptions = [];
        this.orderAllOptions = [];
        this.changeOrderAll = new EventEmitter();
        this.changeOrderSelected = new EventEmitter();
        this.moveAllToAll = new EventEmitter();
        this.moveAllToSelected = new EventEmitter();
        this.moveToAll = new EventEmitter();
        this.moveToSelected = new EventEmitter();
        this.scrollBottomAll = new EventEmitter();
        this.searchOverAll = new EventEmitter();
        this.searchOverSelected = new EventEmitter();
        this.selectAllElement = new EventEmitter();
        this.selectExtraLabelAll = new EventEmitter();
        this.selectExtraLabelSelected = new EventEmitter();
        this.selectItemNonEditable = new EventEmitter();
        this.selectSelectedElement = new EventEmitter();
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    StTwoListSelectionViewComponent.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        if (changes) {
            this.refreshButtons();
        }
    };
    /**
     * @return {?}
     */
    StTwoListSelectionViewComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.refreshButtons();
    };
    Object.defineProperty(StTwoListSelectionViewComponent.prototype, "allTitle", {
        get: /**
         * @return {?}
         */
        function () {
            return this.config && this.config.allElementsListTitle || '';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(StTwoListSelectionViewComponent.prototype, "allSubtitle", {
        get: /**
         * @return {?}
         */
        function () {
            return this.config && this.config.allElementsListSubtitle || '';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(StTwoListSelectionViewComponent.prototype, "allPlaceholder", {
        get: /**
         * @return {?}
         */
        function () {
            return this.config && this.config.allElementsSearchPlaceholder || '';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(StTwoListSelectionViewComponent.prototype, "fetchingDataText", {
        get: /**
         * @return {?}
         */
        function () {
            return this.config && this.config.fetchingDataText || '';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(StTwoListSelectionViewComponent.prototype, "menuOptionList", {
        get: /**
         * @return {?}
         */
        function () {
            return this.config && this.config.menuOptionList;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(StTwoListSelectionViewComponent.prototype, "orderPlaceholder", {
        get: /**
         * @return {?}
         */
        function () {
            return this.config && this.config.orderPlaceholder || '';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(StTwoListSelectionViewComponent.prototype, "selectedTitle", {
        get: /**
         * @return {?}
         */
        function () {
            return this.config && this.config.selectedElementsListTitle || '';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(StTwoListSelectionViewComponent.prototype, "selectedSubtitle", {
        get: /**
         * @return {?}
         */
        function () {
            return this.config && this.config.selectedElementsListSubtitle || '';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(StTwoListSelectionViewComponent.prototype, "allQaTag", {
        get: /**
         * @return {?}
         */
        function () {
            return this.qaTag + '-all-elements';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(StTwoListSelectionViewComponent.prototype, "selectedPlaceholder", {
        get: /**
         * @return {?}
         */
        function () {
            return this.config && this.config.selectedElementsSearchPlaceholder || '';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(StTwoListSelectionViewComponent.prototype, "selectedQaTag", {
        get: /**
         * @return {?}
         */
        function () {
            return this.qaTag + '-selected-elements';
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    StTwoListSelectionViewComponent.prototype.checkMoveToSelected = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var existDisabledElements = (this.allElements) ? this.allElements.filter((/**
         * @param {?} elem
         * @return {?}
         */
        function (elem) { return elem.disabled; })) : [];
        // tslint:disable-next-line:max-line-length
        this.checkMoveToSelectedValue = (this.allElements) ? (this.allElements.length === 0) || (existDisabledElements && existDisabledElements.length === this.allElements.length) : false;
        this._cd.markForCheck();
    };
    /**
     * @return {?}
     */
    StTwoListSelectionViewComponent.prototype.checkMoveToAll = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var existDisabledElements = (this.selectedElements) ? this.selectedElements.filter((/**
         * @param {?} elem
         * @return {?}
         */
        function (elem) { return elem.disabled; })) : [];
        // tslint:disable-next-line:max-line-length
        this.checkMoveToAllValue = (this.selectedElements) ? (this.selectedElements.length === 0) || (existDisabledElements && existDisabledElements.length === this.selectedElements.length) : false;
        this._cd.markForCheck();
    };
    /**
     * @return {?}
     */
    StTwoListSelectionViewComponent.prototype.refreshButtons = /**
     * @return {?}
     */
    function () {
        this.checkMoveToSelected();
        this.checkMoveToAll();
    };
    StTwoListSelectionViewComponent.ctorParameters = function () { return [
        { type: ChangeDetectorRef }
    ]; };
    StTwoListSelectionViewComponent.decorators = [
        { type: Component, args: [{
                    selector: 'st-two-list-selection-view',
                    template: "<!--\n\n    \u00A9 2017 Stratio Big Data Inc., Sucursal en Espa\u00F1a.\n\n    This software is licensed under the Apache License, Version 2.0.\n    This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;\n    without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.\n    See the terms of the License for more details.\n\n    SPDX-License-Identifier: Apache-2.0.\n\n-->\n<div class=\"st-two-list sth-two-list\">\n   <!--ALL ELEMENTS-->\n   <list-selection *ngIf=\"editable\" class=\"list sth-two-list__list sth-two-list__all-list\" [list]=\"allElements\" [itemAll]=\"itemAll\" [title]=\"allTitle\" [subtitle]=\"allSubtitle\" [searchPlaceholder]=\"allPlaceholder\" [orderPlaceholder]=\"orderPlaceholder\"\n       [qaTag]=\"allQaTag\" [editable]=\"editable\" [showSearchNumber]=\"showSearchNumber\" (search)=\"searchOverAll.emit($event)\" (selectItem)=\"selectAllElement.emit($event)\" [hasAllList]=\"hasAllListAll\" [important]=\"false\" [hasSearch]=\"hasSearch\"\n       (selectExtraLabel)=\"selectExtraLabelAll.emit($event)\" [orderOptions]=\"orderAllOptions\" (changeOrder)=\"changeOrderAll.emit($event)\" [mode]=\"mode\" (scrollBottom)=\"scrollBottomAll.emit($event)\" [isLoading]=\"isLoading\" [fetchingDataText]=\"fetchingDataText\">\n   </list-selection>\n\n   <!--SELECTORS-->\n   <div class=\"pick-buttons\" *ngIf=\"editable\">\n      <button *ngIf=\"moveAllToSelectedButton\" class=\"button button-toolbar\" type=\"button\" title=\"Move all to selected\" (click)=\"moveAllToSelected.emit($event)\" [attr.id]=\"qaTag + '-left-all-button'\"><i class=\"icon-fast-forward\"></i></button>\n      <button class=\"button button-toolbar\" type=\"button\" title=\"Move to selected\" (click)=\"moveToSelected.emit($event)\" [attr.id]=\"qaTag + '-left-button'\" [ngClass]=\"{'empty': checkMoveToSelectedValue }\" [disabled]=\"!moveToAllButton\"><i class=\"icon-arrow2_right\" ></i></button>\n      <button class=\"button button-toolbar\" type=\"button\" title=\"Move to all\" (click)=\"moveToAll.emit($event)\" [attr.id]=\"qaTag + '-right-button'\" [ngClass]=\"{'empty': checkMoveToAllValue }\" [disabled]=\"!moveToSelectedButton\"><i class=\"icon-arrow2_left\"></i></button>\n      <button *ngIf=\"moveAllToAllButton\" class=\"button button-toolbar\" type=\"button\" title=\"Move all to all\" (click)=\"moveAllToAll.emit($event)\" [attr.id]=\"qaTag + '-right-all-button'\"><i class=\"icon-rewind\"></i></button>\n   </div>\n\n   <!--SELECTED ELEMENTS-->\n   <list-selection class=\"list sth-two-list__list sth-two-list__selected-list\" [ngClass]=\"{'not-editable': !editable}\" [list]=\"selectedElements\" [itemAll]=\"itemAll\" [title]=\"selectedTitle\" [subtitle]=\"selectedSubtitle\"\n       [searchPlaceholder]=\"selectedPlaceholder\" [orderPlaceholder]=\"orderPlaceholder\" [showSearchNumber]=\"showSearchNumber\" [qaTag]=\"selectedQaTag\" [editable]=\"editable\" [menuOptionList]=\"menuOptionList\" (search)=\"searchOverSelected.emit($event)\" (selectItem)=\"selectSelectedElement.emit($event)\"\n       [important]=\"true\" [hasSearch]=\"hasSearch\" [hasAllList]=\"hasAllListSelected\" (selectExtraLabel)=\"selectExtraLabelSelected.emit($event)\" (selectItemNonEditable)=\"selectItemNonEditable.emit($event)\" [orderOptions]=\"orderSelectedOptions\" (changeOrder)=\"changeOrderSelected.emit($event)\" [mode]=\"mode\">\n   </list-selection>\n</div>\n",
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    styles: ["@charset \"UTF-8\";:host{display:flex;flex:1;flex-grow:1;max-height:100%;min-height:100%}.st-two-list{width:100%;display:flex;justify-content:space-between;max-height:100%}.list{flex-grow:1}.not-editable{width:100%}.pick-buttons{display:flex;flex-direction:column;align-items:center;width:73px;min-width:135px;max-width:135px;margin:auto}.pick-buttons button{margin-bottom:10px}.sth-two-list__all-list,.sth-two-list__selected-list{border-radius:4px;border:1px solid #dbdbdb}button.button-toolbar{width:40px;height:40px;border-radius:4px;background-color:#fff;border:1px solid #ececec;color:#707070}button.button-toolbar.empty{background-color:#ececec;color:#cfcfcf}"]
                }] }
    ];
    /** @nocollapse */
    StTwoListSelectionViewComponent.ctorParameters = function () { return [
        { type: ChangeDetectorRef }
    ]; };
    StTwoListSelectionViewComponent.propDecorators = {
        qaTag: [{ type: Input }],
        selectedElements: [{ type: Input }],
        allElements: [{ type: Input }],
        config: [{ type: Input }],
        editable: [{ type: Input }],
        hasAllListAll: [{ type: Input }],
        hasAllListSelected: [{ type: Input }],
        hasSearch: [{ type: Input }],
        isLoading: [{ type: Input }],
        itemAll: [{ type: Input }],
        mode: [{ type: Input }],
        moveAllToSelectedButton: [{ type: Input }],
        moveAllToAllButton: [{ type: Input }],
        moveToSelectedButton: [{ type: Input }],
        moveToAllButton: [{ type: Input }],
        orderSelectedOptions: [{ type: Input }],
        orderAllOptions: [{ type: Input }],
        showSearchNumber: [{ type: Input }],
        changeOrderAll: [{ type: Output }],
        changeOrderSelected: [{ type: Output }],
        moveAllToAll: [{ type: Output }],
        moveAllToSelected: [{ type: Output }],
        moveToAll: [{ type: Output }],
        moveToSelected: [{ type: Output }],
        scrollBottomAll: [{ type: Output }],
        searchOverAll: [{ type: Output }],
        searchOverSelected: [{ type: Output }],
        selectAllElement: [{ type: Output }],
        selectExtraLabelAll: [{ type: Output }],
        selectExtraLabelSelected: [{ type: Output }],
        selectItemNonEditable: [{ type: Output }],
        selectSelectedElement: [{ type: Output }]
    };
    tslib_1.__decorate([
        StRequired(),
        tslib_1.__metadata("design:type", String)
    ], StTwoListSelectionViewComponent.prototype, "qaTag", void 0);
    tslib_1.__decorate([
        StRequired(),
        tslib_1.__metadata("design:type", Array)
    ], StTwoListSelectionViewComponent.prototype, "selectedElements", void 0);
    StTwoListSelectionViewComponent = tslib_1.__decorate([
        StEgeo(),
        tslib_1.__metadata("design:paramtypes", [ChangeDetectorRef])
    ], StTwoListSelectionViewComponent);
    return StTwoListSelectionViewComponent;
}());
export { StTwoListSelectionViewComponent };
if (false) {
    /** @type {?} */
    StTwoListSelectionViewComponent.prototype.qaTag;
    /** @type {?} */
    StTwoListSelectionViewComponent.prototype.selectedElements;
    /** @type {?} */
    StTwoListSelectionViewComponent.prototype.allElements;
    /** @type {?} */
    StTwoListSelectionViewComponent.prototype.config;
    /** @type {?} */
    StTwoListSelectionViewComponent.prototype.editable;
    /** @type {?} */
    StTwoListSelectionViewComponent.prototype.hasAllListAll;
    /** @type {?} */
    StTwoListSelectionViewComponent.prototype.hasAllListSelected;
    /** @type {?} */
    StTwoListSelectionViewComponent.prototype.hasSearch;
    /** @type {?} */
    StTwoListSelectionViewComponent.prototype.isLoading;
    /** @type {?} */
    StTwoListSelectionViewComponent.prototype.itemAll;
    /** @type {?} */
    StTwoListSelectionViewComponent.prototype.mode;
    /** @type {?} */
    StTwoListSelectionViewComponent.prototype.moveAllToSelectedButton;
    /** @type {?} */
    StTwoListSelectionViewComponent.prototype.moveAllToAllButton;
    /** @type {?} */
    StTwoListSelectionViewComponent.prototype.moveToSelectedButton;
    /** @type {?} */
    StTwoListSelectionViewComponent.prototype.moveToAllButton;
    /** @type {?} */
    StTwoListSelectionViewComponent.prototype.orderSelectedOptions;
    /** @type {?} */
    StTwoListSelectionViewComponent.prototype.orderAllOptions;
    /** @type {?} */
    StTwoListSelectionViewComponent.prototype.showSearchNumber;
    /** @type {?} */
    StTwoListSelectionViewComponent.prototype.changeOrderAll;
    /** @type {?} */
    StTwoListSelectionViewComponent.prototype.changeOrderSelected;
    /** @type {?} */
    StTwoListSelectionViewComponent.prototype.moveAllToAll;
    /** @type {?} */
    StTwoListSelectionViewComponent.prototype.moveAllToSelected;
    /** @type {?} */
    StTwoListSelectionViewComponent.prototype.moveToAll;
    /** @type {?} */
    StTwoListSelectionViewComponent.prototype.moveToSelected;
    /** @type {?} */
    StTwoListSelectionViewComponent.prototype.scrollBottomAll;
    /** @type {?} */
    StTwoListSelectionViewComponent.prototype.searchOverAll;
    /** @type {?} */
    StTwoListSelectionViewComponent.prototype.searchOverSelected;
    /** @type {?} */
    StTwoListSelectionViewComponent.prototype.selectAllElement;
    /** @type {?} */
    StTwoListSelectionViewComponent.prototype.selectExtraLabelAll;
    /** @type {?} */
    StTwoListSelectionViewComponent.prototype.selectExtraLabelSelected;
    /** @type {?} */
    StTwoListSelectionViewComponent.prototype.selectItemNonEditable;
    /** @type {?} */
    StTwoListSelectionViewComponent.prototype.selectSelectedElement;
    /** @type {?} */
    StTwoListSelectionViewComponent.prototype.checkMoveToSelectedValue;
    /** @type {?} */
    StTwoListSelectionViewComponent.prototype.checkMoveToAllValue;
    /**
     * @type {?}
     * @private
     */
    StTwoListSelectionViewComponent.prototype._cd;
}
