// @stratio/egeo/esm5/lib/st-two-list-selection/st-two-list-selection.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/st-two-list-selection/st-two-list-selection.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
/*
 * © 2017 Stratio Big Data Inc., Sucursal en España.
 *
 * This software is licensed under the Apache License, Version 2.0.
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the terms of the License for more details.
 *
 * SPDX-License-Identifier: Apache-2.0.
 */
import { ChangeDetectionStrategy, ChangeDetectorRef, Component, EventEmitter, Input, OnChanges, OnInit, Output, SimpleChanges } from '@angular/core';
import { StEgeo, StRequired } from '../decorators/require-decorators';
import { StTwoListSelection } from './st-two-list-selection';
var StTwoListSelectionComponent = /** @class */ (function (_super) {
    tslib_1.__extends(StTwoListSelectionComponent, _super);
    function StTwoListSelectionComponent(cd) {
        var _this = _super.call(this, cd) || this;
        _this.cd = cd;
        _this.editable = false;
        _this.hasAllListAll = false;
        _this.hasAllListSelected = false;
        _this.hasSearch = true;
        _this.isLoading = false;
        _this.mode = 'normal';
        _this.moveAllToSelectedButton = true;
        _this.moveAllToAllButton = true;
        _this.moveToSelectedButton = true;
        _this.moveToAllButton = true;
        _this.orderAllOptions = [];
        _this.orderSelectedOptions = [];
        _this.sortBy = 'id';
        _this.change = new EventEmitter();
        _this.changeOrderAll = new EventEmitter();
        _this.changeOrderSelected = new EventEmitter();
        _this.numItemsSelectedAll = new EventEmitter();
        _this.numItemsSelectedSelected = new EventEmitter();
        _this.scrollBottomAll = new EventEmitter();
        _this.searchOverAll = new EventEmitter();
        _this.searchOverSelected = new EventEmitter();
        _this.selectedElementsChange = new EventEmitter();
        _this.selectExtraLabelAll = new EventEmitter();
        _this.selectExtraLabelSelected = new EventEmitter();
        _this.selectItemNonEditable = new EventEmitter();
        return _this;
    }
    /**
     * @return {?}
     */
    StTwoListSelectionComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.init(this.allElements, this.selectedElements, this.selectedElementsChange, this.sortBy, this.hasAllListAll, this.hasAllListSelected, this.itemAll);
    };
    /**
     * @param {?} changes
     * @return {?}
     */
    StTwoListSelectionComponent.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        this.checkChanges(changes, 'allElements', 'selectedElements');
        this.change.emit();
    };
    Object.defineProperty(StTwoListSelectionComponent.prototype, "allList", {
        get: /**
         * @return {?}
         */
        function () {
            return this.copyAllElement;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(StTwoListSelectionComponent.prototype, "selectedList", {
        get: /**
         * @return {?}
         */
        function () {
            return this.copySelectedElements;
        },
        enumerable: true,
        configurable: true
    });
    StTwoListSelectionComponent.ctorParameters = function () { return [
        { type: ChangeDetectorRef }
    ]; };
    StTwoListSelectionComponent.decorators = [
        { type: Component, args: [{
                    selector: 'st-two-list-selection',
                    template: "\n      <st-two-list-selection-view\n         [allElements]=\"allList | stFilterList:searchBy:allSearch\"\n         [selectedElements]=\"selectedList | stFilterList:searchBy:selectedSearch\"\n         [config]=\"config\"\n         [editable]=\"editable\"\n         [hasAllListAll]=\"hasAllListAll\"\n         [hasAllListSelected]=\"hasAllListSelected\"\n         [itemAll]=\"itemAll\"\n         [qaTag]=\"qaTag\"\n         [isLoading]=\"isLoading\"\n         [hasSearch]=\"hasSearch\"\n         [orderAllOptions]=\"orderAllOptions\"\n         [orderSelectedOptions]=\"orderSelectedOptions\"\n         [mode]=\"mode\"\n         [moveAllToSelectedButton]=\"moveAllToSelectedButton\"\n         [moveAllToAllButton]=\"moveAllToAllButton\"\n         [moveToAllButton]=\"moveToAllButton\"\n         [moveToSelectedButton]=\"moveToSelectedButton\"\n         [showSearchNumber]=\"showSearchNumber\"\n\n         (selectAllElement)=\"onSelectAllElement($event)\"\n         (selectSelectedElement)=\"onSelectSelectedElement($event)\"\n         (searchOverAll)=\"onSearchOverAll($event)\"\n         (searchOverSelected)=\"onSearchOverSelected($event)\"\n         (moveToSelected)=\"onMoveToSelected($event)\"\n         (moveToAll)=\"onMoveToAll($event)\"\n         (moveAllToSelected)=\"onMoveAllToSelected($event)\"\n         (moveAllToAll)=\"onMoveAllToAll($event)\"\n         (numItemsSelectedAll)=\"numItemsSelectedAll.emit($event)\"\n         (numItemsSelectedSelected)=\"numItemsSelectedSelected.emit($event)\"\n         (selectExtraLabelAll)=\"selectExtraLabelAll.emit($event)\"\n         (selectExtraLabelSelected)=\"selectExtraLabelAll.emit($event)\"\n         (selectItemNonEditable)=\"selectItemNonEditable.emit($event)\"\n         (changeOrderAll)=\"changeOrderAll.emit($event)\"\n         (changeOrderSelected)=\"changeOrderSelected.emit($event)\"\n         (scrollBottomAll)=\"scrollBottomAll.emit($event)\"\n      ></st-two-list-selection-view>\n   ",
                    changeDetection: ChangeDetectionStrategy.OnPush
                }] }
    ];
    /** @nocollapse */
    StTwoListSelectionComponent.ctorParameters = function () { return [
        { type: ChangeDetectorRef }
    ]; };
    StTwoListSelectionComponent.propDecorators = {
        allElements: [{ type: Input }],
        config: [{ type: Input }],
        editable: [{ type: Input }],
        hasAllListAll: [{ type: Input }],
        hasAllListSelected: [{ type: Input }],
        hasSearch: [{ type: Input }],
        isLoading: [{ type: Input }],
        itemAll: [{ type: Input }],
        mode: [{ type: Input }],
        moveAllToSelectedButton: [{ type: Input }],
        moveAllToAllButton: [{ type: Input }],
        moveToSelectedButton: [{ type: Input }],
        moveToAllButton: [{ type: Input }],
        orderAllOptions: [{ type: Input }],
        orderSelectedOptions: [{ type: Input }],
        qaTag: [{ type: Input }],
        selectedElements: [{ type: Input }],
        showSearchNumber: [{ type: Input }],
        sortBy: [{ type: Input }],
        change: [{ type: Output }],
        changeOrderAll: [{ type: Output }],
        changeOrderSelected: [{ type: Output }],
        numItemsSelectedAll: [{ type: Output }],
        numItemsSelectedSelected: [{ type: Output }],
        scrollBottomAll: [{ type: Output }],
        searchOverAll: [{ type: Output }],
        searchOverSelected: [{ type: Output }],
        selectedElementsChange: [{ type: Output }],
        selectExtraLabelAll: [{ type: Output }],
        selectExtraLabelSelected: [{ type: Output }],
        selectItemNonEditable: [{ type: Output }]
    };
    tslib_1.__decorate([
        StRequired('editable'),
        tslib_1.__metadata("design:type", Array)
    ], StTwoListSelectionComponent.prototype, "allElements", void 0);
    tslib_1.__decorate([
        StRequired(),
        tslib_1.__metadata("design:type", String)
    ], StTwoListSelectionComponent.prototype, "qaTag", void 0);
    tslib_1.__decorate([
        StRequired(),
        tslib_1.__metadata("design:type", Array)
    ], StTwoListSelectionComponent.prototype, "selectedElements", void 0);
    StTwoListSelectionComponent = tslib_1.__decorate([
        StEgeo(),
        tslib_1.__metadata("design:paramtypes", [ChangeDetectorRef])
    ], StTwoListSelectionComponent);
    return StTwoListSelectionComponent;
}(StTwoListSelection));
export { StTwoListSelectionComponent };
if (false) {
    /** @type {?} */
    StTwoListSelectionComponent.prototype.allElements;
    /** @type {?} */
    StTwoListSelectionComponent.prototype.config;
    /** @type {?} */
    StTwoListSelectionComponent.prototype.editable;
    /** @type {?} */
    StTwoListSelectionComponent.prototype.hasAllListAll;
    /** @type {?} */
    StTwoListSelectionComponent.prototype.hasAllListSelected;
    /** @type {?} */
    StTwoListSelectionComponent.prototype.hasSearch;
    /** @type {?} */
    StTwoListSelectionComponent.prototype.isLoading;
    /** @type {?} */
    StTwoListSelectionComponent.prototype.itemAll;
    /** @type {?} */
    StTwoListSelectionComponent.prototype.mode;
    /** @type {?} */
    StTwoListSelectionComponent.prototype.moveAllToSelectedButton;
    /** @type {?} */
    StTwoListSelectionComponent.prototype.moveAllToAllButton;
    /** @type {?} */
    StTwoListSelectionComponent.prototype.moveToSelectedButton;
    /** @type {?} */
    StTwoListSelectionComponent.prototype.moveToAllButton;
    /** @type {?} */
    StTwoListSelectionComponent.prototype.orderAllOptions;
    /** @type {?} */
    StTwoListSelectionComponent.prototype.orderSelectedOptions;
    /** @type {?} */
    StTwoListSelectionComponent.prototype.qaTag;
    /** @type {?} */
    StTwoListSelectionComponent.prototype.selectedElements;
    /** @type {?} */
    StTwoListSelectionComponent.prototype.showSearchNumber;
    /** @type {?} */
    StTwoListSelectionComponent.prototype.sortBy;
    /** @type {?} */
    StTwoListSelectionComponent.prototype.change;
    /** @type {?} */
    StTwoListSelectionComponent.prototype.changeOrderAll;
    /** @type {?} */
    StTwoListSelectionComponent.prototype.changeOrderSelected;
    /** @type {?} */
    StTwoListSelectionComponent.prototype.numItemsSelectedAll;
    /** @type {?} */
    StTwoListSelectionComponent.prototype.numItemsSelectedSelected;
    /** @type {?} */
    StTwoListSelectionComponent.prototype.scrollBottomAll;
    /** @type {?} */
    StTwoListSelectionComponent.prototype.searchOverAll;
    /** @type {?} */
    StTwoListSelectionComponent.prototype.searchOverSelected;
    /** @type {?} */
    StTwoListSelectionComponent.prototype.selectedElementsChange;
    /** @type {?} */
    StTwoListSelectionComponent.prototype.selectExtraLabelAll;
    /** @type {?} */
    StTwoListSelectionComponent.prototype.selectExtraLabelSelected;
    /** @type {?} */
    StTwoListSelectionComponent.prototype.selectItemNonEditable;
    /**
     * @type {?}
     * @private
     */
    StTwoListSelectionComponent.prototype.cd;
}
