// @stratio/egeo/esm5/lib/st-select/st-check-validations.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/st-select/st-check-validations.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/*
 * © 2017 Stratio Big Data Inc., Sucursal en España.
 *
 * This software is licensed under the Apache License, Version 2.0.
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the terms of the License for more details.
 *
 * SPDX-License-Identifier: Apache-2.0.
 */
import { Directive, Input } from '@angular/core';
import { NgControl } from '@angular/forms';
var StCheckValidationsDirective = /** @class */ (function () {
    function StCheckValidationsDirective(control) {
        this.control = control;
    }
    /**
     * @param {?} fn
     * @return {?}
     */
    StCheckValidationsDirective.prototype.registerOnChange = /**
     * @param {?} fn
     * @return {?}
     */
    function (fn) {
        this.onChange = fn;
    };
    /**
     * @return {?}
     */
    StCheckValidationsDirective.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.sub = this.control.statusChanges.subscribe((/**
         * @return {?}
         */
        function () { return _this.notifyErrorsIfExists(); }));
    };
    /**
     * @return {?}
     */
    StCheckValidationsDirective.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        if (this.sub) {
            this.sub.unsubscribe();
        }
    };
    // TODO implement error check and notify correct error
    // TODO implement error check and notify correct error
    /**
     * @private
     * @return {?}
     */
    StCheckValidationsDirective.prototype.notifyErrorsIfExists = 
    // TODO implement error check and notify correct error
    /**
     * @private
     * @return {?}
     */
    function () {
        if (this.control.valid && this.onChange) {
            this.onChange(null);
        }
        else if (this.onChange) {
            this.onChange(this.errorMessage);
        }
    };
    StCheckValidationsDirective.decorators = [
        { type: Directive, args: [{ selector: '[stCheckValidations]' },] }
    ];
    /** @nocollapse */
    StCheckValidationsDirective.ctorParameters = function () { return [
        { type: NgControl }
    ]; };
    StCheckValidationsDirective.propDecorators = {
        errorMessage: [{ type: Input }]
    };
    return StCheckValidationsDirective;
}());
export { StCheckValidationsDirective };
if (false) {
    /** @type {?} */
    StCheckValidationsDirective.prototype.errorMessage;
    /** @type {?} */
    StCheckValidationsDirective.prototype.onChange;
    /**
     * @type {?}
     * @private
     */
    StCheckValidationsDirective.prototype.sub;
    /**
     * @type {?}
     * @private
     */
    StCheckValidationsDirective.prototype.control;
}
