// @stratio/egeo/esm5/lib/st-select/st-select.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/st-select/st-select.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/*
 * © 2017 Stratio Big Data Inc., Sucursal en España.
 *
 * This software is licensed under the Apache License, Version 2.0.
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the terms of the License for more details.
 *
 * SPDX-License-Identifier: Apache-2.0.
 */
import { ChangeDetectionStrategy, ChangeDetectorRef, Component, ElementRef, EventEmitter, forwardRef, HostBinding, Injector, Input, Output, ViewChild } from '@angular/core';
import { FormControl, NG_VALUE_ACCESSOR } from '@angular/forms';
import { cloneDeep as _cloneDeep, flatten as _flatten, has as _has } from 'lodash';
import { StCheckValidationsDirective } from './st-check-validations';
import { StDropDownMenuItem } from '../st-dropdown-menu/st-dropdown-menu.interface';
var StSelectComponent = /** @class */ (function () {
    function StSelectComponent(_selectElement, _injector, _cd) {
        this._selectElement = _selectElement;
        this._injector = _injector;
        this._cd = _cd;
        this.placeholder = '';
        this.name = '';
        this.label = '';
        this.tooltip = null;
        this.selected = undefined;
        this.itemsBeforeScroll = 8;
        this.enabledSearcher = false;
        this.isLoading = false;
        this.forceValidations = false;
        this.placeholderSearch = 'Search...';
        this.keyBoardMove = false;
        this.expand = new EventEmitter();
        this.select = new EventEmitter();
        this.scrollAtBottom = new EventEmitter();
        this.search = new EventEmitter();
        this.expandedMenu = false;
        this.searchInput = new FormControl();
        this.inputFormControl = new FormControl();
        this._inputHTMLElement = undefined;
        this._isDisabled = false;
        this._options = [];
    }
    // TODO: MOVE THIS TO FORM-BASE
    // TODO: MOVE THIS TO FORM-BASE
    /**
     * @param {?} errorMessage
     * @return {?}
     */
    StSelectComponent.prototype.notifyError = 
    // TODO: MOVE THIS TO FORM-BASE
    /**
     * @param {?} errorMessage
     * @return {?}
     */
    function (errorMessage) {
        this.errorMessage = errorMessage;
    };
    Object.defineProperty(StSelectComponent.prototype, "disabled", {
        get: /**
         * @return {?}
         */
        function () {
            return this._isDisabled;
        },
        /*
         ****** getters && setters ******
         */
        set: /*
            ****** getters && setters ******
            */
        /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._isDisabled = value;
            this._cd.markForCheck();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(StSelectComponent.prototype, "options", {
        get: /**
         * @return {?}
         */
        function () {
            return this._options;
        },
        set: /**
         * @param {?} options
         * @return {?}
         */
        function (options) {
            this._options = _cloneDeep(options);
            /** @type {?} */
            var selectedItem = this.findByProperty('selected', true);
            this.removeAllSelected();
            if (selectedItem) {
                this.selected = selectedItem;
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(StSelectComponent.prototype, "selectedValue", {
        get: /**
         * @return {?}
         */
        function () {
            return this.selected && this.selected.label ? this.selected.label : '';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(StSelectComponent.prototype, "disableValue", {
        get: /**
         * @return {?}
         */
        function () {
            return this._isDisabled === true ? '' : null;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(StSelectComponent.prototype, "selectId", {
        get: /**
         * @return {?}
         */
        function () {
            /** @type {?} */
            var select = this._selectElement.nativeElement;
            return select.getAttribute('id') !== null ? select.id : null;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(StSelectComponent.prototype, "inputId", {
        get: /**
         * @return {?}
         */
        function () {
            return this.selectId !== null ? this.selectId + "-input" : null;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(StSelectComponent.prototype, "labelId", {
        get: /**
         * @return {?}
         */
        function () {
            return this.selectId !== null ? this.selectId + "-label" : null;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(StSelectComponent.prototype, "optionsId", {
        get: /**
         * @return {?}
         */
        function () {
            return this.selectId !== null ? this.selectId + "-options" : null;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(StSelectComponent.prototype, "inputName", {
        get: /**
         * @return {?}
         */
        function () {
            return this.name && this.name.length > 0 ? this.name : null;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(StSelectComponent.prototype, "hasLabel", {
        get: /**
         * @return {?}
         */
        function () {
            return this.label !== undefined && this.label !== null && this.label.length > 0;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    StSelectComponent.prototype.showError = /**
     * @return {?}
     */
    function () {
        return this.errorMessage && this.errorMessage.length && (this.inputFormControl.touched || this.forceValidations) && !this._isDisabled;
    };
    /*
     ****** Control value accessor && validate methods ******
     */
    /*
        ****** Control value accessor && validate methods ******
        */
    /**
     * @return {?}
     */
    StSelectComponent.prototype.onSearch = /*
        ****** Control value accessor && validate methods ******
        */
    /**
     * @return {?}
     */
    function () {
        this.search.emit(this.searchInput.value);
        this._cd.markForCheck();
    };
    /**
     * @param {?} event
     * @return {?}
     */
    StSelectComponent.prototype.onSearchClick = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        event.preventDefault();
        event.stopPropagation();
    };
    // Set the function to be called when the control receives a change event.
    // Set the function to be called when the control receives a change event.
    /**
     * @param {?} fn
     * @return {?}
     */
    StSelectComponent.prototype.registerOnChange = 
    // Set the function to be called when the control receives a change event.
    /**
     * @param {?} fn
     * @return {?}
     */
    function (fn) {
        this.onChange = fn;
    };
    // Set the function to be called when the control receives a touch event.
    // Set the function to be called when the control receives a touch event.
    /**
     * @param {?} fn
     * @return {?}
     */
    StSelectComponent.prototype.registerOnTouched = 
    // Set the function to be called when the control receives a touch event.
    /**
     * @param {?} fn
     * @return {?}
     */
    function (fn) {
        this.onTouched = fn;
    };
    /**
     * @param {?} disabled
     * @return {?}
     */
    StSelectComponent.prototype.setDisabledState = /**
     * @param {?} disabled
     * @return {?}
     */
    function (disabled) {
        this._isDisabled = disabled;
        this._cd.markForCheck();
    };
    // Write a new value to the element.
    // Write a new value to the element.
    /**
     * @param {?} newValue
     * @return {?}
     */
    StSelectComponent.prototype.writeValue = 
    // Write a new value to the element.
    /**
     * @param {?} newValue
     * @return {?}
     */
    function (newValue) {
        if (!this.selected || this.selected.value !== newValue) {
            this.selected = this.findByProperty('value', newValue);
            this._cd.markForCheck();
        }
    };
    /*
     ****** Component methods ******
     */
    /*
        ****** Component methods ******
        */
    /**
     * @return {?}
     */
    StSelectComponent.prototype.ngAfterViewInit = /*
        ****** Component methods ******
        */
    /**
     * @return {?}
     */
    function () {
        this._inputHTMLElement = this.inputElement.nativeElement;
        /** @type {?} */
        var directive = this._injector.get(StCheckValidationsDirective, null);
        if (directive) {
            directive.registerOnChange(this.notifyError.bind(this));
        }
    };
    /**
     * @return {?}
     */
    StSelectComponent.prototype.onButtonClick = /**
     * @return {?}
     */
    function () {
        if (!this._isDisabled) {
            this.toggleButton();
            this.expandedMenu ? this._inputHTMLElement.focus() : this._inputHTMLElement.blur();
        }
    };
    /**
     * @param {?} event
     * @return {?}
     */
    StSelectComponent.prototype.onButtonKeyPress = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        if (event.code === 'Enter') {
            this.toggleButton();
        }
    };
    /**
     * @return {?}
     */
    StSelectComponent.prototype.clearSearchInput = /**
     * @return {?}
     */
    function () {
        this.searchInput.setValue('');
        this.search.emit('');
    };
    /**
     * @return {?}
     */
    StSelectComponent.prototype.createResetButton = /**
     * @return {?}
     */
    function () {
        return this.default !== undefined && ((!this.selected && this.inputFormControl.touched) || (this.selected && this.selected.value !== this.default));
    };
    /**
     * @return {?}
     */
    StSelectComponent.prototype.resetToDefault = /**
     * @return {?}
     */
    function () {
        this.writeValue(this.default);
        this.select.emit(this.default);
        if (this.onChange) {
            this.onChange(this.default);
        }
        this._cd.markForCheck();
    };
    /**
     * @return {?}
     */
    StSelectComponent.prototype.onClickOutside = /**
     * @return {?}
     */
    function () {
        this.expandedMenu = false;
        this.expand.emit(this.expandedMenu); // Notify expand change
    };
    /**
     * @param {?} option
     * @return {?}
     */
    StSelectComponent.prototype.onChangeOption = /**
     * @param {?} option
     * @return {?}
     */
    function (option) {
        this.selected = option && option.value !== undefined ? option : undefined;
        /** @type {?} */
        var value = option && option.value !== undefined ? option.value : undefined;
        if (this.onChange) {
            this.onChange(value);
        }
        if (this.onTouched) {
            this.onTouched();
        }
        this.select.emit(value);
        if (value || (option && option.hasOwnProperty('value') && !option.value)) {
            this.onClickOutside();
        }
        this._cd.markForCheck();
    };
    /**
     * @return {?}
     */
    StSelectComponent.prototype.onScrollAtBottom = /**
     * @return {?}
     */
    function () {
        this.scrollAtBottom.emit();
    };
    /*
     ****** Util component methods ******
     */
    // Search element by property in option list
    /*
        ****** Util component methods ******
        */
    // Search element by property in option list
    /**
     * @private
     * @param {?} propName
     * @param {?} propValue
     * @return {?}
     */
    StSelectComponent.prototype.findByProperty = /*
        ****** Util component methods ******
        */
    // Search element by property in option list
    /**
     * @private
     * @param {?} propName
     * @param {?} propValue
     * @return {?}
     */
    function (propName, propValue) {
        if (this.isStDropdownItemList(this.options)) {
            return this.options.find((/**
             * @param {?} item
             * @return {?}
             */
            function (item) { return _has(item, propName) && item[propName] === propValue; }));
        }
        else if (this.isStDropdownGroupList(this.options)) {
            return _flatten(this.options.map((/**
             * @param {?} group
             * @return {?}
             */
            function (group) { return group.items; }))).find((/**
             * @param {?} item
             * @return {?}
             */
            function (item) { return _has(item, propName) && item[propName] === propValue; }));
        }
    };
    // Check if options are a instance of StDropDownMenuItem[]
    // Check if options are a instance of StDropDownMenuItem[]
    /**
     * @private
     * @param {?} items
     * @return {?}
     */
    StSelectComponent.prototype.isStDropdownItemList = 
    // Check if options are a instance of StDropDownMenuItem[]
    /**
     * @private
     * @param {?} items
     * @return {?}
     */
    function (items) {
        return this.options && this.options.length > 0 && !_has(((/** @type {?} */ (items)))[0], 'items');
    };
    // Check if options are a instance of StDropDownMenuGroup[]
    // Check if options are a instance of StDropDownMenuGroup[]
    /**
     * @private
     * @param {?} items
     * @return {?}
     */
    StSelectComponent.prototype.isStDropdownGroupList = 
    // Check if options are a instance of StDropDownMenuGroup[]
    /**
     * @private
     * @param {?} items
     * @return {?}
     */
    function (items) {
        return this.options && this.options.length > 0 && _has(((/** @type {?} */ (items)))[0], 'items');
    };
    /**
     * @private
     * @return {?}
     */
    StSelectComponent.prototype.toggleButton = /**
     * @private
     * @return {?}
     */
    function () {
        this.expandedMenu = !this.expandedMenu;
        this.expand.emit(this.expandedMenu); // Notify expand change
        this._cd.markForCheck();
    };
    // TODO: Remove when remove from StDropDownMenuItem model the selected property
    // TODO: Remove when remove from StDropDownMenuItem model the selected property
    /**
     * @private
     * @return {?}
     */
    StSelectComponent.prototype.removeAllSelected = 
    // TODO: Remove when remove from StDropDownMenuItem model the selected property
    /**
     * @private
     * @return {?}
     */
    function () {
        if (this.isStDropdownItemList(this.options)) {
            return this.options.forEach((/**
             * @param {?} item
             * @return {?}
             */
            function (item) {
                if (item.selected) {
                    delete item.selected;
                }
            }));
        }
        else if (this.isStDropdownGroupList(this.options)) {
            this.options.forEach((/**
             * @param {?} group
             * @return {?}
             */
            function (group) { return group.items.forEach((/**
             * @param {?} item
             * @return {?}
             */
            function (item) {
                if (item.selected) {
                    delete item.selected;
                }
            })); }));
        }
    };
    StSelectComponent.decorators = [
        { type: Component, args: [{
                    selector: 'st-select',
                    template: "<!--\n\n    \u00A9 2017 Stratio Big Data Inc., Sucursal en Espa\u00F1a.\n\n    This software is licensed under the Apache License, Version 2.0.\n    This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;\n    without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.\n    See the terms of the License for more details.\n\n    SPDX-License-Identifier: Apache-2.0.\n\n-->\n<st-dropdown-menu\n   class=\"st-select-menu\"\n   [items]=\"options\"\n   [attr.id]=\"optionsId\"\n   [active]=\"expandedMenu\"\n   [placement]=\"'bottom-start'\"\n   [keyBoardMove]=\"keyBoardMove\"\n   [selectedItem]=\"selected\"\n   (change)=\"onChangeOption($event)\"\n   (scrollAtBottom)=\"onScrollAtBottom()\"\n   [isLoading]=\"isLoading\"\n   [itemsBeforeScroll]=\"itemsBeforeScroll\"\n>\n   <div (click)=\"onButtonClick()\" (keypress)=\"onButtonKeyPress($event)\" #button class=\"button-container\"\n        (clickOutside)=\"onClickOutside()\">\n      <div>\n         <label *ngIf=\"hasLabel\" [ngClass]=\"{disabled: disabled, error: showError()}\"\n                st-label\n                for=\"inputId\"\n                class=\"st-label\"\n                [attr.title]=\"tooltip\"\n                [attr.id]=\"labelId\"\n                [attr.disabled]=\"disableValue\">{{label}}</label>\n         <div class=\"input-container\">\n            <input\n               #input\n               [formControl]=\"inputFormControl\"\n               autocomplete=\"off\"\n               aria-haspopup=\"true\"\n               class=\"st-input\"\n               type=\"text\"\n               [readonly]=\"true\"\n               [attr.aria-expanded]=\"expandedMenu\"\n               [attr.disabled]=\"disableValue\"\n               [attr.id]=\"inputId\"\n               [attr.name]=\"inputName\"\n               [placeholder]=\"placeholder\"\n               [value]=\"selectedValue\"\n\n            />\n            <span *ngIf=\"createResetButton()\" class=\"st-form-control-reset-button\" (click)=\"resetToDefault()\"\n                  [style.opacity]=\"expandedMenu ? 1 : 0\">\n               <i class=\"icon-reload2\"></i>\n            </span>\n            <i class=\"icon-arrow2_down arrow\" [attr.disabled]=\"disableValue\"></i>\n         </div>\n      </div>\n   </div>\n   <div dropdown-header *ngIf=\"enabledSearcher\"\n        (click)=\"onSearchClick($event)\"\n        class=\"search-input\">\n      <input\n         id=\"searchInput\"\n         autocomplete=\"off\"\n         aria-haspopup=\"true\"\n         class=\"st-input\"\n         type=\"text\"\n         [attr.aria-expanded]=\"expandedMenu\"\n         [attr.disabled]=\"disableValue\"\n         [attr.id]=\"searchInput\"\n         [placeholder]=\"placeholderSearch\"\n         [formControl]=\"searchInput\"\n         (focus)=\"enabledSearcher\"\n         (input)=\"onSearch()\"\n      />\n      <i class=\"st-search-icon sth-search-icon icon-search\" tabindex=\"0\" *ngIf=\"!searchInput.value?.length\"></i>\n      <i class=\"st-search-icon sth-search-icon icon-cross\" tabindex=\"0\" *ngIf=\"searchInput.value?.length\"\n         (click)=\"clearSearchInput()\" (keyup.enter)=\"clearSearchInput()\"></i>\n   </div>\n\n</st-dropdown-menu>\n\n<div class=\"st-input-error-layout\" *ngIf=\"showError()\">\n   <span class=\"st-input-error-message\">{{errorMessage}}</span>\n</div>\n",
                    host: {
                        'class': 'st-select'
                    },
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    providers: [
                        { provide: NG_VALUE_ACCESSOR, useExisting: forwardRef((/**
                             * @return {?}
                             */
                            function () { return StSelectComponent; })), multi: true }
                    ],
                    styles: ["@charset \"UTF-8\";.button-container{position:relative}.arrow{position:absolute;right:8px;top:8px;padding:5px 5px 5px 9px;cursor:pointer}.st-select-menu{position:relative}.input-container{position:relative;display:inline-block;width:100%}.search-input{margin:15px}.st-search-icon{position:absolute;right:28px;top:28px;cursor:pointer}.st-search-icon:focus{outline:0}"]
                }] }
    ];
    /** @nocollapse */
    StSelectComponent.ctorParameters = function () { return [
        { type: ElementRef },
        { type: Injector },
        { type: ChangeDetectorRef }
    ]; };
    StSelectComponent.propDecorators = {
        placeholder: [{ type: Input }],
        name: [{ type: Input }],
        label: [{ type: Input }],
        tooltip: [{ type: Input }],
        errorMessage: [{ type: Input }],
        selected: [{ type: Input }],
        default: [{ type: Input }],
        itemsBeforeScroll: [{ type: Input }],
        enabledSearcher: [{ type: Input }],
        isLoading: [{ type: Input }],
        forceValidations: [{ type: Input }],
        placeholderSearch: [{ type: Input }],
        keyBoardMove: [{ type: Input }],
        expand: [{ type: Output }],
        select: [{ type: Output }],
        scrollAtBottom: [{ type: Output }],
        search: [{ type: Output }],
        inputElement: [{ type: ViewChild, args: ['input', { static: false },] }],
        buttonElement: [{ type: ViewChild, args: ['button', { static: false },] }],
        expandedMenu: [{ type: HostBinding, args: ['class.st-select-opened',] }],
        disabled: [{ type: Input }],
        options: [{ type: Input }]
    };
    return StSelectComponent;
}());
export { StSelectComponent };
if (false) {
    /** @type {?} */
    StSelectComponent.prototype.placeholder;
    /** @type {?} */
    StSelectComponent.prototype.name;
    /** @type {?} */
    StSelectComponent.prototype.label;
    /** @type {?} */
    StSelectComponent.prototype.tooltip;
    /** @type {?} */
    StSelectComponent.prototype.errorMessage;
    /** @type {?} */
    StSelectComponent.prototype.selected;
    /** @type {?} */
    StSelectComponent.prototype.default;
    /** @type {?} */
    StSelectComponent.prototype.itemsBeforeScroll;
    /** @type {?} */
    StSelectComponent.prototype.enabledSearcher;
    /** @type {?} */
    StSelectComponent.prototype.isLoading;
    /** @type {?} */
    StSelectComponent.prototype.forceValidations;
    /** @type {?} */
    StSelectComponent.prototype.placeholderSearch;
    /** @type {?} */
    StSelectComponent.prototype.keyBoardMove;
    /** @type {?} */
    StSelectComponent.prototype.expand;
    /** @type {?} */
    StSelectComponent.prototype.select;
    /** @type {?} */
    StSelectComponent.prototype.scrollAtBottom;
    /** @type {?} */
    StSelectComponent.prototype.search;
    /** @type {?} */
    StSelectComponent.prototype.inputElement;
    /** @type {?} */
    StSelectComponent.prototype.buttonElement;
    /** @type {?} */
    StSelectComponent.prototype.expandedMenu;
    /** @type {?} */
    StSelectComponent.prototype.searchInput;
    /** @type {?} */
    StSelectComponent.prototype.inputFormControl;
    /** @type {?} */
    StSelectComponent.prototype.onChange;
    /** @type {?} */
    StSelectComponent.prototype.onTouched;
    /**
     * @type {?}
     * @private
     */
    StSelectComponent.prototype._inputHTMLElement;
    /**
     * @type {?}
     * @private
     */
    StSelectComponent.prototype._isDisabled;
    /**
     * @type {?}
     * @private
     */
    StSelectComponent.prototype._options;
    /**
     * @type {?}
     * @private
     */
    StSelectComponent.prototype._selectElement;
    /**
     * @type {?}
     * @private
     */
    StSelectComponent.prototype._injector;
    /**
     * @type {?}
     * @private
     */
    StSelectComponent.prototype._cd;
}
