// @stratio/egeo/esm2015/lib/utils/egeo-resolver/egeo-resolve.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/utils/egeo-resolver/egeo-resolve.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/*
 * © 2017 Stratio Big Data Inc., Sucursal en España.
 *
 * This software is licensed under the Apache License, Version 2.0.
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the terms of the License for more details.
 *
 * SPDX-License-Identifier: Apache-2.0.
 */
import { Injectable } from '@angular/core';
import { has as _has, set as _set, cloneDeep as _cloneDeep, forEach as _forEach, keys as _keys, values as _values, omit as _omit } from 'lodash';
import { Observable } from 'rxjs';
import { map } from 'rxjs/operators';
export class EgeoResolveService {
    /**
     * @param {?} object
     * @param {?} key
     * @param {?=} searchedValue
     * @return {?}
     */
    getKeys(object, key, searchedValue) {
        return this.searchInDeep(object, key, searchedValue);
    }
    /**
     * @param {?} object
     * @param {?} resolved
     * @return {?}
     */
    setKeys(object, resolved) {
        resolved.forEach((/**
         * @param {?} element
         * @return {?}
         */
        (element) => _set(object, element.path, element.resolved)));
    }
    /**
     * @param {?} object
     * @param {?} translateService
     * @return {?}
     */
    translate(object, translateService) {
        /** @type {?} */
        let keys = this.getKeys(object, 'translate');
        // If not found translateable elements, return the same because if not, translate service broke on try to translate an empty array.
        if (keys && keys.length > 0) {
            /** @type {?} */
            let toTranslate = this.extractTranslationKeys(keys);
            return translateService.get(toTranslate).pipe(// return object translation
            map((/**
             * @param {?} translation
             * @return {?}
             */
            (translation) => this.remapObjectWithTranslations(translation, keys, object))));
        }
        else {
            return Observable.create((/**
             * @param {?} observer
             * @return {?}
             */
            (observer) => {
                observer.next(object);
                observer.complete();
            }));
        }
    }
    /**
     * @param {?} keys
     * @param {?} translateService
     * @return {?}
     */
    translateArrayOfKeys(keys, translateService) {
        return translateService.get(keys).pipe(map((/**
         * @param {?} translation
         * @return {?}
         */
        (translation) => this.remapArrayWithTranslations(translation, keys))));
    }
    /**
     * @private
     * @param {?} translations
     * @param {?} originalArray
     * @return {?}
     */
    remapArrayWithTranslations(translations, originalArray) {
        return originalArray.map((/**
         * @param {?} key
         * @return {?}
         */
        (key) => translations[key] ? translations[key] : key));
    }
    /**
     * @private
     * @param {?} translations
     * @param {?} resolverKeys
     * @param {?} object
     * @return {?}
     */
    remapObjectWithTranslations(translations, resolverKeys, object) {
        /** @type {?} */
        let newObj = _cloneDeep(object);
        if (translations || _keys(translations).length > 0) {
            _forEach(resolverKeys, (/**
             * @param {?} resolvKey
             * @return {?}
             */
            (resolvKey) => {
                _set(newObj, resolvKey.path, this.getTranslationFromTranslatedKey(translations, resolvKey));
            }));
        }
        return newObj;
    }
    /**
     * @private
     * @param {?} translations
     * @param {?} resolverKey
     * @return {?}
     */
    getTranslationFromTranslatedKey(translations, resolverKey) {
        /** @type {?} */
        let translationElementKey = resolverKey && resolverKey.toResolve && resolverKey.toResolve.key ? resolverKey.toResolve.key : '';
        return translations[translationElementKey] ? translations[translationElementKey] : translationElementKey;
    }
    /**
     * @private
     * @param {?} list
     * @return {?}
     */
    extractTranslationKeys(list) {
        return list.map((/**
         * @param {?} element
         * @return {?}
         */
        (element) => (/** @type {?} */ (_values(_omit(element.toResolve, 'translate'))[0]))));
    }
    /**
     * @private
     * @param {?} object
     * @param {?} key
     * @param {?} searchedValue
     * @param {?=} path
     * @return {?}
     */
    searchInDeep(object, key, searchedValue, path = '') {
        /** @type {?} */
        let result = [];
        if (_has(object, key)) { // If we found key, return object.
            if (searchedValue !== undefined) {
                if (object[key] === searchedValue) {
                    result.push({ path, toResolve: object });
                }
            }
            else {
                result.push({ path, toResolve: object });
            }
        }
        /** @type {?} */
        let i = 0;
        _forEach(object, (/**
         * @param {?} value
         * @param {?} objKey
         * @return {?}
         */
        (value, objKey) => {
            if (typeof value === 'object') {
                result = [...result, ...this.searchInDeep(value, key, searchedValue, this.getPath(path, object, i, objKey))];
            }
            i++;
        }));
        return result;
    }
    // Build path for future replace
    /**
     * @private
     * @param {?} actualPath
     * @param {?} obj
     * @param {?} pos
     * @param {?} key
     * @return {?}
     */
    getPath(actualPath, obj, pos, key) {
        if (this.isArray(obj)) {
            actualPath = `${actualPath}[${pos}]`;
        }
        else {
            actualPath = actualPath === '' ? key : `${actualPath}.${key}`;
        }
        return actualPath;
    }
    /**
     * @private
     * @param {?} value
     * @return {?}
     */
    isArray(value) {
        return Object.prototype.toString.call(value) === '[object Array]';
    }
}
EgeoResolveService.decorators = [
    { type: Injectable }
];
