// @stratio/egeo/esm2015/lib/pipes/search-filter/search-filter.pipe.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/pipes/search-filter/search-filter.pipe.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/*
 * © 2017 Stratio Big Data Inc., Sucursal en España.
 *
 * This software is licensed under the Apache License, Version 2.0.
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the terms of the License for more details.
 *
 * SPDX-License-Identifier: Apache-2.0.
 */
import { Pipe } from '@angular/core';
export class StFilterList {
    /**
     * @param {?} list
     * @param {?} field
     * @param {?} search
     * @return {?}
     */
    transform(list, field, search) {
        this.checkParams(field);
        if (!search) {
            return list;
        }
        if (list && list.length > 0 && field) {
            return list.filter((/**
             * @param {?} element
             * @return {?}
             */
            (element) => this.contains(element, field, search)));
        }
        else {
            return [];
        }
    }
    /**
     * @private
     * @param {?} element
     * @param {?} field
     * @param {?} search
     * @return {?}
     */
    contains(element, field, search) {
        if (this.isDefined(element) && this.isString(field) && this.isString(search) && this.isString(element[field])) {
            return element[field].toUpperCase().search(search.toUpperCase()) > -1;
        }
        return false;
    }
    /**
     * @private
     * @param {?} value
     * @return {?}
     */
    isDefined(value) {
        return value !== undefined && value !== null;
    }
    /**
     * @private
     * @param {?} value
     * @return {?}
     */
    isString(value) {
        return this.isDefined(value) && typeof value === 'string';
    }
    /**
     * @private
     * @param {?} field
     * @return {?}
     */
    checkParams(field) {
        if (!this.isString(field)) {
            throw new Error('PIPE: stFilterList: "field" is a required and string parameter');
        }
    }
}
StFilterList.decorators = [
    { type: Pipe, args: [{
                name: 'stFilterList'
            },] }
];
