// @stratio/egeo/esm5/lib/utils/egeo-resolver/egeo-resolve.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/utils/egeo-resolver/egeo-resolve.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
/*
 * © 2017 Stratio Big Data Inc., Sucursal en España.
 *
 * This software is licensed under the Apache License, Version 2.0.
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the terms of the License for more details.
 *
 * SPDX-License-Identifier: Apache-2.0.
 */
import { Injectable } from '@angular/core';
import { has as _has, set as _set, cloneDeep as _cloneDeep, forEach as _forEach, keys as _keys, values as _values, omit as _omit } from 'lodash';
import { Observable } from 'rxjs';
import { map } from 'rxjs/operators';
var EgeoResolveService = /** @class */ (function () {
    function EgeoResolveService() {
    }
    /**
     * @param {?} object
     * @param {?} key
     * @param {?=} searchedValue
     * @return {?}
     */
    EgeoResolveService.prototype.getKeys = /**
     * @param {?} object
     * @param {?} key
     * @param {?=} searchedValue
     * @return {?}
     */
    function (object, key, searchedValue) {
        return this.searchInDeep(object, key, searchedValue);
    };
    /**
     * @param {?} object
     * @param {?} resolved
     * @return {?}
     */
    EgeoResolveService.prototype.setKeys = /**
     * @param {?} object
     * @param {?} resolved
     * @return {?}
     */
    function (object, resolved) {
        resolved.forEach((/**
         * @param {?} element
         * @return {?}
         */
        function (element) { return _set(object, element.path, element.resolved); }));
    };
    /**
     * @param {?} object
     * @param {?} translateService
     * @return {?}
     */
    EgeoResolveService.prototype.translate = /**
     * @param {?} object
     * @param {?} translateService
     * @return {?}
     */
    function (object, translateService) {
        var _this = this;
        /** @type {?} */
        var keys = this.getKeys(object, 'translate');
        // If not found translateable elements, return the same because if not, translate service broke on try to translate an empty array.
        if (keys && keys.length > 0) {
            /** @type {?} */
            var toTranslate = this.extractTranslationKeys(keys);
            return translateService.get(toTranslate).pipe(// return object translation
            map((/**
             * @param {?} translation
             * @return {?}
             */
            function (translation) { return _this.remapObjectWithTranslations(translation, keys, object); })));
        }
        else {
            return Observable.create((/**
             * @param {?} observer
             * @return {?}
             */
            function (observer) {
                observer.next(object);
                observer.complete();
            }));
        }
    };
    /**
     * @param {?} keys
     * @param {?} translateService
     * @return {?}
     */
    EgeoResolveService.prototype.translateArrayOfKeys = /**
     * @param {?} keys
     * @param {?} translateService
     * @return {?}
     */
    function (keys, translateService) {
        var _this = this;
        return translateService.get(keys).pipe(map((/**
         * @param {?} translation
         * @return {?}
         */
        function (translation) { return _this.remapArrayWithTranslations(translation, keys); })));
    };
    /**
     * @private
     * @param {?} translations
     * @param {?} originalArray
     * @return {?}
     */
    EgeoResolveService.prototype.remapArrayWithTranslations = /**
     * @private
     * @param {?} translations
     * @param {?} originalArray
     * @return {?}
     */
    function (translations, originalArray) {
        return originalArray.map((/**
         * @param {?} key
         * @return {?}
         */
        function (key) { return translations[key] ? translations[key] : key; }));
    };
    /**
     * @private
     * @param {?} translations
     * @param {?} resolverKeys
     * @param {?} object
     * @return {?}
     */
    EgeoResolveService.prototype.remapObjectWithTranslations = /**
     * @private
     * @param {?} translations
     * @param {?} resolverKeys
     * @param {?} object
     * @return {?}
     */
    function (translations, resolverKeys, object) {
        var _this = this;
        /** @type {?} */
        var newObj = _cloneDeep(object);
        if (translations || _keys(translations).length > 0) {
            _forEach(resolverKeys, (/**
             * @param {?} resolvKey
             * @return {?}
             */
            function (resolvKey) {
                _set(newObj, resolvKey.path, _this.getTranslationFromTranslatedKey(translations, resolvKey));
            }));
        }
        return newObj;
    };
    /**
     * @private
     * @param {?} translations
     * @param {?} resolverKey
     * @return {?}
     */
    EgeoResolveService.prototype.getTranslationFromTranslatedKey = /**
     * @private
     * @param {?} translations
     * @param {?} resolverKey
     * @return {?}
     */
    function (translations, resolverKey) {
        /** @type {?} */
        var translationElementKey = resolverKey && resolverKey.toResolve && resolverKey.toResolve.key ? resolverKey.toResolve.key : '';
        return translations[translationElementKey] ? translations[translationElementKey] : translationElementKey;
    };
    /**
     * @private
     * @param {?} list
     * @return {?}
     */
    EgeoResolveService.prototype.extractTranslationKeys = /**
     * @private
     * @param {?} list
     * @return {?}
     */
    function (list) {
        return list.map((/**
         * @param {?} element
         * @return {?}
         */
        function (element) { return (/** @type {?} */ (_values(_omit(element.toResolve, 'translate'))[0])); }));
    };
    /**
     * @private
     * @param {?} object
     * @param {?} key
     * @param {?} searchedValue
     * @param {?=} path
     * @return {?}
     */
    EgeoResolveService.prototype.searchInDeep = /**
     * @private
     * @param {?} object
     * @param {?} key
     * @param {?} searchedValue
     * @param {?=} path
     * @return {?}
     */
    function (object, key, searchedValue, path) {
        var _this = this;
        if (path === void 0) { path = ''; }
        /** @type {?} */
        var result = [];
        if (_has(object, key)) { // If we found key, return object.
            if (searchedValue !== undefined) {
                if (object[key] === searchedValue) {
                    result.push({ path: path, toResolve: object });
                }
            }
            else {
                result.push({ path: path, toResolve: object });
            }
        }
        /** @type {?} */
        var i = 0;
        _forEach(object, (/**
         * @param {?} value
         * @param {?} objKey
         * @return {?}
         */
        function (value, objKey) {
            if (typeof value === 'object') {
                result = tslib_1.__spread(result, _this.searchInDeep(value, key, searchedValue, _this.getPath(path, object, i, objKey)));
            }
            i++;
        }));
        return result;
    };
    // Build path for future replace
    // Build path for future replace
    /**
     * @private
     * @param {?} actualPath
     * @param {?} obj
     * @param {?} pos
     * @param {?} key
     * @return {?}
     */
    EgeoResolveService.prototype.getPath = 
    // Build path for future replace
    /**
     * @private
     * @param {?} actualPath
     * @param {?} obj
     * @param {?} pos
     * @param {?} key
     * @return {?}
     */
    function (actualPath, obj, pos, key) {
        if (this.isArray(obj)) {
            actualPath = actualPath + "[" + pos + "]";
        }
        else {
            actualPath = actualPath === '' ? key : actualPath + "." + key;
        }
        return actualPath;
    };
    /**
     * @private
     * @param {?} value
     * @return {?}
     */
    EgeoResolveService.prototype.isArray = /**
     * @private
     * @param {?} value
     * @return {?}
     */
    function (value) {
        return Object.prototype.toString.call(value) === '[object Array]';
    };
    EgeoResolveService.decorators = [
        { type: Injectable }
    ];
    return EgeoResolveService;
}());
export { EgeoResolveService };
