// @stratio/egeo/esm2015/lib/st-menu/st-menu.model.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/st-menu/st-menu.model.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/*
 * © 2017 Stratio Big Data Inc., Sucursal en España.
 *
 * This software is licensed under the Apache License, Version 2.0.
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the terms of the License for more details.
 *
 * SPDX-License-Identifier: Apache-2.0.
 */
/**
 * @record
 * @template T
 */
export function StMenuModel() { }
if (false) {
    /** @type {?} */
    StMenuModel.prototype.name;
    /** @type {?} */
    StMenuModel.prototype.value;
    /** @type {?|undefined} */
    StMenuModel.prototype.icon;
    /** @type {?|undefined} */
    StMenuModel.prototype.subMenus;
    /** @type {?|undefined} */
    StMenuModel.prototype.separator;
    /** @type {?|undefined} */
    StMenuModel.prototype.status;
    /** @type {?|undefined} */
    StMenuModel.prototype.disabled;
}
/** @enum {string} */
const StMenuStatus = {
    success: "success",
    warning: "warning",
    critical: "critical",
};
export { StMenuStatus };
/** @enum {string} */
const StMenuPosition = {
    left: "left",
    right: "right",
};
export { StMenuPosition };
