// @stratio/egeo/esm5/lib/st-item-list/st-item-list.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/st-item-list/st-item-list.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
/*
 * © 2017 Stratio Big Data Inc., Sucursal en España.
 *
 * This software is licensed under the Apache License, Version 2.0.
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the terms of the License for more details.
 *
 * SPDX-License-Identifier: Apache-2.0.
 */
import { ChangeDetectionStrategy, Component, EventEmitter, Input, Output } from '@angular/core';
import { StRequired } from '../decorators/require-decorators';
var StItemListComponent = /** @class */ (function () {
    function StItemListComponent() {
        this.qaTag = '';
        this.hasSearch = false;
        this.align = 'left';
        this.theme = 'themeA';
        this.selectItem = new EventEmitter();
        this.search = new EventEmitter();
    }
    Object.defineProperty(StItemListComponent.prototype, "listTitle", {
        get: /**
         * @return {?}
         */
        function () {
            return this.config && this.config.title || '';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(StItemListComponent.prototype, "searchPlaceholder", {
        get: /**
         * @return {?}
         */
        function () {
            return this.config && this.config.searchPlaceholder || '';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(StItemListComponent.prototype, "listQaTag", {
        get: /**
         * @return {?}
         */
        function () {
            return this.qaTag || '';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(StItemListComponent.prototype, "searchQaTag", {
        get: /**
         * @return {?}
         */
        function () {
            return this.qaTag && this.qaTag + '-search' || '';
        },
        enumerable: true,
        configurable: true
    });
    StItemListComponent.decorators = [
        { type: Component, args: [{
                    selector: 'st-item-list',
                    template: "<!--\n\n    \u00A9 2017 Stratio Big Data Inc., Sucursal en Espa\u00F1a.\n\n    This software is licensed under the Apache License, Version 2.0.\n    This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;\n    without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.\n    See the terms of the License for more details.\n\n    SPDX-License-Identifier: Apache-2.0.\n\n-->\n<div class=\"st-item-list sth-item-list\" [ngClass]=\"theme\">\n   <p *ngIf=\"listTitle\" class=\"st-item-list__title sth-item-list__title\">{{listTitle}}</p>\n   <st-search *ngIf=\"hasSearch\" class=\"st-item-list__search\" [placeholder]=\"searchPlaceholder\" [qaTag]=\"searchQaTag\" (search)=\"search.emit($event)\"></st-search>\n   <item-list-scroll [list]=\"list\" [align]=\"align\" [qaTag]=\"listQaTag\" (selectItem)=\"selectItem.emit($event)\"></item-list-scroll>\n</div>\n",
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    styles: ["@charset \"UTF-8\";.st-item-list{display:flex;flex-direction:column;flex-wrap:nowrap;overflow:hidden;width:100%;height:100%;padding:30px 20px}.st-item-list__title{width:100%;margin-bottom:10px;padding:0}.st-item-list__search{margin-bottom:10px;display:block}virtual-scroll{display:block;width:100%;height:100%}"]
                }] }
    ];
    StItemListComponent.propDecorators = {
        list: [{ type: Input }],
        config: [{ type: Input }],
        qaTag: [{ type: Input }],
        hasSearch: [{ type: Input }],
        align: [{ type: Input }],
        theme: [{ type: Input }],
        selectItem: [{ type: Output }],
        search: [{ type: Output }]
    };
    tslib_1.__decorate([
        StRequired(),
        tslib_1.__metadata("design:type", Array)
    ], StItemListComponent.prototype, "list", void 0);
    return StItemListComponent;
}());
export { StItemListComponent };
if (false) {
    /** @type {?} */
    StItemListComponent.prototype.list;
    /** @type {?} */
    StItemListComponent.prototype.config;
    /** @type {?} */
    StItemListComponent.prototype.qaTag;
    /** @type {?} */
    StItemListComponent.prototype.hasSearch;
    /** @type {?} */
    StItemListComponent.prototype.align;
    /** @type {?} */
    StItemListComponent.prototype.theme;
    /** @type {?} */
    StItemListComponent.prototype.selectItem;
    /** @type {?} */
    StItemListComponent.prototype.search;
}
